{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "17941e3b-7c57-43c7-a0f7-a14a168b1ec7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import importlib\n",
    "\n",
    "#import autoreload\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "import pwd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "from tqdm import tqdm, trange\n",
    "from tqdm.contrib.concurrent import process_map\n",
    "sys.path.append('../')\n",
    "from meth import tools as mtools\n",
    "import matplotlib.pyplot as plt\n",
    "import GEOparse\n",
    "import scanpy as sc\n",
    "from scipy import sparse as sp\n",
    "# %autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84b52f3f-4e3a-476d-adf8-c8091c0669fa",
   "metadata": {},
   "source": [
    "# GSE 121690"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4e950585-8737-42ef-b64c-b7af03346ab1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwd = Path.cwd()\n",
    "# cwd = Path('/Users/tarkhov/Dropbox/Harvard/Science/Church/')\n",
    "project_dir = Path.joinpath(cwd, Path('data/GSE121690/'))\n",
    "project_id = 'GSE121690'\n",
    "project_samples_dir = 'GSE121690_RAW'\n",
    "metainfo_path = 'GSE121690_series_matrix.txt'\n",
    "gl_metainfo_path = Path.joinpath(project_dir, metainfo_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "373d9cc1-84ff-41f9-8120-16a0d9491cd1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "26-Sep-2022 14:23:02 DEBUG utils - Directory ./data/GSE121690/ already exists. Skipping.\n",
      "26-Sep-2022 14:23:02 INFO GEOparse - File already exist: using local version.\n",
      "26-Sep-2022 14:23:02 INFO GEOparse - Parsing ./data/GSE121690/GSE121690_family.soft.gz: \n",
      "26-Sep-2022 14:23:02 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "26-Sep-2022 14:23:02 DEBUG GEOparse - SERIES: GSE121690\n",
      "26-Sep-2022 14:23:02 DEBUG GEOparse - PLATFORM: GPL19057\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443369\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443370\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443371\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443372\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443373\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443374\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443375\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443376\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443377\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443378\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443379\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443380\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443381\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443382\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443383\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443384\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443385\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443386\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443387\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443388\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443389\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443390\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443391\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443392\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443393\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443394\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443395\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443396\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443397\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443398\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443399\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443400\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443401\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443402\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443403\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443404\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443405\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443406\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443407\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443408\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443409\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443410\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443411\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443412\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443413\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443414\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443415\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443416\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443417\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443418\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443419\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443420\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443421\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443422\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443423\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443424\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443425\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443426\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443427\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443428\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443429\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443430\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443431\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443432\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443433\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443434\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443435\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443436\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443437\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443438\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443439\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443440\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443441\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443442\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443443\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443444\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443445\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443446\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443447\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443448\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443449\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443450\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443451\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443452\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443453\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443454\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443455\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443456\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443457\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443458\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443459\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443460\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443461\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443462\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443463\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443464\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443465\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443466\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443467\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443468\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443469\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443470\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443471\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443472\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443473\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443474\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443475\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443476\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443477\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443478\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443479\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443480\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443481\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443482\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443483\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443484\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443485\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443486\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443487\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443488\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443489\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443490\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443491\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443492\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443493\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443494\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443495\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443496\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443497\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443498\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443499\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443500\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443501\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443502\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443503\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443504\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443505\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443506\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443507\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443508\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443509\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443510\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443511\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443512\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443513\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443514\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443515\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443516\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443517\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443518\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443519\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443520\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443521\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443522\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443523\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443524\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443525\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443526\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443527\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443528\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443529\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443530\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443531\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443532\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443533\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443534\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443535\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443536\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443537\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443538\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443539\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443540\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443541\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443542\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443543\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443544\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443545\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443546\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443547\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443548\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443549\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443550\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443551\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443552\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443553\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443554\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443555\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443556\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443557\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443558\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443559\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443560\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443561\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443562\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443563\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443564\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443565\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443566\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443567\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443568\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443569\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443570\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443571\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443572\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443573\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443574\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443575\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443576\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443577\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443578\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443579\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443580\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443581\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443582\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443583\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443584\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443585\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443586\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443587\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443588\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443589\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443590\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443591\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443592\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443593\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443594\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443595\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443596\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443597\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443598\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443599\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443600\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443601\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443602\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443603\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443604\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443605\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443606\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443607\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443608\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443609\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443610\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443611\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443612\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443613\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443614\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443615\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443616\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443617\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443618\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443619\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443620\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443621\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443622\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443623\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443624\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443625\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443626\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443627\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443628\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443629\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443630\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443631\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443632\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443633\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443634\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443635\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443636\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443637\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443638\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443639\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443640\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443641\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443642\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443643\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443644\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443645\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443646\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443647\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443648\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443649\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443650\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443651\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443652\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443653\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443654\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443655\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443656\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443657\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443658\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443659\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443660\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443661\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443662\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443663\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443664\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443665\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443666\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443667\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443668\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443669\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443670\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443671\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443672\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443673\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443674\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443675\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443676\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443677\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443678\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443679\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443680\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443681\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443682\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443683\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443684\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443685\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443686\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443687\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443688\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443689\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443690\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443691\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443692\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443693\n",
      "26-Sep-2022 14:23:03 DEBUG GEOparse - SAMPLE: GSM3443694\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443695\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443696\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443697\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443698\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443699\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443700\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443701\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443702\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443703\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443704\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443705\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443706\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443707\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443708\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443709\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443710\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443711\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443712\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443713\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443714\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443715\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443716\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443717\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443718\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443719\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443720\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443721\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443722\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443723\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443724\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443725\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443726\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443727\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443728\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443729\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443730\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443731\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443732\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443733\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443734\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443735\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443736\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443737\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443738\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443739\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443740\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443741\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443742\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443743\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443744\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443745\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443746\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443747\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443748\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443749\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443750\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443751\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443752\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443753\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443754\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443755\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443756\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443757\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443758\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443759\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443760\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443761\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443762\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443763\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443764\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443765\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443766\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443767\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443768\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443769\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443770\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443771\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443772\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443773\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443774\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443775\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443776\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443777\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443778\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443779\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443780\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443781\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443782\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443783\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443784\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443785\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443786\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443787\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443788\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443789\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443790\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443791\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443792\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443793\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443794\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443795\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443796\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443797\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443798\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443799\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443800\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443801\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443802\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443803\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443804\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443805\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443806\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443807\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443808\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443809\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443810\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443811\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443812\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443813\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443814\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443815\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443816\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443817\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443818\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443819\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443820\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443821\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443822\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443823\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443824\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443825\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443826\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443827\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443828\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443829\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443830\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443831\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443832\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443833\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443834\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443835\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443836\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443837\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443838\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443839\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443840\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443841\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443842\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443843\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443844\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443845\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443846\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443847\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443848\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443849\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443850\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443851\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443852\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443853\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443854\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443855\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443856\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443857\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443858\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443859\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443860\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443861\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443862\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443863\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443864\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443865\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443866\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443867\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443868\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443869\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443870\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443871\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443872\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443873\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443874\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443875\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443876\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443877\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443878\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443879\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443880\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443881\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443882\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443883\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443884\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443885\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443886\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443887\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443888\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443889\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443890\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443891\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443892\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443893\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443894\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443895\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443896\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443897\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443898\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443899\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443900\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443901\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443902\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443903\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443904\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443905\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443906\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443907\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443908\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443909\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443910\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443911\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443912\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443913\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443914\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443915\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443916\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443917\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443918\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443919\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443920\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443921\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443922\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443923\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443924\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443925\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443926\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443927\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443928\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443929\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443930\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443931\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443932\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443933\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443934\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443935\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443936\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443937\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443938\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443939\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443940\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443941\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443942\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443943\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443944\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443945\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443946\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443947\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443948\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443949\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443950\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443951\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443952\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443953\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443954\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443955\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443956\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443957\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443958\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443959\n",
      "26-Sep-2022 14:23:04 DEBUG GEOparse - SAMPLE: GSM3443960\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443961\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443962\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443963\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443964\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443965\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443966\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443967\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443968\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443969\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443970\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443971\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443972\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443973\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443974\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443975\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443976\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443977\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443978\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443979\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443980\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443981\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443982\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443983\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443984\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443985\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443986\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443987\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443988\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443989\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443990\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443991\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443992\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443993\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443994\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443995\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443996\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443997\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443998\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3443999\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444000\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444001\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444002\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444003\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444004\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444005\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444006\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444007\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444008\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444009\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444010\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444011\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444012\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444013\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444014\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444015\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444016\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444017\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444018\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444019\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444020\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444021\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444022\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444023\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444024\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444025\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444026\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444027\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444028\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444029\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444030\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444031\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444032\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444033\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444034\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444035\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444036\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444037\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444038\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444039\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444040\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444041\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444042\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444043\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444044\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444045\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444046\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444047\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444048\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444049\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444050\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444051\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444052\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444053\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444054\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444055\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444056\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444057\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444058\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444059\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444060\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444061\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444062\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444063\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444064\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444065\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444066\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444067\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444068\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444069\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444070\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444071\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444072\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444073\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444074\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444075\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444076\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444077\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444078\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444079\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444080\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444081\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444082\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444083\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444084\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444085\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444086\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444087\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444088\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444089\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444090\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444091\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444092\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444093\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444094\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444095\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444096\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444097\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444098\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444099\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444100\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444101\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444102\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444103\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444104\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444105\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444106\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444107\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444108\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444109\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444110\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444111\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444112\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444113\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444114\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444115\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444116\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444117\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444118\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444119\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444120\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444121\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444122\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444123\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444124\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444125\n",
      "26-Sep-2022 14:23:05 DEBUG GEOparse - SAMPLE: GSM3444126\n"
     ]
    }
   ],
   "source": [
    "import GEOparse\n",
    "gse = GEOparse.get_GEO(geo=\"GSE121690\", destdir=\"./data/GSE121690/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fb3e81c5-fdc6-4124-9ad8-b46af0a48826",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>geo_accession</th>\n",
       "      <th>status</th>\n",
       "      <th>submission_date</th>\n",
       "      <th>last_update_date</th>\n",
       "      <th>type</th>\n",
       "      <th>channel_count</th>\n",
       "      <th>source_name_ch1</th>\n",
       "      <th>organism_ch1</th>\n",
       "      <th>taxid_ch1</th>\n",
       "      <th>...</th>\n",
       "      <th>contact_country</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>relation</th>\n",
       "      <th>supplementary_file_1</th>\n",
       "      <th>supplementary_file_2</th>\n",
       "      <th>series_id</th>\n",
       "      <th>data_row_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM3443369</th>\n",
       "      <td>E4.5-5.5_new_Plate1_A02 (scBS-Seq)</td>\n",
       "      <td>GSM3443369</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3443370</th>\n",
       "      <td>E4.5-5.5_new_Plate1_A03 (scBS-Seq)</td>\n",
       "      <td>GSM3443370</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3443371</th>\n",
       "      <td>E4.5-5.5_new_Plate1_A04 (scBS-Seq)</td>\n",
       "      <td>GSM3443371</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3443372</th>\n",
       "      <td>E4.5-5.5_new_Plate1_A05 (scBS-Seq)</td>\n",
       "      <td>GSM3443372</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3443373</th>\n",
       "      <td>E4.5-5.5_new_Plate1_A07 (scBS-Seq)</td>\n",
       "      <td>GSM3443373</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         title geo_accession  \\\n",
       "GSM3443369  E4.5-5.5_new_Plate1_A02 (scBS-Seq)    GSM3443369   \n",
       "GSM3443370  E4.5-5.5_new_Plate1_A03 (scBS-Seq)    GSM3443370   \n",
       "GSM3443371  E4.5-5.5_new_Plate1_A04 (scBS-Seq)    GSM3443371   \n",
       "GSM3443372  E4.5-5.5_new_Plate1_A05 (scBS-Seq)    GSM3443372   \n",
       "GSM3443373  E4.5-5.5_new_Plate1_A07 (scBS-Seq)    GSM3443373   \n",
       "\n",
       "                           status submission_date last_update_date type  \\\n",
       "GSM3443369  Public on Sep 30 2019     Oct 23 2018      Sep 30 2019  SRA   \n",
       "GSM3443370  Public on Sep 30 2019     Oct 23 2018      Sep 30 2019  SRA   \n",
       "GSM3443371  Public on Sep 30 2019     Oct 23 2018      Sep 30 2019  SRA   \n",
       "GSM3443372  Public on Sep 30 2019     Oct 23 2018      Sep 30 2019  SRA   \n",
       "GSM3443373  Public on Sep 30 2019     Oct 23 2018      Sep 30 2019  SRA   \n",
       "\n",
       "           channel_count source_name_ch1  organism_ch1 taxid_ch1  ...  \\\n",
       "GSM3443369             1    Early embryo  Mus musculus     10090  ...   \n",
       "GSM3443370             1    Early embryo  Mus musculus     10090  ...   \n",
       "GSM3443371             1    Early embryo  Mus musculus     10090  ...   \n",
       "GSM3443372             1    Early embryo  Mus musculus     10090  ...   \n",
       "GSM3443373             1    Early embryo  Mus musculus     10090  ...   \n",
       "\n",
       "           contact_country      instrument_model library_selection  \\\n",
       "GSM3443369  United Kingdom  Illumina NextSeq 500            RANDOM   \n",
       "GSM3443370  United Kingdom  Illumina NextSeq 500            RANDOM   \n",
       "GSM3443371  United Kingdom  Illumina NextSeq 500            RANDOM   \n",
       "GSM3443372  United Kingdom  Illumina NextSeq 500            RANDOM   \n",
       "GSM3443373  United Kingdom  Illumina NextSeq 500            RANDOM   \n",
       "\n",
       "           library_source library_strategy  \\\n",
       "GSM3443369        genomic    Bisulfite-Seq   \n",
       "GSM3443370        genomic    Bisulfite-Seq   \n",
       "GSM3443371        genomic    Bisulfite-Seq   \n",
       "GSM3443372        genomic    Bisulfite-Seq   \n",
       "GSM3443373        genomic    Bisulfite-Seq   \n",
       "\n",
       "                                                     relation  \\\n",
       "GSM3443369  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM3443370  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM3443371  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM3443372  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM3443373  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "\n",
       "                                         supplementary_file_1  \\\n",
       "GSM3443369  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443370  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443371  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443372  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443373  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "\n",
       "                                         supplementary_file_2  \\\n",
       "GSM3443369  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443370  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443371  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443372  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443373  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "\n",
       "                      series_id data_row_count  \n",
       "GSM3443369  GSE121690,GSE121708              0  \n",
       "GSM3443370  GSE121690,GSE121708              0  \n",
       "GSM3443371  GSE121690,GSE121708              0  \n",
       "GSM3443372  GSE121690,GSE121708              0  \n",
       "GSM3443373  GSE121690,GSE121708              0  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = gse.phenotype_data.copy()\n",
    "df_meta.shape\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3ec9f679-1be1-4cd6-a566-a262aef8c198",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1516, 35)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>geo_accession</th>\n",
       "      <th>status</th>\n",
       "      <th>submission_date</th>\n",
       "      <th>last_update_date</th>\n",
       "      <th>type</th>\n",
       "      <th>channel_count</th>\n",
       "      <th>source_name_ch1</th>\n",
       "      <th>organism_ch1</th>\n",
       "      <th>taxid_ch1</th>\n",
       "      <th>...</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>relation</th>\n",
       "      <th>supplementary_file_1</th>\n",
       "      <th>supplementary_file_2</th>\n",
       "      <th>series_id</th>\n",
       "      <th>data_row_count</th>\n",
       "      <th>Data Type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name PerCov</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM3443369_percentage</th>\n",
       "      <td>E4.5-5.5_new_Plate1_A02 (scBS-Seq)</td>\n",
       "      <td>GSM3443369</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3443370_percentage</th>\n",
       "      <td>E4.5-5.5_new_Plate1_A03 (scBS-Seq)</td>\n",
       "      <td>GSM3443370</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3443371_percentage</th>\n",
       "      <td>E4.5-5.5_new_Plate1_A04 (scBS-Seq)</td>\n",
       "      <td>GSM3443371</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3443372_percentage</th>\n",
       "      <td>E4.5-5.5_new_Plate1_A05 (scBS-Seq)</td>\n",
       "      <td>GSM3443372</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3443373_percentage</th>\n",
       "      <td>E4.5-5.5_new_Plate1_A07 (scBS-Seq)</td>\n",
       "      <td>GSM3443373</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    title geo_accession  \\\n",
       "Name PerCov                                                               \n",
       "GSM3443369_percentage  E4.5-5.5_new_Plate1_A02 (scBS-Seq)    GSM3443369   \n",
       "GSM3443370_percentage  E4.5-5.5_new_Plate1_A03 (scBS-Seq)    GSM3443370   \n",
       "GSM3443371_percentage  E4.5-5.5_new_Plate1_A04 (scBS-Seq)    GSM3443371   \n",
       "GSM3443372_percentage  E4.5-5.5_new_Plate1_A05 (scBS-Seq)    GSM3443372   \n",
       "GSM3443373_percentage  E4.5-5.5_new_Plate1_A07 (scBS-Seq)    GSM3443373   \n",
       "\n",
       "                                      status submission_date last_update_date  \\\n",
       "Name PerCov                                                                     \n",
       "GSM3443369_percentage  Public on Sep 30 2019     Oct 23 2018      Sep 30 2019   \n",
       "GSM3443370_percentage  Public on Sep 30 2019     Oct 23 2018      Sep 30 2019   \n",
       "GSM3443371_percentage  Public on Sep 30 2019     Oct 23 2018      Sep 30 2019   \n",
       "GSM3443372_percentage  Public on Sep 30 2019     Oct 23 2018      Sep 30 2019   \n",
       "GSM3443373_percentage  Public on Sep 30 2019     Oct 23 2018      Sep 30 2019   \n",
       "\n",
       "                      type channel_count source_name_ch1  organism_ch1  \\\n",
       "Name PerCov                                                              \n",
       "GSM3443369_percentage  SRA             1    Early embryo  Mus musculus   \n",
       "GSM3443370_percentage  SRA             1    Early embryo  Mus musculus   \n",
       "GSM3443371_percentage  SRA             1    Early embryo  Mus musculus   \n",
       "GSM3443372_percentage  SRA             1    Early embryo  Mus musculus   \n",
       "GSM3443373_percentage  SRA             1    Early embryo  Mus musculus   \n",
       "\n",
       "                      taxid_ch1  ...      instrument_model library_selection  \\\n",
       "Name PerCov                      ...                                           \n",
       "GSM3443369_percentage     10090  ...  Illumina NextSeq 500            RANDOM   \n",
       "GSM3443370_percentage     10090  ...  Illumina NextSeq 500            RANDOM   \n",
       "GSM3443371_percentage     10090  ...  Illumina NextSeq 500            RANDOM   \n",
       "GSM3443372_percentage     10090  ...  Illumina NextSeq 500            RANDOM   \n",
       "GSM3443373_percentage     10090  ...  Illumina NextSeq 500            RANDOM   \n",
       "\n",
       "                      library_source library_strategy  \\\n",
       "Name PerCov                                             \n",
       "GSM3443369_percentage        genomic    Bisulfite-Seq   \n",
       "GSM3443370_percentage        genomic    Bisulfite-Seq   \n",
       "GSM3443371_percentage        genomic    Bisulfite-Seq   \n",
       "GSM3443372_percentage        genomic    Bisulfite-Seq   \n",
       "GSM3443373_percentage        genomic    Bisulfite-Seq   \n",
       "\n",
       "                                                                relation  \\\n",
       "Name PerCov                                                                \n",
       "GSM3443369_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM3443370_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM3443371_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM3443372_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM3443373_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "\n",
       "                                                    supplementary_file_1  \\\n",
       "Name PerCov                                                                \n",
       "GSM3443369_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443370_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443371_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443372_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443373_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "\n",
       "                                                    supplementary_file_2  \\\n",
       "Name PerCov                                                                \n",
       "GSM3443369_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443370_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443371_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443372_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443373_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "\n",
       "                                 series_id data_row_count   Data Type  \n",
       "Name PerCov                                                            \n",
       "GSM3443369_percentage  GSE121690,GSE121708              0  percentage  \n",
       "GSM3443370_percentage  GSE121690,GSE121708              0  percentage  \n",
       "GSM3443371_percentage  GSE121690,GSE121708              0  percentage  \n",
       "GSM3443372_percentage  GSE121690,GSE121708              0  percentage  \n",
       "GSM3443373_percentage  GSE121690,GSE121708              0  percentage  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = gse.phenotype_data.copy()\n",
    "df_meta_2 = df_meta.copy()\n",
    "df_meta['Name PerCov'] = df_meta['geo_accession'].str.cat(['percentage'] * df_meta.shape[0], sep='_')\n",
    "df_meta['Data Type'] = 'percentage'\n",
    "df_meta_2['Name PerCov'] = df_meta_2['geo_accession'].str.cat(['coverage'] * df_meta.shape[0], sep='_')\n",
    "df_meta_2['Data Type'] = 'coverage'\n",
    "df_meta['Name PerCov'] = df_meta['geo_accession'].str.cat(['percentage'] * df_meta.shape[0], sep='_')\n",
    "df_meta = pd.concat((df_meta, df_meta_2), axis=0)\n",
    "df_meta.set_index('Name PerCov', inplace=True)\n",
    "print(df_meta.shape)\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "295f6a95-7f9c-49ab-8a6e-7a8521642500",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_meta['characteristics_ch1.1.developmental stage'].value_counts()\n",
    "df_meta['Age'] = df_meta['characteristics_ch1.1.developmental stage'].apply(lambda x: float(x[1:]) / 365. * 12)\n",
    "df_meta['Mouse_ID'] = df_meta['characteristics_ch1.1.developmental stage']\n",
    "mice = df_meta['Mouse_ID'].unique()\n",
    "mice = np.array(['E4.5', 'E5.5', 'E6.5', 'E7.5'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bef0871-efa1-48c8-b28b-0374392ab284",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9500ee5e-998a-4148-ad65-04b9478ad393",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/Dropbox/Harvard/Science/scDNAm/data/GSE121690/GSE121690_coverage_0_binned_by_1.h5ad\n"
     ]
    }
   ],
   "source": [
    "data_path = Path.joinpath(cwd, project_dir, \"{}_coverage_0_binned_by_1.h5ad\".format(project_id))\n",
    "print(data_path)\n",
    "ad = sc.read_h5ad(data_path)\n",
    "ad.var = pd.concat((ad.var, df_meta.loc[ad.var.index,:]), axis=1)\n",
    "ad.X = ad.X.astype(np.float32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a86e1e98-3a29-4bb8-8721-2352aa4de959",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad.obs_names_make_unique()\n",
    "ad.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4d41f938-c508-44a6-af24-3759d4b71425",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'title' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'geo_accession' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'status' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'submission_date' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'last_update_date' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'channel_count' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'source_name_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'organism_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'taxid_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'characteristics_ch1.0.strain' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'characteristics_ch1.1.developmental stage' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'characteristics_ch1.2.sample type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'molecule_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'extract_protocol_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'data_processing' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'platform_id' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_email' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_department' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_institute' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_address' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_city' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_zip/postal_code' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_country' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'instrument_model' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'library_selection' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'library_source' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'library_strategy' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'relation' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'supplementary_file_1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'supplementary_file_2' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'series_id' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'data_row_count' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Data Type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Mouse_ID' as categorical\n"
     ]
    }
   ],
   "source": [
    "ad.obs_names_make_unique()\n",
    "ad.var_names_make_unique()\n",
    "ad.write_h5ad(Path.joinpath(project_dir, \"{}_wmeta.h5ad\".format(project_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "1c600cf9-71f1-43b5-a8b1-a998aa0c032e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>channel_count</th>\n",
       "      <th>source_name_ch1</th>\n",
       "      <th>organism_ch1</th>\n",
       "      <th>taxid_ch1</th>\n",
       "      <th>characteristics_ch1.0.strain</th>\n",
       "      <th>characteristics_ch1.1.developmental stage</th>\n",
       "      <th>characteristics_ch1.2.sample type</th>\n",
       "      <th>molecule_ch1</th>\n",
       "      <th>extract_protocol_ch1</th>\n",
       "      <th>data_processing</th>\n",
       "      <th>...</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>relation</th>\n",
       "      <th>supplementary_file_1</th>\n",
       "      <th>supplementary_file_2</th>\n",
       "      <th>series_id</th>\n",
       "      <th>data_row_count</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Age</th>\n",
       "      <th>Mouse_ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name PerCov</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM3443369_percentage</th>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>C57BL/6Babr</td>\n",
       "      <td>E5.5</td>\n",
       "      <td>single cell</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>Single-cells were sorted either manually (E4.5...</td>\n",
       "      <td>Libraries were sequenced on the Illumina Nexts...</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>percentage</td>\n",
       "      <td>0.180822</td>\n",
       "      <td>E5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3443370_percentage</th>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>C57BL/6Babr</td>\n",
       "      <td>E5.5</td>\n",
       "      <td>single cell</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>Single-cells were sorted either manually (E4.5...</td>\n",
       "      <td>Libraries were sequenced on the Illumina Nexts...</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>percentage</td>\n",
       "      <td>0.180822</td>\n",
       "      <td>E5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3443371_percentage</th>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>C57BL/6Babr</td>\n",
       "      <td>E5.5</td>\n",
       "      <td>single cell</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>Single-cells were sorted either manually (E4.5...</td>\n",
       "      <td>Libraries were sequenced on the Illumina Nexts...</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>percentage</td>\n",
       "      <td>0.180822</td>\n",
       "      <td>E5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3443372_percentage</th>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>C57BL/6Babr</td>\n",
       "      <td>E5.5</td>\n",
       "      <td>single cell</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>Single-cells were sorted either manually (E4.5...</td>\n",
       "      <td>Libraries were sequenced on the Illumina Nexts...</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>percentage</td>\n",
       "      <td>0.180822</td>\n",
       "      <td>E5.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3443373_percentage</th>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>C57BL/6Babr</td>\n",
       "      <td>E5.5</td>\n",
       "      <td>single cell</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>Single-cells were sorted either manually (E4.5...</td>\n",
       "      <td>Libraries were sequenced on the Illumina Nexts...</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>percentage</td>\n",
       "      <td>0.180822</td>\n",
       "      <td>E5.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      channel_count source_name_ch1  organism_ch1 taxid_ch1  \\\n",
       "Name PerCov                                                                   \n",
       "GSM3443369_percentage             1    Early embryo  Mus musculus     10090   \n",
       "GSM3443370_percentage             1    Early embryo  Mus musculus     10090   \n",
       "GSM3443371_percentage             1    Early embryo  Mus musculus     10090   \n",
       "GSM3443372_percentage             1    Early embryo  Mus musculus     10090   \n",
       "GSM3443373_percentage             1    Early embryo  Mus musculus     10090   \n",
       "\n",
       "                      characteristics_ch1.0.strain  \\\n",
       "Name PerCov                                          \n",
       "GSM3443369_percentage                  C57BL/6Babr   \n",
       "GSM3443370_percentage                  C57BL/6Babr   \n",
       "GSM3443371_percentage                  C57BL/6Babr   \n",
       "GSM3443372_percentage                  C57BL/6Babr   \n",
       "GSM3443373_percentage                  C57BL/6Babr   \n",
       "\n",
       "                      characteristics_ch1.1.developmental stage  \\\n",
       "Name PerCov                                                       \n",
       "GSM3443369_percentage                                      E5.5   \n",
       "GSM3443370_percentage                                      E5.5   \n",
       "GSM3443371_percentage                                      E5.5   \n",
       "GSM3443372_percentage                                      E5.5   \n",
       "GSM3443373_percentage                                      E5.5   \n",
       "\n",
       "                      characteristics_ch1.2.sample type molecule_ch1  \\\n",
       "Name PerCov                                                            \n",
       "GSM3443369_percentage                       single cell  genomic DNA   \n",
       "GSM3443370_percentage                       single cell  genomic DNA   \n",
       "GSM3443371_percentage                       single cell  genomic DNA   \n",
       "GSM3443372_percentage                       single cell  genomic DNA   \n",
       "GSM3443373_percentage                       single cell  genomic DNA   \n",
       "\n",
       "                                                    extract_protocol_ch1  \\\n",
       "Name PerCov                                                                \n",
       "GSM3443369_percentage  Single-cells were sorted either manually (E4.5...   \n",
       "GSM3443370_percentage  Single-cells were sorted either manually (E4.5...   \n",
       "GSM3443371_percentage  Single-cells were sorted either manually (E4.5...   \n",
       "GSM3443372_percentage  Single-cells were sorted either manually (E4.5...   \n",
       "GSM3443373_percentage  Single-cells were sorted either manually (E4.5...   \n",
       "\n",
       "                                                         data_processing  ...  \\\n",
       "Name PerCov                                                               ...   \n",
       "GSM3443369_percentage  Libraries were sequenced on the Illumina Nexts...  ...   \n",
       "GSM3443370_percentage  Libraries were sequenced on the Illumina Nexts...  ...   \n",
       "GSM3443371_percentage  Libraries were sequenced on the Illumina Nexts...  ...   \n",
       "GSM3443372_percentage  Libraries were sequenced on the Illumina Nexts...  ...   \n",
       "GSM3443373_percentage  Libraries were sequenced on the Illumina Nexts...  ...   \n",
       "\n",
       "                      library_source library_strategy  \\\n",
       "Name PerCov                                             \n",
       "GSM3443369_percentage        genomic    Bisulfite-Seq   \n",
       "GSM3443370_percentage        genomic    Bisulfite-Seq   \n",
       "GSM3443371_percentage        genomic    Bisulfite-Seq   \n",
       "GSM3443372_percentage        genomic    Bisulfite-Seq   \n",
       "GSM3443373_percentage        genomic    Bisulfite-Seq   \n",
       "\n",
       "                                                                relation  \\\n",
       "Name PerCov                                                                \n",
       "GSM3443369_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM3443370_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM3443371_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM3443372_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM3443373_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "\n",
       "                                                    supplementary_file_1  \\\n",
       "Name PerCov                                                                \n",
       "GSM3443369_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443370_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443371_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443372_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443373_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "\n",
       "                                                    supplementary_file_2  \\\n",
       "Name PerCov                                                                \n",
       "GSM3443369_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443370_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443371_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443372_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "GSM3443373_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "\n",
       "                                 series_id data_row_count   Data Type  \\\n",
       "Name PerCov                                                             \n",
       "GSM3443369_percentage  GSE121690,GSE121708              0  percentage   \n",
       "GSM3443370_percentage  GSE121690,GSE121708              0  percentage   \n",
       "GSM3443371_percentage  GSE121690,GSE121708              0  percentage   \n",
       "GSM3443372_percentage  GSE121690,GSE121708              0  percentage   \n",
       "GSM3443373_percentage  GSE121690,GSE121708              0  percentage   \n",
       "\n",
       "                            Age Mouse_ID  \n",
       "Name PerCov                               \n",
       "GSM3443369_percentage  0.180822     E5.5  \n",
       "GSM3443370_percentage  0.180822     E5.5  \n",
       "GSM3443371_percentage  0.180822     E5.5  \n",
       "GSM3443372_percentage  0.180822     E5.5  \n",
       "GSM3443373_percentage  0.180822     E5.5  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta.iloc[:,6:].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f61072b6-6abf-4205-8674-cbd4c46d4803",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c39e2a9-da96-47ea-9fe9-fd7fee8dca0e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e67d861b-82a5-4a7f-8b4c-0603cd4a6b01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['E4.5' 'E5.5' 'E6.5' 'E7.5']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/v_/gtsrj3xs0_131ltxwg51r_f40000gn/T/ipykernel_6997/1600779323.py:12: RuntimeWarning: invalid value encountered in true_divide\n",
      "  xbulk[:,2*i] = xmeth[:,idx].sum(axis=1).flatten() / xcov[:,idx].sum(axis=1).flatten()\n"
     ]
    }
   ],
   "source": [
    "x = ad.X[:,::2]\n",
    "xcov = ad.X[:,1::2]\n",
    "xmeth = x.multiply(xcov)\n",
    "\n",
    "mouse_ids = ad.var.loc[::2,:].Mouse_ID.values\n",
    "mice = np.unique(mouse_ids)\n",
    "print(mice)\n",
    "xbulk = np.zeros((ad.X.shape[0], 2* len(mice)))\n",
    "#xcovbulk = np.zeros((ad.X.shape[0], len(mice)))\n",
    "for i,mouse_id in enumerate(mice):\n",
    "    idx = mouse_ids == mouse_id\n",
    "    xbulk[:,2*i] = xmeth[:,idx].sum(axis=1).flatten() / xcov[:,idx].sum(axis=1).flatten()\n",
    "    xbulk[:,2*i+1] = xcov[:,idx].sum(axis=1).flatten()\n",
    "    \n",
    "mice_index = np.array([[f\"{x}_percentage\", f\"{x}_coverage\"] for x in mice]).flatten()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "f4ded4fa-68a0-4baf-8550-668ed058b7ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8, 37)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mouse_ID</th>\n",
       "      <th>title</th>\n",
       "      <th>geo_accession</th>\n",
       "      <th>status</th>\n",
       "      <th>submission_date</th>\n",
       "      <th>last_update_date</th>\n",
       "      <th>type</th>\n",
       "      <th>channel_count</th>\n",
       "      <th>source_name_ch1</th>\n",
       "      <th>organism_ch1</th>\n",
       "      <th>...</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>relation</th>\n",
       "      <th>supplementary_file_1</th>\n",
       "      <th>supplementary_file_2</th>\n",
       "      <th>series_id</th>\n",
       "      <th>data_row_count</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>Age</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name PerCov</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>E4.5_percentage</th>\n",
       "      <td>E4.5</td>\n",
       "      <td>E4.5-5.5_new_Plate3_A02 (scBS-Seq)</td>\n",
       "      <td>GSM3443503</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>...</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>percentage</td>\n",
       "      <td>0.147945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E4.5_coverage</th>\n",
       "      <td>E4.5</td>\n",
       "      <td>E4.5-5.5_new_Plate3_A02 (scBS-Seq)</td>\n",
       "      <td>GSM3443503</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>...</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>coverage</td>\n",
       "      <td>0.147945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E5.5_percentage</th>\n",
       "      <td>E5.5</td>\n",
       "      <td>E4.5-5.5_new_Plate1_A02 (scBS-Seq)</td>\n",
       "      <td>GSM3443369</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>...</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>percentage</td>\n",
       "      <td>0.180822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E5.5_coverage</th>\n",
       "      <td>E5.5</td>\n",
       "      <td>E4.5-5.5_new_Plate1_A02 (scBS-Seq)</td>\n",
       "      <td>GSM3443369</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>...</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>coverage</td>\n",
       "      <td>0.180822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E6.5_percentage</th>\n",
       "      <td>E6.5</td>\n",
       "      <td>E6.5_Plate2_G6 (scBS-Seq)</td>\n",
       "      <td>GSM3443636</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>...</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>percentage</td>\n",
       "      <td>0.213699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E6.5_coverage</th>\n",
       "      <td>E6.5</td>\n",
       "      <td>E6.5_Plate2_G6 (scBS-Seq)</td>\n",
       "      <td>GSM3443636</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>...</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>coverage</td>\n",
       "      <td>0.213699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E7.5_percentage</th>\n",
       "      <td>E7.5</td>\n",
       "      <td>E7.5_Plate1_A3 (scBS-Seq)</td>\n",
       "      <td>GSM3443831</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>...</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>percentage</td>\n",
       "      <td>0.246575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E7.5_coverage</th>\n",
       "      <td>E7.5</td>\n",
       "      <td>E7.5_Plate1_A3 (scBS-Seq)</td>\n",
       "      <td>GSM3443831</td>\n",
       "      <td>Public on Sep 30 2019</td>\n",
       "      <td>Oct 23 2018</td>\n",
       "      <td>Sep 30 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Early embryo</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>...</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...</td>\n",
       "      <td>GSE121690,GSE121708</td>\n",
       "      <td>0</td>\n",
       "      <td>coverage</td>\n",
       "      <td>0.246575</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows  37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Mouse_ID                               title geo_accession  \\\n",
       "Name PerCov                                                                  \n",
       "E4.5_percentage     E4.5  E4.5-5.5_new_Plate3_A02 (scBS-Seq)    GSM3443503   \n",
       "E4.5_coverage       E4.5  E4.5-5.5_new_Plate3_A02 (scBS-Seq)    GSM3443503   \n",
       "E5.5_percentage     E5.5  E4.5-5.5_new_Plate1_A02 (scBS-Seq)    GSM3443369   \n",
       "E5.5_coverage       E5.5  E4.5-5.5_new_Plate1_A02 (scBS-Seq)    GSM3443369   \n",
       "E6.5_percentage     E6.5           E6.5_Plate2_G6 (scBS-Seq)    GSM3443636   \n",
       "E6.5_coverage       E6.5           E6.5_Plate2_G6 (scBS-Seq)    GSM3443636   \n",
       "E7.5_percentage     E7.5           E7.5_Plate1_A3 (scBS-Seq)    GSM3443831   \n",
       "E7.5_coverage       E7.5           E7.5_Plate1_A3 (scBS-Seq)    GSM3443831   \n",
       "\n",
       "                                status submission_date last_update_date type  \\\n",
       "Name PerCov                                                                    \n",
       "E4.5_percentage  Public on Sep 30 2019     Oct 23 2018      Sep 30 2019  SRA   \n",
       "E4.5_coverage    Public on Sep 30 2019     Oct 23 2018      Sep 30 2019  SRA   \n",
       "E5.5_percentage  Public on Sep 30 2019     Oct 23 2018      Sep 30 2019  SRA   \n",
       "E5.5_coverage    Public on Sep 30 2019     Oct 23 2018      Sep 30 2019  SRA   \n",
       "E6.5_percentage  Public on Sep 30 2019     Oct 23 2018      Sep 30 2019  SRA   \n",
       "E6.5_coverage    Public on Sep 30 2019     Oct 23 2018      Sep 30 2019  SRA   \n",
       "E7.5_percentage  Public on Sep 30 2019     Oct 23 2018      Sep 30 2019  SRA   \n",
       "E7.5_coverage    Public on Sep 30 2019     Oct 23 2018      Sep 30 2019  SRA   \n",
       "\n",
       "                channel_count source_name_ch1  organism_ch1  ...  \\\n",
       "Name PerCov                                                  ...   \n",
       "E4.5_percentage             1    Early embryo  Mus musculus  ...   \n",
       "E4.5_coverage               1    Early embryo  Mus musculus  ...   \n",
       "E5.5_percentage             1    Early embryo  Mus musculus  ...   \n",
       "E5.5_coverage               1    Early embryo  Mus musculus  ...   \n",
       "E6.5_percentage             1    Early embryo  Mus musculus  ...   \n",
       "E6.5_coverage               1    Early embryo  Mus musculus  ...   \n",
       "E7.5_percentage             1    Early embryo  Mus musculus  ...   \n",
       "E7.5_coverage               1    Early embryo  Mus musculus  ...   \n",
       "\n",
       "                library_selection library_source library_strategy  \\\n",
       "Name PerCov                                                         \n",
       "E4.5_percentage            RANDOM        genomic    Bisulfite-Seq   \n",
       "E4.5_coverage              RANDOM        genomic    Bisulfite-Seq   \n",
       "E5.5_percentage            RANDOM        genomic    Bisulfite-Seq   \n",
       "E5.5_coverage              RANDOM        genomic    Bisulfite-Seq   \n",
       "E6.5_percentage            RANDOM        genomic    Bisulfite-Seq   \n",
       "E6.5_coverage              RANDOM        genomic    Bisulfite-Seq   \n",
       "E7.5_percentage            RANDOM        genomic    Bisulfite-Seq   \n",
       "E7.5_coverage              RANDOM        genomic    Bisulfite-Seq   \n",
       "\n",
       "                                                          relation  \\\n",
       "Name PerCov                                                          \n",
       "E4.5_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "E4.5_coverage    BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "E5.5_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "E5.5_coverage    BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "E6.5_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "E6.5_coverage    BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "E7.5_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "E7.5_coverage    BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "\n",
       "                                              supplementary_file_1  \\\n",
       "Name PerCov                                                          \n",
       "E4.5_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "E4.5_coverage    ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "E5.5_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "E5.5_coverage    ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "E6.5_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "E6.5_coverage    ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "E7.5_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "E7.5_coverage    ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "\n",
       "                                              supplementary_file_2  \\\n",
       "Name PerCov                                                          \n",
       "E4.5_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "E4.5_coverage    ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "E5.5_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "E5.5_coverage    ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "E6.5_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "E6.5_coverage    ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "E7.5_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "E7.5_coverage    ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3443...   \n",
       "\n",
       "                           series_id data_row_count   Data Type       Age  \n",
       "Name PerCov                                                                \n",
       "E4.5_percentage  GSE121690,GSE121708              0  percentage  0.147945  \n",
       "E4.5_coverage    GSE121690,GSE121708              0    coverage  0.147945  \n",
       "E5.5_percentage  GSE121690,GSE121708              0  percentage  0.180822  \n",
       "E5.5_coverage    GSE121690,GSE121708              0    coverage  0.180822  \n",
       "E6.5_percentage  GSE121690,GSE121708              0  percentage  0.213699  \n",
       "E6.5_coverage    GSE121690,GSE121708              0    coverage  0.213699  \n",
       "E7.5_percentage  GSE121690,GSE121708              0  percentage  0.246575  \n",
       "E7.5_coverage    GSE121690,GSE121708              0    coverage  0.246575  \n",
       "\n",
       "[8 rows x 37 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta_bulk = df_meta.drop_duplicates('Mouse_ID').set_index('Mouse_ID').loc[mice].copy()\n",
    "df_meta_bulk.reset_index(inplace=True)\n",
    "df_meta_2 = df_meta_bulk.copy()\n",
    "df_meta_bulk['Name PerCov'] = df_meta_bulk['Mouse_ID'].str.cat(['percentage'] * df_meta_bulk.shape[0], sep='_')\n",
    "df_meta_bulk['Data Type'] = 'percentage'\n",
    "df_meta_2['Name PerCov'] = df_meta_2['Mouse_ID'].str.cat(['coverage'] * df_meta_bulk.shape[0], sep='_')\n",
    "df_meta_2['Data Type'] = 'coverage'\n",
    "#df_meta['Name PerCov'] = df_meta['Sample Name'].str.cat(['percentage'] * df_meta.shape[0], sep='_')\n",
    "df_meta_bulk = pd.concat((df_meta_bulk, df_meta_2), axis=0)\n",
    "df_meta_bulk.set_index('Name PerCov', inplace=True)\n",
    "print(df_meta_bulk.shape)\n",
    "df_meta_bulk = df_meta_bulk.loc[mice_index]\n",
    "df_meta_bulk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "fa6866b4-5b06-48d3-a0c0-b718475a8c1e",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Mouse_ID' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'title' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'geo_accession' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'status' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'submission_date' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'last_update_date' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'channel_count' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'source_name_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'organism_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'taxid_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'characteristics_ch1.0.strain' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'characteristics_ch1.1.developmental stage' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'characteristics_ch1.2.sample type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'molecule_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'extract_protocol_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'data_processing' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'platform_id' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_email' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_department' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_institute' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_address' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_city' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_zip/postal_code' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_country' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'instrument_model' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'library_selection' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'library_source' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'library_strategy' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'relation' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'supplementary_file_1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'supplementary_file_2' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'series_id' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'data_row_count' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Data Type' as categorical\n"
     ]
    }
   ],
   "source": [
    "adbu = sc.AnnData(sp.csr_matrix(np.nan_to_num(xbulk)), ad.obs, df_meta_bulk)\n",
    "adbu.obs_names_make_unique()\n",
    "adbu.var_names_make_unique()\n",
    "\n",
    "adbu.write_h5ad(Path.joinpath(project_dir, \"{}_bulked_wmeta_mm10.h5ad\".format(project_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "cfaad484-7832-4a8c-ae5e-6e88817aedea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(AnnData object with n_obs  n_vars = 20073742  8\n",
       "     var: 'Mouse_ID', 'title', 'geo_accession', 'status', 'submission_date', 'last_update_date', 'type', 'channel_count', 'source_name_ch1', 'organism_ch1', 'taxid_ch1', 'characteristics_ch1.0.strain', 'characteristics_ch1.1.developmental stage', 'characteristics_ch1.2.sample type', 'molecule_ch1', 'extract_protocol_ch1', 'data_processing', 'platform_id', 'contact_name', 'contact_email', 'contact_department', 'contact_institute', 'contact_address', 'contact_city', 'contact_zip/postal_code', 'contact_country', 'instrument_model', 'library_selection', 'library_source', 'library_strategy', 'relation', 'supplementary_file_1', 'supplementary_file_2', 'series_id', 'data_row_count', 'Data Type', 'Age',\n",
       " AnnData object with n_obs  n_vars = 20073742  1516\n",
       "     var: 'title', 'geo_accession', 'status', 'submission_date', 'last_update_date', 'type', 'channel_count', 'source_name_ch1', 'organism_ch1', 'taxid_ch1', 'characteristics_ch1.0.strain', 'characteristics_ch1.1.developmental stage', 'characteristics_ch1.2.sample type', 'molecule_ch1', 'extract_protocol_ch1', 'data_processing', 'platform_id', 'contact_name', 'contact_email', 'contact_department', 'contact_institute', 'contact_address', 'contact_city', 'contact_zip/postal_code', 'contact_country', 'instrument_model', 'library_selection', 'library_source', 'library_strategy', 'relation', 'supplementary_file_1', 'supplementary_file_2', 'series_id', 'data_row_count', 'Data Type', 'Age', 'Mouse_ID')"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adbu, ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "96ef1935-5caf-4880-9152-2642d345f0c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad[:, ad.var.Age * 365 / 12 > 5.].write_h5ad(Path.joinpath(project_dir, \"{}_wmeta_postE55.h5ad\".format(project_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e216ee11-4e18-45b3-919c-0e1c9c0646d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "adbu[:, adbu.var.Age * 365 / 12 > 5.].write_h5ad(Path.joinpath(project_dir, \"{}_bulked_wmeta_postE55.h5ad\".format(project_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "a8fc13a9-2ad7-43a9-ab10-922276b3ff40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GSM3443850_percentage    7.5\n",
       "GSM3443850_coverage      7.5\n",
       "GSM3443761_percentage    6.5\n",
       "GSM3443761_coverage      6.5\n",
       "GSM3443588_percentage    4.5\n",
       "                        ... \n",
       "GSM3443956_coverage      7.5\n",
       "GSM3443841_percentage    7.5\n",
       "GSM3443841_coverage      7.5\n",
       "GSM3443717_percentage    6.5\n",
       "GSM3443717_coverage      6.5\n",
       "Name: Age, Length: 1516, dtype: float64"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad.var.Age * 365 / 12"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31202e47-4fbe-4a61-a8c5-06344abd2394",
   "metadata": {},
   "source": [
    "# GSE56879"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "996e6275-44df-4cf9-873d-eafc3a6614b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "gse = 'GSE56879'\n",
    "cwd = Path.cwd()\n",
    "# cwd = Path('/Users/tarkhov/Dropbox/Harvard/Science/Church/')\n",
    "project_dir = Path.joinpath(cwd, Path(f'data/{gse}/'))\n",
    "project_id = f'{gse}'\n",
    "project_samples_dir = f'{gse}_RAW'\n",
    "metainfo_path = f'{gse}_series_matrix.txt'\n",
    "gl_metainfo_path = Path.joinpath(project_dir, metainfo_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "51389cce-d69e-4a0e-950f-8ec5118463f9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "03-Aug-2022 20:54:02 DEBUG utils - Directory ./data/GSE56879/ already exists. Skipping.\n",
      "03-Aug-2022 20:54:02 INFO GEOparse - File already exist: using local version.\n",
      "03-Aug-2022 20:54:02 INFO GEOparse - Parsing ./data/GSE56879/GSE56879_family.soft.gz: \n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SERIES: GSE56879\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - PLATFORM: GPL13112\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370522\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370523\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370524\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370525\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370526\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370527\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370528\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370529\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370530\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370531\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370532\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370533\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370534\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370535\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370536\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370537\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370538\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370539\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370540\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370541\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370542\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370543\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370544\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370545\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370546\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370547\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370548\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370549\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370550\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370551\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370552\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370553\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370554\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370555\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370556\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370557\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370558\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370559\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370560\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370561\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370562\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370563\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370564\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370565\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370566\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370567\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370568\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370569\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370570\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370571\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370572\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370573\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370574\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1370575\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1413827\n",
      "03-Aug-2022 20:54:02 DEBUG GEOparse - SAMPLE: GSM1413828\n"
     ]
    }
   ],
   "source": [
    "import GEOparse\n",
    "gse = GEOparse.get_GEO(geo=f\"{gse}\", destdir=f\"./data/{gse}/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "8d21fb89-604a-4c2a-b90c-4759c7772a6a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(112, 39)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>geo_accession</th>\n",
       "      <th>status</th>\n",
       "      <th>submission_date</th>\n",
       "      <th>last_update_date</th>\n",
       "      <th>type</th>\n",
       "      <th>channel_count</th>\n",
       "      <th>source_name_ch1</th>\n",
       "      <th>organism_ch1</th>\n",
       "      <th>taxid_ch1</th>\n",
       "      <th>...</th>\n",
       "      <th>relation</th>\n",
       "      <th>supplementary_file_1</th>\n",
       "      <th>supplementary_file_2</th>\n",
       "      <th>series_id</th>\n",
       "      <th>data_row_count</th>\n",
       "      <th>characteristics_ch1.1.treatment</th>\n",
       "      <th>characteristics_ch1.2.strain</th>\n",
       "      <th>characteristics_ch1.0.strain</th>\n",
       "      <th>description</th>\n",
       "      <th>Data Type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name PerCov</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1370522_percentage</th>\n",
       "      <td>MII_1</td>\n",
       "      <td>GSM1370522</td>\n",
       "      <td>Public on Jul 18 2014</td>\n",
       "      <td>Apr 17 2014</td>\n",
       "      <td>May 15 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>oocytes</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...</td>\n",
       "      <td>GSE56879</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1370523_percentage</th>\n",
       "      <td>MII_2</td>\n",
       "      <td>GSM1370523</td>\n",
       "      <td>Public on Jul 18 2014</td>\n",
       "      <td>Apr 17 2014</td>\n",
       "      <td>May 15 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>oocytes</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...</td>\n",
       "      <td>GSE56879</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1370524_percentage</th>\n",
       "      <td>MII_3</td>\n",
       "      <td>GSM1370524</td>\n",
       "      <td>Public on Jul 18 2014</td>\n",
       "      <td>Apr 17 2014</td>\n",
       "      <td>May 15 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>oocytes</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...</td>\n",
       "      <td>GSE56879</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1370525_percentage</th>\n",
       "      <td>MII_4</td>\n",
       "      <td>GSM1370525</td>\n",
       "      <td>Public on Jul 18 2014</td>\n",
       "      <td>Apr 17 2014</td>\n",
       "      <td>May 15 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>oocytes</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...</td>\n",
       "      <td>GSE56879</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1370526_percentage</th>\n",
       "      <td>MII_5</td>\n",
       "      <td>GSM1370526</td>\n",
       "      <td>Public on Jul 18 2014</td>\n",
       "      <td>Apr 17 2014</td>\n",
       "      <td>May 15 2019</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>oocytes</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>10090</td>\n",
       "      <td>...</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...</td>\n",
       "      <td>GSE56879</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  39 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       title geo_accession                 status  \\\n",
       "Name PerCov                                                         \n",
       "GSM1370522_percentage  MII_1    GSM1370522  Public on Jul 18 2014   \n",
       "GSM1370523_percentage  MII_2    GSM1370523  Public on Jul 18 2014   \n",
       "GSM1370524_percentage  MII_3    GSM1370524  Public on Jul 18 2014   \n",
       "GSM1370525_percentage  MII_4    GSM1370525  Public on Jul 18 2014   \n",
       "GSM1370526_percentage  MII_5    GSM1370526  Public on Jul 18 2014   \n",
       "\n",
       "                      submission_date last_update_date type channel_count  \\\n",
       "Name PerCov                                                                 \n",
       "GSM1370522_percentage     Apr 17 2014      May 15 2019  SRA             1   \n",
       "GSM1370523_percentage     Apr 17 2014      May 15 2019  SRA             1   \n",
       "GSM1370524_percentage     Apr 17 2014      May 15 2019  SRA             1   \n",
       "GSM1370525_percentage     Apr 17 2014      May 15 2019  SRA             1   \n",
       "GSM1370526_percentage     Apr 17 2014      May 15 2019  SRA             1   \n",
       "\n",
       "                      source_name_ch1  organism_ch1 taxid_ch1  ...  \\\n",
       "Name PerCov                                                    ...   \n",
       "GSM1370522_percentage         oocytes  Mus musculus     10090  ...   \n",
       "GSM1370523_percentage         oocytes  Mus musculus     10090  ...   \n",
       "GSM1370524_percentage         oocytes  Mus musculus     10090  ...   \n",
       "GSM1370525_percentage         oocytes  Mus musculus     10090  ...   \n",
       "GSM1370526_percentage         oocytes  Mus musculus     10090  ...   \n",
       "\n",
       "                                                                relation  \\\n",
       "Name PerCov                                                                \n",
       "GSM1370522_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM1370523_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM1370524_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM1370525_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM1370526_percentage  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "\n",
       "                                                    supplementary_file_1  \\\n",
       "Name PerCov                                                                \n",
       "GSM1370522_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...   \n",
       "GSM1370523_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...   \n",
       "GSM1370524_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...   \n",
       "GSM1370525_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...   \n",
       "GSM1370526_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...   \n",
       "\n",
       "                                                    supplementary_file_2  \\\n",
       "Name PerCov                                                                \n",
       "GSM1370522_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...   \n",
       "GSM1370523_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...   \n",
       "GSM1370524_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...   \n",
       "GSM1370525_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...   \n",
       "GSM1370526_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1370...   \n",
       "\n",
       "                      series_id data_row_count  \\\n",
       "Name PerCov                                      \n",
       "GSM1370522_percentage  GSE56879              0   \n",
       "GSM1370523_percentage  GSE56879              0   \n",
       "GSM1370524_percentage  GSE56879              0   \n",
       "GSM1370525_percentage  GSE56879              0   \n",
       "GSM1370526_percentage  GSE56879              0   \n",
       "\n",
       "                      characteristics_ch1.1.treatment  \\\n",
       "Name PerCov                                             \n",
       "GSM1370522_percentage                             NaN   \n",
       "GSM1370523_percentage                             NaN   \n",
       "GSM1370524_percentage                             NaN   \n",
       "GSM1370525_percentage                             NaN   \n",
       "GSM1370526_percentage                             NaN   \n",
       "\n",
       "                      characteristics_ch1.2.strain  \\\n",
       "Name PerCov                                          \n",
       "GSM1370522_percentage                          NaN   \n",
       "GSM1370523_percentage                          NaN   \n",
       "GSM1370524_percentage                          NaN   \n",
       "GSM1370525_percentage                          NaN   \n",
       "GSM1370526_percentage                          NaN   \n",
       "\n",
       "                      characteristics_ch1.0.strain description   Data Type  \n",
       "Name PerCov                                                                 \n",
       "GSM1370522_percentage                          NaN         NaN  percentage  \n",
       "GSM1370523_percentage                          NaN         NaN  percentage  \n",
       "GSM1370524_percentage                          NaN         NaN  percentage  \n",
       "GSM1370525_percentage                          NaN         NaN  percentage  \n",
       "GSM1370526_percentage                          NaN         NaN  percentage  \n",
       "\n",
       "[5 rows x 39 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = gse.phenotype_data.copy()\n",
    "df_meta_2 = df_meta.copy()\n",
    "df_meta['Name PerCov'] = df_meta['geo_accession'].str.cat(['percentage'] * df_meta.shape[0], sep='_')\n",
    "df_meta['Data Type'] = 'percentage'\n",
    "df_meta_2['Name PerCov'] = df_meta_2['geo_accession'].str.cat(['coverage'] * df_meta.shape[0], sep='_')\n",
    "df_meta_2['Data Type'] = 'coverage'\n",
    "df_meta['Name PerCov'] = df_meta['geo_accession'].str.cat(['percentage'] * df_meta.shape[0], sep='_')\n",
    "df_meta = pd.concat((df_meta, df_meta_2), axis=0)\n",
    "df_meta.set_index('Name PerCov', inplace=True)\n",
    "print(df_meta.shape)\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "id": "117c9011-0ad1-4e85-a4a7-dc9252b2a664",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Run', 'Assay Type', 'AvgSpotLen', 'Bases', 'BioProject', 'BioSample',\n",
       "       'Bytes', 'Center Name', 'Consent', 'DATASTORE filetype',\n",
       "       'DATASTORE provider', 'DATASTORE region', 'Experiment',\n",
       "       'GEO_Accession (exp)', 'Instrument', 'LibraryLayout',\n",
       "       'LibrarySelection', 'LibrarySource', 'Organism', 'Platform_x',\n",
       "       'ReleaseDate', 'Sample Name', 'SRA Study', 'Age', 'Cell_type',\n",
       "       'Mouse_ID', 'source_name', 'STRAIN', 'Accession', 'Title',\n",
       "       'Sample Type', 'Taxonomy', 'Channels', 'Platform_y', 'Series',\n",
       "       'Supplementary Types', 'Supplementary Links', 'SRA Accession',\n",
       "       'Contact', 'Release Date'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "679c7c2b-231b-4724-9ac1-3a51af609ea0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/Dropbox/Harvard/Science/scDNAm/data/GSE56879/GSE56879_coverage_0_binned_by_1.h5ad\n"
     ]
    }
   ],
   "source": [
    "data_path = Path.joinpath(cwd, project_dir, \"{}_coverage_0_binned_by_1.h5ad\".format(project_id))\n",
    "print(data_path)\n",
    "ad = sc.read_h5ad(data_path)\n",
    "ad.var = pd.concat((ad.var, df_meta.loc[ad.var.index,:]), axis=1)\n",
    "ad.X = ad.X.astype(np.float32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "11bb32c6-3f1a-4173-b89b-314b5ae0df4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad.obs_names_make_unique()\n",
    "ad.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f5140682-3070-4c05-a4fb-5e03391f9c67",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'title' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'geo_accession' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'status' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'submission_date' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'last_update_date' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'channel_count' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'source_name_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'organism_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'taxid_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'characteristics_ch1.0.cell line' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'characteristics_ch1.1.strain' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'growth_protocol_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'molecule_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'extract_protocol_ch1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'data_processing' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'platform_id' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_email' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_department' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_institute' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_address' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_city' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_zip/postal_code' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'contact_country' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'instrument_model' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'library_selection' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'library_source' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'library_strategy' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'relation' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'supplementary_file_1' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'supplementary_file_2' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'series_id' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'data_row_count' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'characteristics_ch1.1.treatment' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'characteristics_ch1.2.strain' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'characteristics_ch1.0.strain' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'description' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Data Type' as categorical\n"
     ]
    }
   ],
   "source": [
    "ad.write_h5ad(Path.joinpath(project_dir, \"{}_wmeta.h5ad\".format(project_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd9a8099-46f9-4816-96ce-5da93a5a5511",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ec416b37-86e8-4c6e-a063-86f1d5c8b42e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_meta_MII = df_meta.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc4d35c6-12d7-4443-9045-1dd112c7b676",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5ad2cd48-fff8-4de0-bf93-c74a6c32cff0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>3021291</td>\n",
       "      <td>3021291</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>3021293</td>\n",
       "      <td>3021293</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8</td>\n",
       "      <td>3021452</td>\n",
       "      <td>3021452</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>3036686</td>\n",
       "      <td>3036686</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>3068337</td>\n",
       "      <td>3068337</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0        1        2      3  4  5\n",
       "0  8  3021291  3021291  100.0  1  0\n",
       "1  8  3021293  3021293  100.0  1  0\n",
       "2  8  3021452  3021452    0.0  0  1\n",
       "3  8  3036686  3036686  100.0  1  0\n",
       "4  8  3068337  3068337  100.0  1  0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(path_sample, sep='\\t', header=None, skiprows=1,\n",
    "                 dtype={0: 'str', 1: 'int', 2: 'int', 3: 'float', 4: 'int', 5: 'int'})\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cf2e1ad6-ef24-4d5d-8921-8e71d79be8f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>3021291</td>\n",
       "      <td>3021291</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>3021293</td>\n",
       "      <td>3021293</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8</td>\n",
       "      <td>3021452</td>\n",
       "      <td>3021452</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>3036686</td>\n",
       "      <td>3036686</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>3068337</td>\n",
       "      <td>3068337</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7644136</th>\n",
       "      <td>Y</td>\n",
       "      <td>90829081</td>\n",
       "      <td>90829081</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7644137</th>\n",
       "      <td>Y</td>\n",
       "      <td>90829087</td>\n",
       "      <td>90829087</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7644138</th>\n",
       "      <td>Y</td>\n",
       "      <td>90829099</td>\n",
       "      <td>90829099</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7644139</th>\n",
       "      <td>Y</td>\n",
       "      <td>90829119</td>\n",
       "      <td>90829119</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7644140</th>\n",
       "      <td>Y</td>\n",
       "      <td>90829139</td>\n",
       "      <td>90829139</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7644141 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         0         1         2      3  4  5\n",
       "0        8   3021291   3021291  100.0  1  0\n",
       "1        8   3021293   3021293  100.0  1  0\n",
       "2        8   3021452   3021452    0.0  0  1\n",
       "3        8   3036686   3036686  100.0  1  0\n",
       "4        8   3068337   3068337  100.0  1  0\n",
       "...     ..       ...       ...    ... .. ..\n",
       "7644136  Y  90829081  90829081    0.0  0  1\n",
       "7644137  Y  90829087  90829087    0.0  0  1\n",
       "7644138  Y  90829099  90829099    0.0  0  1\n",
       "7644139  Y  90829119  90829119    0.0  0  1\n",
       "7644140  Y  90829139  90829139    0.0  0  1\n",
       "\n",
       "[7644141 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5f0e78cf-b263-422c-b596-a1becdcf2fc5",
   "metadata": {},
   "source": [
    "# GSE186458"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "13c8adeb-c5b6-4583-b4fd-582b597786e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwd = Path.cwd()\n",
    "# cwd = Path('/Users/tarkhov/Dropbox/Harvard/Science/Church/')\n",
    "project_dir = Path.joinpath(cwd, Path('data/GSE186458/'))\n",
    "project_id = 'GSE186458'\n",
    "project_samples_dir = 'GSE186458_RAW'\n",
    "metainfo_path = 'GSE186458_series_matrix.txt'\n",
    "gl_metainfo_path = Path.joinpath(project_dir, metainfo_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "837573c4-3a07-46d5-8fc2-053dd6d0bf2a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "16-Jun-2022 12:44:03 DEBUG utils - Directory ./data/GSE186458/ already exists. Skipping.\n",
      "16-Jun-2022 12:44:03 INFO GEOparse - Downloading ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE186nnn/GSE186458/soft/GSE186458_family.soft.gz to ./data/GSE186458/GSE186458_family.soft.gz\n",
      "100%|| 22.3k/22.3k [00:00<00:00, 289kB/s]\n",
      "16-Jun-2022 12:44:03 DEBUG downloader - Size validation passed\n",
      "16-Jun-2022 12:44:03 DEBUG downloader - Moving /var/folders/v_/gtsrj3xs0_131ltxwg51r_f40000gn/T/tmppv7byq60 to /Users/tarkhov/Dropbox/Harvard/Science/scDNAm/data/GSE186458/GSE186458_family.soft.gz\n",
      "16-Jun-2022 12:44:03 DEBUG downloader - Successfully downloaded ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE186nnn/GSE186458/soft/GSE186458_family.soft.gz\n",
      "16-Jun-2022 12:44:03 INFO GEOparse - Parsing ./data/GSE186458/GSE186458_family.soft.gz: \n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SERIES: GSE186458\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - PLATFORM: GPL24676\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652176\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652177\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652178\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652179\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652180\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652181\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652182\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652183\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652184\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652185\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652186\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652187\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652188\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652189\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652190\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652191\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652192\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652193\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652194\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652195\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652196\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652197\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652198\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652199\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652200\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652201\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652202\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652203\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652204\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652205\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652206\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652207\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652208\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652209\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652210\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652211\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652212\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652213\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652214\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652215\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652216\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652217\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652218\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652219\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652220\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652221\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652222\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652223\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652224\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652225\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652226\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652227\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652228\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652229\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652230\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652231\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652232\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652233\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652234\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652235\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652236\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652237\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652238\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652239\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652240\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652241\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652242\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652243\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652244\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652245\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652246\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652247\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652248\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652249\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652250\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652251\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652252\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652253\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652254\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652255\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652256\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652257\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652258\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652259\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652260\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652261\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652262\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652263\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652264\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652265\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652266\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652267\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652268\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652269\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652270\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652271\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652272\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652273\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652274\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652275\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652276\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652277\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652278\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652279\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652280\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652281\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652282\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652283\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652284\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652285\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652286\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652287\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652288\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652289\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652290\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652291\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652292\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652293\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652294\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652295\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652296\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652297\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652298\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652299\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652300\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652301\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652302\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652303\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652304\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652305\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652306\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652307\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652308\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652309\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652310\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652311\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652312\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652313\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652314\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652315\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652316\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652317\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652318\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652319\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652320\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652321\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652322\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652323\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652324\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652325\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652326\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652327\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652328\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652329\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652330\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652331\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652332\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652333\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652334\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652335\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652336\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652337\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652338\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652339\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652340\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652341\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652342\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652343\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652344\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652345\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652346\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652347\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652348\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652349\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652350\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652351\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652352\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652353\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652354\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652355\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652356\n",
      "16-Jun-2022 12:44:03 DEBUG GEOparse - SAMPLE: GSM5652357\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652358\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652359\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652360\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652361\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652362\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652363\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652364\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652365\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652366\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652367\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652368\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652369\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652370\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652371\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652372\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652373\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652374\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652375\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652376\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652377\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652378\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652379\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652380\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652381\n",
      "16-Jun-2022 12:44:04 DEBUG GEOparse - SAMPLE: GSM5652382\n"
     ]
    }
   ],
   "source": [
    "import GEOparse\n",
    "gse = GEOparse.get_GEO(geo=\"GSE186458\", destdir=\"./data/GSE186458/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "724b3163-9e32-4cb3-b44a-1a445c7aa3f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>geo_accession</th>\n",
       "      <th>status</th>\n",
       "      <th>submission_date</th>\n",
       "      <th>last_update_date</th>\n",
       "      <th>type</th>\n",
       "      <th>channel_count</th>\n",
       "      <th>source_name_ch1</th>\n",
       "      <th>organism_ch1</th>\n",
       "      <th>taxid_ch1</th>\n",
       "      <th>...</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>relation</th>\n",
       "      <th>supplementary_file_1</th>\n",
       "      <th>supplementary_file_2</th>\n",
       "      <th>supplementary_file_3</th>\n",
       "      <th>supplementary_file_4</th>\n",
       "      <th>series_id</th>\n",
       "      <th>data_row_count</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM5652176</th>\n",
       "      <td>Adipocytes-Z000000T7</td>\n",
       "      <td>GSM5652176</td>\n",
       "      <td>Public on Mar 02 2022</td>\n",
       "      <td>Oct 23 2021</td>\n",
       "      <td>May 18 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Abdominal Subcut.</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>GSE186458</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5652177</th>\n",
       "      <td>Adipocytes-Z000000T9</td>\n",
       "      <td>GSM5652177</td>\n",
       "      <td>Public on Mar 02 2022</td>\n",
       "      <td>Oct 23 2021</td>\n",
       "      <td>May 18 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Abdominal Subcut.</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>GSE186458</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5652178</th>\n",
       "      <td>Adipocytes-Z000000T5</td>\n",
       "      <td>GSM5652178</td>\n",
       "      <td>Public on Mar 02 2022</td>\n",
       "      <td>Oct 23 2021</td>\n",
       "      <td>May 18 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Abdominal Subcut.</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>GSE186458</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5652179</th>\n",
       "      <td>Aorta-Endothelium-Z00000422</td>\n",
       "      <td>GSM5652179</td>\n",
       "      <td>Public on Mar 02 2022</td>\n",
       "      <td>Oct 23 2021</td>\n",
       "      <td>May 18 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Aorta</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>GSE186458</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5652180</th>\n",
       "      <td>Aorta-Endothelium-Z0000043G</td>\n",
       "      <td>GSM5652180</td>\n",
       "      <td>Public on Mar 02 2022</td>\n",
       "      <td>Oct 23 2021</td>\n",
       "      <td>May 18 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Aorta</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>GSE186458</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  title geo_accession                 status  \\\n",
       "GSM5652176         Adipocytes-Z000000T7    GSM5652176  Public on Mar 02 2022   \n",
       "GSM5652177         Adipocytes-Z000000T9    GSM5652177  Public on Mar 02 2022   \n",
       "GSM5652178         Adipocytes-Z000000T5    GSM5652178  Public on Mar 02 2022   \n",
       "GSM5652179  Aorta-Endothelium-Z00000422    GSM5652179  Public on Mar 02 2022   \n",
       "GSM5652180  Aorta-Endothelium-Z0000043G    GSM5652180  Public on Mar 02 2022   \n",
       "\n",
       "           submission_date last_update_date type channel_count  \\\n",
       "GSM5652176     Oct 23 2021      May 18 2022  SRA             1   \n",
       "GSM5652177     Oct 23 2021      May 18 2022  SRA             1   \n",
       "GSM5652178     Oct 23 2021      May 18 2022  SRA             1   \n",
       "GSM5652179     Oct 23 2021      May 18 2022  SRA             1   \n",
       "GSM5652180     Oct 23 2021      May 18 2022  SRA             1   \n",
       "\n",
       "              source_name_ch1  organism_ch1 taxid_ch1  ... library_source  \\\n",
       "GSM5652176  Abdominal Subcut.  Homo sapiens      9606  ...        genomic   \n",
       "GSM5652177  Abdominal Subcut.  Homo sapiens      9606  ...        genomic   \n",
       "GSM5652178  Abdominal Subcut.  Homo sapiens      9606  ...        genomic   \n",
       "GSM5652179              Aorta  Homo sapiens      9606  ...        genomic   \n",
       "GSM5652180              Aorta  Homo sapiens      9606  ...        genomic   \n",
       "\n",
       "           library_strategy  \\\n",
       "GSM5652176    Bisulfite-Seq   \n",
       "GSM5652177    Bisulfite-Seq   \n",
       "GSM5652178    Bisulfite-Seq   \n",
       "GSM5652179    Bisulfite-Seq   \n",
       "GSM5652180    Bisulfite-Seq   \n",
       "\n",
       "                                                     relation  \\\n",
       "GSM5652176  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM5652177  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM5652178  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM5652179  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM5652180  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "\n",
       "                                         supplementary_file_1  \\\n",
       "GSM5652176  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652177  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652178  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652179  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652180  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "\n",
       "                                         supplementary_file_2  \\\n",
       "GSM5652176  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652177  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652178  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652179  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652180  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "\n",
       "                                         supplementary_file_3  \\\n",
       "GSM5652176  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652177  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652178  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652179  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652180  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "\n",
       "                                         supplementary_file_4  series_id  \\\n",
       "GSM5652176  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...  GSE186458   \n",
       "GSM5652177  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...  GSE186458   \n",
       "GSM5652178  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...  GSE186458   \n",
       "GSM5652179  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...  GSE186458   \n",
       "GSM5652180  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...  GSE186458   \n",
       "\n",
       "           data_row_count description  \n",
       "GSM5652176              0         NaN  \n",
       "GSM5652177              0         NaN  \n",
       "GSM5652178              0         NaN  \n",
       "GSM5652179              0         NaN  \n",
       "GSM5652180              0         NaN  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = gse.phenotype_data.copy()\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9510ae44-162e-41d2-8a18-fb96fd8df58c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "9ee277ab-2992-4d5a-9bf1-d8d1dac11ee0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>geo_accession</th>\n",
       "      <th>status</th>\n",
       "      <th>submission_date</th>\n",
       "      <th>last_update_date</th>\n",
       "      <th>type</th>\n",
       "      <th>channel_count</th>\n",
       "      <th>source_name_ch1</th>\n",
       "      <th>organism_ch1</th>\n",
       "      <th>taxid_ch1</th>\n",
       "      <th>...</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>relation</th>\n",
       "      <th>supplementary_file_1</th>\n",
       "      <th>supplementary_file_2</th>\n",
       "      <th>supplementary_file_3</th>\n",
       "      <th>supplementary_file_4</th>\n",
       "      <th>series_id</th>\n",
       "      <th>data_row_count</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM5652176</th>\n",
       "      <td>Adipocytes-Z000000T7</td>\n",
       "      <td>GSM5652176</td>\n",
       "      <td>Public on Mar 02 2022</td>\n",
       "      <td>Oct 23 2021</td>\n",
       "      <td>May 18 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Abdominal Subcut.</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>GSE186458</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5652177</th>\n",
       "      <td>Adipocytes-Z000000T9</td>\n",
       "      <td>GSM5652177</td>\n",
       "      <td>Public on Mar 02 2022</td>\n",
       "      <td>Oct 23 2021</td>\n",
       "      <td>May 18 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Abdominal Subcut.</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>GSE186458</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5652178</th>\n",
       "      <td>Adipocytes-Z000000T5</td>\n",
       "      <td>GSM5652178</td>\n",
       "      <td>Public on Mar 02 2022</td>\n",
       "      <td>Oct 23 2021</td>\n",
       "      <td>May 18 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Abdominal Subcut.</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>GSE186458</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5652179</th>\n",
       "      <td>Aorta-Endothelium-Z00000422</td>\n",
       "      <td>GSM5652179</td>\n",
       "      <td>Public on Mar 02 2022</td>\n",
       "      <td>Oct 23 2021</td>\n",
       "      <td>May 18 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Aorta</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>GSE186458</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM5652180</th>\n",
       "      <td>Aorta-Endothelium-Z0000043G</td>\n",
       "      <td>GSM5652180</td>\n",
       "      <td>Public on Mar 02 2022</td>\n",
       "      <td>Oct 23 2021</td>\n",
       "      <td>May 18 2022</td>\n",
       "      <td>SRA</td>\n",
       "      <td>1</td>\n",
       "      <td>Aorta</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>9606</td>\n",
       "      <td>...</td>\n",
       "      <td>genomic</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>BioSample: https://www.ncbi.nlm.nih.gov/biosam...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...</td>\n",
       "      <td>GSE186458</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  title geo_accession                 status  \\\n",
       "GSM5652176         Adipocytes-Z000000T7    GSM5652176  Public on Mar 02 2022   \n",
       "GSM5652177         Adipocytes-Z000000T9    GSM5652177  Public on Mar 02 2022   \n",
       "GSM5652178         Adipocytes-Z000000T5    GSM5652178  Public on Mar 02 2022   \n",
       "GSM5652179  Aorta-Endothelium-Z00000422    GSM5652179  Public on Mar 02 2022   \n",
       "GSM5652180  Aorta-Endothelium-Z0000043G    GSM5652180  Public on Mar 02 2022   \n",
       "\n",
       "           submission_date last_update_date type channel_count  \\\n",
       "GSM5652176     Oct 23 2021      May 18 2022  SRA             1   \n",
       "GSM5652177     Oct 23 2021      May 18 2022  SRA             1   \n",
       "GSM5652178     Oct 23 2021      May 18 2022  SRA             1   \n",
       "GSM5652179     Oct 23 2021      May 18 2022  SRA             1   \n",
       "GSM5652180     Oct 23 2021      May 18 2022  SRA             1   \n",
       "\n",
       "              source_name_ch1  organism_ch1 taxid_ch1  ... library_source  \\\n",
       "GSM5652176  Abdominal Subcut.  Homo sapiens      9606  ...        genomic   \n",
       "GSM5652177  Abdominal Subcut.  Homo sapiens      9606  ...        genomic   \n",
       "GSM5652178  Abdominal Subcut.  Homo sapiens      9606  ...        genomic   \n",
       "GSM5652179              Aorta  Homo sapiens      9606  ...        genomic   \n",
       "GSM5652180              Aorta  Homo sapiens      9606  ...        genomic   \n",
       "\n",
       "           library_strategy  \\\n",
       "GSM5652176    Bisulfite-Seq   \n",
       "GSM5652177    Bisulfite-Seq   \n",
       "GSM5652178    Bisulfite-Seq   \n",
       "GSM5652179    Bisulfite-Seq   \n",
       "GSM5652180    Bisulfite-Seq   \n",
       "\n",
       "                                                     relation  \\\n",
       "GSM5652176  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM5652177  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM5652178  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM5652179  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "GSM5652180  BioSample: https://www.ncbi.nlm.nih.gov/biosam...   \n",
       "\n",
       "                                         supplementary_file_1  \\\n",
       "GSM5652176  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652177  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652178  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652179  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652180  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "\n",
       "                                         supplementary_file_2  \\\n",
       "GSM5652176  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652177  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652178  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652179  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652180  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "\n",
       "                                         supplementary_file_3  \\\n",
       "GSM5652176  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652177  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652178  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652179  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "GSM5652180  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...   \n",
       "\n",
       "                                         supplementary_file_4  series_id  \\\n",
       "GSM5652176  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...  GSE186458   \n",
       "GSM5652177  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...  GSE186458   \n",
       "GSM5652178  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...  GSE186458   \n",
       "GSM5652179  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...  GSE186458   \n",
       "GSM5652180  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM5652...  GSE186458   \n",
       "\n",
       "           data_row_count description  \n",
       "GSM5652176              0         NaN  \n",
       "GSM5652177              0         NaN  \n",
       "GSM5652178              0         NaN  \n",
       "GSM5652179              0         NaN  \n",
       "GSM5652180              0         NaN  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4c305af-8914-4287-abcc-db95160b3356",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "826b2093-cbe0-4d2c-9c80-a2ac566c72d7",
   "metadata": {},
   "source": [
    "# SRA344045 liver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "8ac780c7-0a44-4544-9115-ce5989701c39",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwd = Path.cwd()\n",
    "# cwd = Path('/Users/tarkhov/Dropbox/Harvard/Science/Church/')\n",
    "project_dir = Path.joinpath(cwd, Path('data/SRA344045/'))\n",
    "project_id = 'SRA344045'\n",
    "project_samples_dir = 'SRA344045_RAW'\n",
    "metainfo_path = 'SraRunTable.txt'\n",
    "gl_metainfo_path = Path.joinpath(project_dir, metainfo_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "68386224-60ec-4222-8843-44d43102eacb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(34, 32)\n",
      "(68, 33)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AssemblyName</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleModel</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>...</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>sex</th>\n",
       "      <th>SRA Study</th>\n",
       "      <th>strain</th>\n",
       "      <th>tissue</th>\n",
       "      <th>Age</th>\n",
       "      <th>dev_stage</th>\n",
       "      <th>Age_months</th>\n",
       "      <th>Mouse_ID</th>\n",
       "      <th>Data Type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name PerCov</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR3136624_percentage</th>\n",
       "      <td>SRR3136624</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>42</td>\n",
       "      <td>476309309</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449570</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>305774648</td>\n",
       "      <td>ALBERT EINSTEIN COLLEGE OF MEDICINE</td>\n",
       "      <td>...</td>\n",
       "      <td>f11</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>MEFs in culture</td>\n",
       "      <td>0.0</td>\n",
       "      <td>embryonic</td>\n",
       "      <td>0 months</td>\n",
       "      <td>f1</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR3136625_percentage</th>\n",
       "      <td>SRR3136625</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>82</td>\n",
       "      <td>276514225</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449998</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>166102905</td>\n",
       "      <td>ALBERT EINSTEIN COLLEGE OF MEDICINE</td>\n",
       "      <td>...</td>\n",
       "      <td>f23</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>MEFs in culture</td>\n",
       "      <td>0.0</td>\n",
       "      <td>embryonic</td>\n",
       "      <td>0 months</td>\n",
       "      <td>f2</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR3136626_percentage</th>\n",
       "      <td>SRR3136626</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>83</td>\n",
       "      <td>345410020</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449847</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>197726064</td>\n",
       "      <td>ALBERT EINSTEIN COLLEGE OF MEDICINE</td>\n",
       "      <td>...</td>\n",
       "      <td>f22</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>MEFs in culture</td>\n",
       "      <td>0.0</td>\n",
       "      <td>embryonic</td>\n",
       "      <td>0 months</td>\n",
       "      <td>f2</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR3136627_percentage</th>\n",
       "      <td>SRR3136627</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>85</td>\n",
       "      <td>450783127</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449846</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>250903561</td>\n",
       "      <td>ALBERT EINSTEIN COLLEGE OF MEDICINE</td>\n",
       "      <td>...</td>\n",
       "      <td>f21</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>MEFs in culture</td>\n",
       "      <td>0.0</td>\n",
       "      <td>embryonic</td>\n",
       "      <td>0 months</td>\n",
       "      <td>f2</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR3136628_percentage</th>\n",
       "      <td>SRR3136628</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>83</td>\n",
       "      <td>322920190</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449988</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>214862269</td>\n",
       "      <td>ALBERT EINSTEIN COLLEGE OF MEDICINE</td>\n",
       "      <td>...</td>\n",
       "      <td>o11</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>Liver</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26 month</td>\n",
       "      <td>o1</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Run     Assay Type     AssemblyName  AvgSpotLen  \\\n",
       "Name PerCov                                                                     \n",
       "SRR3136624_percentage  SRR3136624  Bisulfite-Seq  GCA_000001635.1          42   \n",
       "SRR3136625_percentage  SRR3136625  Bisulfite-Seq  GCA_000001635.1          82   \n",
       "SRR3136626_percentage  SRR3136626  Bisulfite-Seq  GCA_000001635.1          83   \n",
       "SRR3136627_percentage  SRR3136627  Bisulfite-Seq  GCA_000001635.1          85   \n",
       "SRR3136628_percentage  SRR3136628  Bisulfite-Seq  GCA_000001635.1          83   \n",
       "\n",
       "                           Bases   BioProject     BioSample  \\\n",
       "Name PerCov                                                   \n",
       "SRR3136624_percentage  476309309  PRJNA310298  SAMN04449570   \n",
       "SRR3136625_percentage  276514225  PRJNA310298  SAMN04449998   \n",
       "SRR3136626_percentage  345410020  PRJNA310298  SAMN04449847   \n",
       "SRR3136627_percentage  450783127  PRJNA310298  SAMN04449846   \n",
       "SRR3136628_percentage  322920190  PRJNA310298  SAMN04449988   \n",
       "\n",
       "                                 BioSampleModel      Bytes  \\\n",
       "Name PerCov                                                  \n",
       "SRR3136624_percentage  Model organism or animal  305774648   \n",
       "SRR3136625_percentage  Model organism or animal  166102905   \n",
       "SRR3136626_percentage  Model organism or animal  197726064   \n",
       "SRR3136627_percentage  Model organism or animal  250903561   \n",
       "SRR3136628_percentage  Model organism or animal  214862269   \n",
       "\n",
       "                                               Center Name  ... Sample Name  \\\n",
       "Name PerCov                                                 ...               \n",
       "SRR3136624_percentage  ALBERT EINSTEIN COLLEGE OF MEDICINE  ...         f11   \n",
       "SRR3136625_percentage  ALBERT EINSTEIN COLLEGE OF MEDICINE  ...         f23   \n",
       "SRR3136626_percentage  ALBERT EINSTEIN COLLEGE OF MEDICINE  ...         f22   \n",
       "SRR3136627_percentage  ALBERT EINSTEIN COLLEGE OF MEDICINE  ...         f21   \n",
       "SRR3136628_percentage  ALBERT EINSTEIN COLLEGE OF MEDICINE  ...         o11   \n",
       "\n",
       "                        sex  SRA Study   strain           tissue   Age  \\\n",
       "Name PerCov                                                              \n",
       "SRR3136624_percentage  male  SRP069120  C57BL/6  MEFs in culture   0.0   \n",
       "SRR3136625_percentage  male  SRP069120  C57BL/6  MEFs in culture   0.0   \n",
       "SRR3136626_percentage  male  SRP069120  C57BL/6  MEFs in culture   0.0   \n",
       "SRR3136627_percentage  male  SRP069120  C57BL/6  MEFs in culture   0.0   \n",
       "SRR3136628_percentage  male  SRP069120  C57BL/6            Liver  26.0   \n",
       "\n",
       "                       dev_stage Age_months Mouse_ID   Data Type  \n",
       "Name PerCov                                                       \n",
       "SRR3136624_percentage  embryonic   0 months       f1  percentage  \n",
       "SRR3136625_percentage  embryonic   0 months       f2  percentage  \n",
       "SRR3136626_percentage  embryonic   0 months       f2  percentage  \n",
       "SRR3136627_percentage  embryonic   0 months       f2  percentage  \n",
       "SRR3136628_percentage        NaN   26 month       o1  percentage  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_csv(gl_metainfo_path, sep=',')\n",
    "df_meta['Age_months'] = df_meta['Age'].copy()\n",
    "df_meta['Age_months'].fillna('0 months', inplace=True)\n",
    "df_meta['Age'] = df_meta['Age_months'].apply(lambda x: float(x.split(' ')[0]))\n",
    "df_meta['Mouse_ID'] = df_meta['Sample Name'].apply(lambda x: x[:2])\n",
    "print(df_meta.shape)\n",
    "df_meta_2 = df_meta.copy()\n",
    "df_meta['Name PerCov'] = df_meta['Run'].str.cat(['percentage'] * df_meta.shape[0], sep='_')\n",
    "df_meta['Data Type'] = 'percentage'\n",
    "df_meta_2['Name PerCov'] = df_meta_2['Run'].str.cat(['coverage'] * df_meta.shape[0], sep='_')\n",
    "df_meta_2['Data Type'] = 'coverage'\n",
    "df_meta['Name PerCov'] = df_meta['Run'].str.cat(['percentage'] * df_meta.shape[0], sep='_')\n",
    "df_meta = pd.concat((df_meta, df_meta_2), axis=0)\n",
    "df_meta.set_index('Name PerCov', inplace=True)\n",
    "print(df_meta.shape)\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "b48559f5-2bc3-430d-b715-5aff25b3ca8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 68 entries, SRR3136624_percentage to SRR3136673_coverage\n",
      "Data columns (total 33 columns):\n",
      " #   Column              Non-Null Count  Dtype  \n",
      "---  ------              --------------  -----  \n",
      " 0   Run                 68 non-null     object \n",
      " 1   Assay Type          68 non-null     object \n",
      " 2   AssemblyName        68 non-null     object \n",
      " 3   AvgSpotLen          68 non-null     int64  \n",
      " 4   Bases               68 non-null     int64  \n",
      " 5   BioProject          68 non-null     object \n",
      " 6   BioSample           68 non-null     object \n",
      " 7   BioSampleModel      68 non-null     object \n",
      " 8   Bytes               68 non-null     int64  \n",
      " 9   Center Name         68 non-null     object \n",
      " 10  Consent             68 non-null     object \n",
      " 11  DATASTORE filetype  68 non-null     object \n",
      " 12  DATASTORE provider  68 non-null     object \n",
      " 13  DATASTORE region    68 non-null     object \n",
      " 14  Experiment          68 non-null     object \n",
      " 15  Instrument          68 non-null     object \n",
      " 16  Library Name        0 non-null      float64\n",
      " 17  LibraryLayout       68 non-null     object \n",
      " 18  LibrarySelection    68 non-null     object \n",
      " 19  LibrarySource       68 non-null     object \n",
      " 20  Organism            68 non-null     object \n",
      " 21  Platform            68 non-null     object \n",
      " 22  ReleaseDate         68 non-null     object \n",
      " 23  Sample Name         68 non-null     object \n",
      " 24  sex                 68 non-null     object \n",
      " 25  SRA Study           68 non-null     object \n",
      " 26  strain              68 non-null     object \n",
      " 27  tissue              68 non-null     object \n",
      " 28  Age                 68 non-null     float64\n",
      " 29  dev_stage           14 non-null     object \n",
      " 30  Age_months          68 non-null     object \n",
      " 31  Mouse_ID            68 non-null     object \n",
      " 32  Data Type           68 non-null     object \n",
      "dtypes: float64(2), int64(3), object(28)\n",
      "memory usage: 18.1+ KB\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 68 entries, SRR3136624_percentage to SRR3136673_coverage\n",
      "Data columns (total 33 columns):\n",
      " #   Column              Non-Null Count  Dtype   \n",
      "---  ------              --------------  -----   \n",
      " 0   Run                 68 non-null     category\n",
      " 1   Assay Type          68 non-null     category\n",
      " 2   AssemblyName        68 non-null     category\n",
      " 3   AvgSpotLen          68 non-null     category\n",
      " 4   Bases               68 non-null     category\n",
      " 5   BioProject          68 non-null     category\n",
      " 6   BioSample           68 non-null     category\n",
      " 7   BioSampleModel      68 non-null     category\n",
      " 8   Bytes               68 non-null     category\n",
      " 9   Center Name         68 non-null     category\n",
      " 10  Consent             68 non-null     category\n",
      " 11  DATASTORE filetype  68 non-null     category\n",
      " 12  DATASTORE provider  68 non-null     category\n",
      " 13  DATASTORE region    68 non-null     category\n",
      " 14  Experiment          68 non-null     category\n",
      " 15  Instrument          68 non-null     category\n",
      " 16  Library Name        0 non-null      category\n",
      " 17  LibraryLayout       68 non-null     category\n",
      " 18  LibrarySelection    68 non-null     category\n",
      " 19  LibrarySource       68 non-null     category\n",
      " 20  Organism            68 non-null     category\n",
      " 21  Platform            68 non-null     category\n",
      " 22  ReleaseDate         68 non-null     category\n",
      " 23  Sample Name         68 non-null     category\n",
      " 24  sex                 68 non-null     category\n",
      " 25  SRA Study           68 non-null     category\n",
      " 26  strain              68 non-null     category\n",
      " 27  tissue              68 non-null     category\n",
      " 28  Age                 68 non-null     float64 \n",
      " 29  dev_stage           14 non-null     category\n",
      " 30  Age_months          68 non-null     category\n",
      " 31  Mouse_ID            68 non-null     category\n",
      " 32  Data Type           68 non-null     category\n",
      "dtypes: category(32), float64(1)\n",
      "memory usage: 14.8+ KB\n"
     ]
    }
   ],
   "source": [
    "df_meta.info()\n",
    "to_category = np.setdiff1d(df_meta.columns, ['Age'])\n",
    "df_meta[to_category] = df_meta[to_category].astype(\"category\")\n",
    "df_meta.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "e8e915c4-b1c8-41ec-a194-872c464e7a06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/Dropbox/Harvard/Science/scDNAm/data/SRA344045/SRA344045_coverage_0_binned_by_1.h5ad\n"
     ]
    }
   ],
   "source": [
    "data_path = Path.joinpath(cwd, project_dir, \"{}_coverage_0_binned_by_1.h5ad\".format(project_id))\n",
    "print(data_path)\n",
    "ad = sc.read_h5ad(data_path)\n",
    "ad.var = pd.concat((ad.var, df_meta.loc[ad.var.index,:]), axis=1)\n",
    "ad.X = ad.X.astype(np.float32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "4489aed1-a931-4d76-92f0-17fae825fadb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ad.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "f2f255c7-74a4-4923-83a8-0fad229617fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad[:,ad.var.tissue.values == 'Liver'].write_h5ad(Path.joinpath(project_dir, \"{}_Liver_wmeta.h5ad\".format(project_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "8d3e6097-bbdc-472c-8449-f82467ad7387",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad[:,ad.var.tissue.values != 'Liver'].write_h5ad(Path.joinpath(project_dir, \"{}_MEF_wmeta.h5ad\".format(project_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "a0825f9a-0ab0-4593-a7dd-6b417722158b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad.write_h5ad(Path.joinpath(project_dir, \"{}_wmeta.h5ad\".format(project_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "ef453719-b4b6-4cb2-8db4-39815e203593",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SRR3136634_percentage    f1\n",
       "SRR3136634_coverage      f1\n",
       "SRR3136624_percentage    f1\n",
       "SRR3136624_coverage      f1\n",
       "SRR3136627_coverage      f2\n",
       "SRR3136627_percentage    f2\n",
       "SRR3136626_coverage      f2\n",
       "SRR3136626_percentage    f2\n",
       "SRR3136625_percentage    f2\n",
       "SRR3136625_coverage      f2\n",
       "SRR3136630_percentage    o1\n",
       "SRR3136653_coverage      o1\n",
       "SRR3136653_percentage    o1\n",
       "SRR3136630_coverage      o1\n",
       "SRR3136629_percentage    o1\n",
       "SRR3136629_coverage      o1\n",
       "SRR3136628_coverage      o1\n",
       "SRR3136646_percentage    o1\n",
       "SRR3136651_percentage    o1\n",
       "SRR3136651_coverage      o1\n",
       "SRR3136646_coverage      o1\n",
       "SRR3136628_percentage    o1\n",
       "SRR3136631_coverage      o2\n",
       "SRR3136631_percentage    o2\n",
       "SRR3136635_coverage      o2\n",
       "SRR3136635_percentage    o2\n",
       "SRR3136647_coverage      o3\n",
       "SRR3136652_percentage    o3\n",
       "SRR3136652_coverage      o3\n",
       "SRR3136647_percentage    o3\n",
       "SRR3136658_coverage      y1\n",
       "SRR3136658_percentage    y1\n",
       "SRR3136655_percentage    y1\n",
       "SRR3136660_coverage      y1\n",
       "SRR3136660_percentage    y1\n",
       "SRR3136655_coverage      y1\n",
       "SRR3136654_coverage      y1\n",
       "SRR3136654_percentage    y1\n",
       "SRR3136656_coverage      y1\n",
       "SRR3136656_percentage    y1\n",
       "SRR3136663_coverage      y1\n",
       "SRR3136663_percentage    y1\n",
       "SRR3136657_coverage      y2\n",
       "SRR3136657_percentage    y2\n",
       "SRR3136659_coverage      y2\n",
       "SRR3136659_percentage    y2\n",
       "SRR3136661_percentage    y3\n",
       "SRR3136662_percentage    y3\n",
       "SRR3136664_coverage      y3\n",
       "SRR3136664_percentage    y3\n",
       "SRR3136661_coverage      y3\n",
       "SRR3136662_coverage      y3\n",
       "Name: Mouse_ID, dtype: object"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mouse_ids = ad.var.loc[::2,:].Mouse_ID.values\n",
    "ad.var['Mouse_ID'] = ad.var['Sample Name'].apply(lambda x: x[:2])\n",
    "# ad.var.loc[::2,:].columns\n",
    "# ad.var.iloc[:,5:].head(3)\n",
    "ad.var['Mouse_ID'].sort_values()#value_counts()#iloc[:,5:].head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "45a3db65-7c47-4bdf-8f41-6cc977df66f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['f1' 'f2' 'o1' 'o2' 'o3' 'y1' 'y2' 'y3']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/v_/gtsrj3xs0_131ltxwg51r_f40000gn/T/ipykernel_1713/3837489246.py:12: RuntimeWarning: invalid value encountered in true_divide\n",
      "  xbulk[:,2*i] = xmeth[:,idx].sum(axis=1).flatten() / xcov[:,idx].sum(axis=1).flatten()\n"
     ]
    }
   ],
   "source": [
    "x = ad.X[:,::2]\n",
    "xcov = ad.X[:,1::2]\n",
    "xmeth = x.multiply(xcov)\n",
    "\n",
    "mouse_ids = ad.var.loc[::2,:].Mouse_ID.values\n",
    "mice = np.unique(mouse_ids)\n",
    "print(mice)\n",
    "xbulk = np.zeros((ad.X.shape[0], 2* len(mice)))\n",
    "#xcovbulk = np.zeros((ad.X.shape[0], len(mice)))\n",
    "for i,mouse_id in enumerate(mice):\n",
    "    idx = mouse_ids == mouse_id\n",
    "    xbulk[:,2*i] = xmeth[:,idx].sum(axis=1).flatten() / xcov[:,idx].sum(axis=1).flatten()\n",
    "    xbulk[:,2*i+1] = xcov[:,idx].sum(axis=1).flatten()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "35f00b17-8d42-4c9d-9241-bb3998c4af5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mice_index = np.array([[f\"{x}_percentage\", f\"{x}_coverage\"] for x in mice]).flatten()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "166f7f32-7158-4335-a1e0-c80e184aeb09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(16, 33)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mouse_ID</th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AssemblyName</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleModel</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>...</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>sex</th>\n",
       "      <th>SRA Study</th>\n",
       "      <th>strain</th>\n",
       "      <th>tissue</th>\n",
       "      <th>Age</th>\n",
       "      <th>dev_stage</th>\n",
       "      <th>Age_months</th>\n",
       "      <th>Data Type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name PerCov</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>f1_percentage</th>\n",
       "      <td>f1</td>\n",
       "      <td>SRR3136624</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>42</td>\n",
       "      <td>476309309</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449570</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>305774648</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>f11</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>MEFs in culture</td>\n",
       "      <td>0.0</td>\n",
       "      <td>embryonic</td>\n",
       "      <td>0 months</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f1_coverage</th>\n",
       "      <td>f1</td>\n",
       "      <td>SRR3136624</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>42</td>\n",
       "      <td>476309309</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449570</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>305774648</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>f11</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>MEFs in culture</td>\n",
       "      <td>0.0</td>\n",
       "      <td>embryonic</td>\n",
       "      <td>0 months</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f2_percentage</th>\n",
       "      <td>f2</td>\n",
       "      <td>SRR3136625</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>82</td>\n",
       "      <td>276514225</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449998</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>166102905</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>f23</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>MEFs in culture</td>\n",
       "      <td>0.0</td>\n",
       "      <td>embryonic</td>\n",
       "      <td>0 months</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f2_coverage</th>\n",
       "      <td>f2</td>\n",
       "      <td>SRR3136625</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>82</td>\n",
       "      <td>276514225</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449998</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>166102905</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>f23</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>MEFs in culture</td>\n",
       "      <td>0.0</td>\n",
       "      <td>embryonic</td>\n",
       "      <td>0 months</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>o1_percentage</th>\n",
       "      <td>o1</td>\n",
       "      <td>SRR3136628</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>83</td>\n",
       "      <td>322920190</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449988</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>214862269</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>o11</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>Liver</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26 month</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>o1_coverage</th>\n",
       "      <td>o1</td>\n",
       "      <td>SRR3136628</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>83</td>\n",
       "      <td>322920190</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449988</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>214862269</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>o11</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>Liver</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26 month</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>o2_percentage</th>\n",
       "      <td>o2</td>\n",
       "      <td>SRR3136631</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>84</td>\n",
       "      <td>204738803</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449994</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>140159990</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>o21</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>Liver</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26 month</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>o2_coverage</th>\n",
       "      <td>o2</td>\n",
       "      <td>SRR3136631</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>84</td>\n",
       "      <td>204738803</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449994</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>140159990</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>o21</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>Liver</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26 month</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>o3_percentage</th>\n",
       "      <td>o3</td>\n",
       "      <td>SRR3136647</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>83</td>\n",
       "      <td>212610316</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449996</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>127381556</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>o31</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>Liver</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26 month</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>o3_coverage</th>\n",
       "      <td>o3</td>\n",
       "      <td>SRR3136647</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>83</td>\n",
       "      <td>212610316</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449996</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>127381556</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>o31</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>Liver</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26 month</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>y1_percentage</th>\n",
       "      <td>y1</td>\n",
       "      <td>SRR3136654</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>81</td>\n",
       "      <td>192636765</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449977</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>137041287</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>y11</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>Liver</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4 month</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>y1_coverage</th>\n",
       "      <td>y1</td>\n",
       "      <td>SRR3136654</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>81</td>\n",
       "      <td>192636765</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449977</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>137041287</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>y11</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>Liver</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4 month</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>y2_percentage</th>\n",
       "      <td>y2</td>\n",
       "      <td>SRR3136657</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>42</td>\n",
       "      <td>58472193</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449984</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>54511577</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>y22</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>Liver</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4 month</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>y2_coverage</th>\n",
       "      <td>y2</td>\n",
       "      <td>SRR3136657</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>42</td>\n",
       "      <td>58472193</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449984</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>54511577</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>y22</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>Liver</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4 month</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>y3_percentage</th>\n",
       "      <td>y3</td>\n",
       "      <td>SRR3136661</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>86</td>\n",
       "      <td>80915262</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449986</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>58393054</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>y32</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>Liver</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4 month</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>y3_coverage</th>\n",
       "      <td>y3</td>\n",
       "      <td>SRR3136661</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>GCA_000001635.1</td>\n",
       "      <td>86</td>\n",
       "      <td>80915262</td>\n",
       "      <td>PRJNA310298</td>\n",
       "      <td>SAMN04449986</td>\n",
       "      <td>Model organism or animal</td>\n",
       "      <td>58393054</td>\n",
       "      <td>...</td>\n",
       "      <td>2016-03-05T00:00:00Z</td>\n",
       "      <td>y32</td>\n",
       "      <td>male</td>\n",
       "      <td>SRP069120</td>\n",
       "      <td>C57BL/6</td>\n",
       "      <td>Liver</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4 month</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16 rows  33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Mouse_ID         Run     Assay Type     AssemblyName AvgSpotLen  \\\n",
       "Name PerCov                                                                     \n",
       "f1_percentage       f1  SRR3136624  Bisulfite-Seq  GCA_000001635.1         42   \n",
       "f1_coverage         f1  SRR3136624  Bisulfite-Seq  GCA_000001635.1         42   \n",
       "f2_percentage       f2  SRR3136625  Bisulfite-Seq  GCA_000001635.1         82   \n",
       "f2_coverage         f2  SRR3136625  Bisulfite-Seq  GCA_000001635.1         82   \n",
       "o1_percentage       o1  SRR3136628  Bisulfite-Seq  GCA_000001635.1         83   \n",
       "o1_coverage         o1  SRR3136628  Bisulfite-Seq  GCA_000001635.1         83   \n",
       "o2_percentage       o2  SRR3136631  Bisulfite-Seq  GCA_000001635.1         84   \n",
       "o2_coverage         o2  SRR3136631  Bisulfite-Seq  GCA_000001635.1         84   \n",
       "o3_percentage       o3  SRR3136647  Bisulfite-Seq  GCA_000001635.1         83   \n",
       "o3_coverage         o3  SRR3136647  Bisulfite-Seq  GCA_000001635.1         83   \n",
       "y1_percentage       y1  SRR3136654  Bisulfite-Seq  GCA_000001635.1         81   \n",
       "y1_coverage         y1  SRR3136654  Bisulfite-Seq  GCA_000001635.1         81   \n",
       "y2_percentage       y2  SRR3136657  Bisulfite-Seq  GCA_000001635.1         42   \n",
       "y2_coverage         y2  SRR3136657  Bisulfite-Seq  GCA_000001635.1         42   \n",
       "y3_percentage       y3  SRR3136661  Bisulfite-Seq  GCA_000001635.1         86   \n",
       "y3_coverage         y3  SRR3136661  Bisulfite-Seq  GCA_000001635.1         86   \n",
       "\n",
       "                   Bases   BioProject     BioSample            BioSampleModel  \\\n",
       "Name PerCov                                                                     \n",
       "f1_percentage  476309309  PRJNA310298  SAMN04449570  Model organism or animal   \n",
       "f1_coverage    476309309  PRJNA310298  SAMN04449570  Model organism or animal   \n",
       "f2_percentage  276514225  PRJNA310298  SAMN04449998  Model organism or animal   \n",
       "f2_coverage    276514225  PRJNA310298  SAMN04449998  Model organism or animal   \n",
       "o1_percentage  322920190  PRJNA310298  SAMN04449988  Model organism or animal   \n",
       "o1_coverage    322920190  PRJNA310298  SAMN04449988  Model organism or animal   \n",
       "o2_percentage  204738803  PRJNA310298  SAMN04449994  Model organism or animal   \n",
       "o2_coverage    204738803  PRJNA310298  SAMN04449994  Model organism or animal   \n",
       "o3_percentage  212610316  PRJNA310298  SAMN04449996  Model organism or animal   \n",
       "o3_coverage    212610316  PRJNA310298  SAMN04449996  Model organism or animal   \n",
       "y1_percentage  192636765  PRJNA310298  SAMN04449977  Model organism or animal   \n",
       "y1_coverage    192636765  PRJNA310298  SAMN04449977  Model organism or animal   \n",
       "y2_percentage   58472193  PRJNA310298  SAMN04449984  Model organism or animal   \n",
       "y2_coverage     58472193  PRJNA310298  SAMN04449984  Model organism or animal   \n",
       "y3_percentage   80915262  PRJNA310298  SAMN04449986  Model organism or animal   \n",
       "y3_coverage     80915262  PRJNA310298  SAMN04449986  Model organism or animal   \n",
       "\n",
       "                   Bytes  ...           ReleaseDate Sample Name   sex  \\\n",
       "Name PerCov               ...                                           \n",
       "f1_percentage  305774648  ...  2016-03-05T00:00:00Z         f11  male   \n",
       "f1_coverage    305774648  ...  2016-03-05T00:00:00Z         f11  male   \n",
       "f2_percentage  166102905  ...  2016-03-05T00:00:00Z         f23  male   \n",
       "f2_coverage    166102905  ...  2016-03-05T00:00:00Z         f23  male   \n",
       "o1_percentage  214862269  ...  2016-03-05T00:00:00Z         o11  male   \n",
       "o1_coverage    214862269  ...  2016-03-05T00:00:00Z         o11  male   \n",
       "o2_percentage  140159990  ...  2016-03-05T00:00:00Z         o21  male   \n",
       "o2_coverage    140159990  ...  2016-03-05T00:00:00Z         o21  male   \n",
       "o3_percentage  127381556  ...  2016-03-05T00:00:00Z         o31  male   \n",
       "o3_coverage    127381556  ...  2016-03-05T00:00:00Z         o31  male   \n",
       "y1_percentage  137041287  ...  2016-03-05T00:00:00Z         y11  male   \n",
       "y1_coverage    137041287  ...  2016-03-05T00:00:00Z         y11  male   \n",
       "y2_percentage   54511577  ...  2016-03-05T00:00:00Z         y22  male   \n",
       "y2_coverage     54511577  ...  2016-03-05T00:00:00Z         y22  male   \n",
       "y3_percentage   58393054  ...  2016-03-05T00:00:00Z         y32  male   \n",
       "y3_coverage     58393054  ...  2016-03-05T00:00:00Z         y32  male   \n",
       "\n",
       "               SRA Study   strain           tissue   Age  dev_stage  \\\n",
       "Name PerCov                                                           \n",
       "f1_percentage  SRP069120  C57BL/6  MEFs in culture   0.0  embryonic   \n",
       "f1_coverage    SRP069120  C57BL/6  MEFs in culture   0.0  embryonic   \n",
       "f2_percentage  SRP069120  C57BL/6  MEFs in culture   0.0  embryonic   \n",
       "f2_coverage    SRP069120  C57BL/6  MEFs in culture   0.0  embryonic   \n",
       "o1_percentage  SRP069120  C57BL/6            Liver  26.0        NaN   \n",
       "o1_coverage    SRP069120  C57BL/6            Liver  26.0        NaN   \n",
       "o2_percentage  SRP069120  C57BL/6            Liver  26.0        NaN   \n",
       "o2_coverage    SRP069120  C57BL/6            Liver  26.0        NaN   \n",
       "o3_percentage  SRP069120  C57BL/6            Liver  26.0        NaN   \n",
       "o3_coverage    SRP069120  C57BL/6            Liver  26.0        NaN   \n",
       "y1_percentage  SRP069120  C57BL/6            Liver   4.0        NaN   \n",
       "y1_coverage    SRP069120  C57BL/6            Liver   4.0        NaN   \n",
       "y2_percentage  SRP069120  C57BL/6            Liver   4.0        NaN   \n",
       "y2_coverage    SRP069120  C57BL/6            Liver   4.0        NaN   \n",
       "y3_percentage  SRP069120  C57BL/6            Liver   4.0        NaN   \n",
       "y3_coverage    SRP069120  C57BL/6            Liver   4.0        NaN   \n",
       "\n",
       "              Age_months   Data Type  \n",
       "Name PerCov                           \n",
       "f1_percentage   0 months  percentage  \n",
       "f1_coverage     0 months    coverage  \n",
       "f2_percentage   0 months  percentage  \n",
       "f2_coverage     0 months    coverage  \n",
       "o1_percentage   26 month  percentage  \n",
       "o1_coverage     26 month    coverage  \n",
       "o2_percentage   26 month  percentage  \n",
       "o2_coverage     26 month    coverage  \n",
       "o3_percentage   26 month  percentage  \n",
       "o3_coverage     26 month    coverage  \n",
       "y1_percentage    4 month  percentage  \n",
       "y1_coverage      4 month    coverage  \n",
       "y2_percentage    4 month  percentage  \n",
       "y2_coverage      4 month    coverage  \n",
       "y3_percentage    4 month  percentage  \n",
       "y3_coverage      4 month    coverage  \n",
       "\n",
       "[16 rows x 33 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta_bulk = df_meta.drop_duplicates('Mouse_ID').set_index('Mouse_ID').loc[mice].copy()\n",
    "df_meta_bulk.reset_index(inplace=True)\n",
    "df_meta_2 = df_meta_bulk.copy()\n",
    "df_meta_bulk['Name PerCov'] = df_meta_bulk['Mouse_ID'].str.cat(['percentage'] * df_meta_bulk.shape[0], sep='_')\n",
    "df_meta_bulk['Data Type'] = 'percentage'\n",
    "df_meta_2['Name PerCov'] = df_meta_2['Mouse_ID'].str.cat(['coverage'] * df_meta_bulk.shape[0], sep='_')\n",
    "df_meta_2['Data Type'] = 'coverage'\n",
    "#df_meta['Name PerCov'] = df_meta['Sample Name'].str.cat(['percentage'] * df_meta.shape[0], sep='_')\n",
    "df_meta_bulk = pd.concat((df_meta_bulk, df_meta_2), axis=0)\n",
    "df_meta_bulk.set_index('Name PerCov', inplace=True)\n",
    "print(df_meta_bulk.shape)\n",
    "df_meta_bulk = df_meta_bulk.loc[mice_index]\n",
    "df_meta_bulk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "139df911-6218-472c-867e-3bea4e1e4cc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Data Type' as categorical\n"
     ]
    }
   ],
   "source": [
    "adbu = sc.AnnData(sp.csr_matrix(np.nan_to_num(xbulk)), ad.obs, df_meta_bulk)\n",
    "adbu.write_h5ad(Path.joinpath(project_dir, \"{}_bulked_wmeta_mm10.h5ad\".format(project_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e76d5c4-9115-423d-96f7-307e2c335cf1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "479b3d8e-9b45-48a9-bfdd-2596359d9c15",
   "metadata": {},
   "source": [
    "# GSE121364 scRNAseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "f1a609c8-9240-4a70-afac-39bbe7731b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwd = Path.cwd()\n",
    "# cwd = Path('/Users/tarkhov/Dropbox/Harvard/Science/Church/')\n",
    "project_dir = Path.joinpath(cwd, Path('data/GSE121364/'))\n",
    "project_id = 'GSE121364'\n",
    "project_samples_dir = 'GSE121364_RAW'\n",
    "metainfo_path = 'SraRunTable.txt'\n",
    "metainfo_path_2 = 'sample.csv'\n",
    "gl_metainfo_path = Path.joinpath(project_dir, metainfo_path)\n",
    "gl_metainfo_path_2 = Path.joinpath(project_dir, metainfo_path_2)\n",
    "data_path = Path.joinpath(project_dir, \"GSE121364_Table_raw_counts_RNA_QSC.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "a234edb0-326f-4630-a48d-4a54e98b73f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(25952, 380)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3251: DtypeWarning: Columns (1) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneName</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>lane3_Tom_Y4_TA_Hi_11_CTATCATG_L003_22540_7_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane3_Tom_Y4_TA_Hi_12_TCTCAGTG_L003_22540_7_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane3_Tom_Y4_TA_Hi_13_GCCTGCCC_L003_22540_7_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane3_Tom_Y4_TA_Hi_1_TTCCTGCT_L003_22540_7_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane3_Tom_Y4_TA_Hi_23_CGCCTGAT_L003_22540_7_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane3_Tom_Y4_TA_Hi_25_TTCCATTG_L003_22540_7_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane3_Tom_Y4_TA_Hi_27_CTAGATAG_L003_22540_7_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>...</th>\n",
       "      <th>lane2_Tom_O5_TA_Hi_84_GTTCCACG_L002_22540_5_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane2_Tom_O5_TA_Hi_85_TTCCTACA_L002_22540_5_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane2_Tom_O5_TA_Hi_86_TCCGGGTA_L002_22540_5_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane2_Tom_O5_TA_Hi_87_GGACACGG_L002_22540_5_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane2_Tom_O5_TA_Hi_89_TACTAGTC_L002_22540_5_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane2_Tom_O5_TA_Hi_90_GATTCATC_L002_22540_5_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane2_Tom_O5_TA_Hi_91_CGACTGCA_L002_22540_5_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane2_Tom_O5_TA_Hi_92_TATTGGAG_L002_22540_5_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane2_Tom_O5_TA_Hi_94_TCGGACCG_L002_22540_5_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "      <th>lane2_Tom_O5_TA_Hi_95_CTTCGGCT_L002_22540_5_R1_val_1_GRCm38_hisat2.bam</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>1</td>\n",
       "      <td>3205901</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gm1992</td>\n",
       "      <td>1</td>\n",
       "      <td>3466587</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows  380 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  geneName chr    start  \\\n",
       "0     Xkr4   1  3205901   \n",
       "1   Gm1992   1  3466587   \n",
       "\n",
       "   lane3_Tom_Y4_TA_Hi_11_CTATCATG_L003_22540_7_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   lane3_Tom_Y4_TA_Hi_12_TCTCAGTG_L003_22540_7_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   lane3_Tom_Y4_TA_Hi_13_GCCTGCCC_L003_22540_7_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   lane3_Tom_Y4_TA_Hi_1_TTCCTGCT_L003_22540_7_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                       \n",
       "1                                                  0                       \n",
       "\n",
       "   lane3_Tom_Y4_TA_Hi_23_CGCCTGAT_L003_22540_7_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   lane3_Tom_Y4_TA_Hi_25_TTCCATTG_L003_22540_7_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   lane3_Tom_Y4_TA_Hi_27_CTAGATAG_L003_22540_7_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   ...  \\\n",
       "0  ...   \n",
       "1  ...   \n",
       "\n",
       "   lane2_Tom_O5_TA_Hi_84_GTTCCACG_L002_22540_5_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   lane2_Tom_O5_TA_Hi_85_TTCCTACA_L002_22540_5_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   lane2_Tom_O5_TA_Hi_86_TCCGGGTA_L002_22540_5_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   lane2_Tom_O5_TA_Hi_87_GGACACGG_L002_22540_5_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   lane2_Tom_O5_TA_Hi_89_TACTAGTC_L002_22540_5_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   lane2_Tom_O5_TA_Hi_90_GATTCATC_L002_22540_5_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   lane2_Tom_O5_TA_Hi_91_CGACTGCA_L002_22540_5_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   lane2_Tom_O5_TA_Hi_92_TATTGGAG_L002_22540_5_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   lane2_Tom_O5_TA_Hi_94_TCGGACCG_L002_22540_5_R1_val_1_GRCm38_hisat2.bam  \\\n",
       "0                                                  0                        \n",
       "1                                                  0                        \n",
       "\n",
       "   lane2_Tom_O5_TA_Hi_95_CTTCGGCT_L002_22540_5_R1_val_1_GRCm38_hisat2.bam  \n",
       "0                                                  0                       \n",
       "1                                                  0                       \n",
       "\n",
       "[2 rows x 380 columns]"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(data_path, sep=' ')\n",
    "print(df.shape)\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "ef8f20cc-0f12-4f9e-9be6-3ee223bcc354",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>Consent</th>\n",
       "      <th>DATASTORE filetype</th>\n",
       "      <th>...</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>SRA Study</th>\n",
       "      <th>Age</th>\n",
       "      <th>Cell_type</th>\n",
       "      <th>Mouse_ID</th>\n",
       "      <th>source_name</th>\n",
       "      <th>STRAIN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8066856</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>1169700</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10249431</td>\n",
       "      <td>494375</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-14T00:00:00Z</td>\n",
       "      <td>GSM3432562</td>\n",
       "      <td>SRP165935</td>\n",
       "      <td>24 months</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>O1</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8066857</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>147409200</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10249430</td>\n",
       "      <td>56823661</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>sra,fastq</td>\n",
       "      <td>...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-14T00:00:00Z</td>\n",
       "      <td>GSM3432563</td>\n",
       "      <td>SRP165935</td>\n",
       "      <td>24 months</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>O1</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows  28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Run Assay Type  AvgSpotLen      Bases   BioProject     BioSample  \\\n",
       "0  SRR8066856    RNA-Seq         150    1169700  PRJNA497128  SAMN10249431   \n",
       "1  SRR8066857    RNA-Seq         150  147409200  PRJNA497128  SAMN10249430   \n",
       "\n",
       "      Bytes Center Name Consent DATASTORE filetype  ...      Organism  \\\n",
       "0    494375         GEO  public          fastq,sra  ...  Mus musculus   \n",
       "1  56823661         GEO  public          sra,fastq  ...  Mus musculus   \n",
       "\n",
       "   Platform           ReleaseDate Sample Name  SRA Study        Age  \\\n",
       "0  ILLUMINA  2019-08-14T00:00:00Z  GSM3432562  SRP165935  24 months   \n",
       "1  ILLUMINA  2019-08-14T00:00:00Z  GSM3432563  SRP165935  24 months   \n",
       "\n",
       "           Cell_type Mouse_ID                source_name        STRAIN  \n",
       "0  Muscle stem cells       O1  Tibialis anterior muscles  MGI: 5308730  \n",
       "1  Muscle stem cells       O1  Tibialis anterior muscles  MGI: 5308730  \n",
       "\n",
       "[2 rows x 28 columns]"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_csv(gl_metainfo_path, sep=',')\n",
    "df_meta.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "49c01cf8-41c5-45ff-93ef-88c3f2dfd77e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Title</th>\n",
       "      <th>Sample Type</th>\n",
       "      <th>Taxonomy</th>\n",
       "      <th>Channels</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Series</th>\n",
       "      <th>Supplementary Types</th>\n",
       "      <th>Supplementary Links</th>\n",
       "      <th>SRA Accession</th>\n",
       "      <th>Contact</th>\n",
       "      <th>Release Date</th>\n",
       "      <th>CellID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSM3432562</td>\n",
       "      <td>O1_TA_Hi_10 [scRNA-seq]</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895195</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "      <td>O1_TA_Hi_10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSM3432563</td>\n",
       "      <td>O1_TA_Hi_11 [scRNA-seq]</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895196</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "      <td>O1_TA_Hi_11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Accession                    Title Sample Type      Taxonomy  Channels  \\\n",
       "0  GSM3432562  O1_TA_Hi_10 [scRNA-seq]         SRA  Mus musculus         1   \n",
       "1  GSM3432563  O1_TA_Hi_11 [scRNA-seq]         SRA  Mus musculus         1   \n",
       "\n",
       "   Platform               Series Supplementary Types  \\\n",
       "0  GPL17021  GSE121364;GSE121437    SRA Run Selector   \n",
       "1  GPL17021  GSE121364;GSE121437    SRA Run Selector   \n",
       "\n",
       "                                 Supplementary Links SRA Accession  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895195   \n",
       "1  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895196   \n",
       "\n",
       "         Contact  Release Date       CellID  \n",
       "0  Felix Krueger  Aug 14, 2019  O1_TA_Hi_10  \n",
       "1  Felix Krueger  Aug 14, 2019  O1_TA_Hi_11  "
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta_2 = pd.read_csv(gl_metainfo_path_2, sep=',')\n",
    "df_meta_2['CellID'] = df_meta_2['Title'].apply(lambda x: x.split(' ')[0])\n",
    "df_meta_2.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "2bd856ec-05e0-42ed-820b-fccb1af01cfc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>Consent</th>\n",
       "      <th>DATASTORE filetype</th>\n",
       "      <th>...</th>\n",
       "      <th>Taxonomy</th>\n",
       "      <th>Channels</th>\n",
       "      <th>Platform_y</th>\n",
       "      <th>Series</th>\n",
       "      <th>Supplementary Types</th>\n",
       "      <th>Supplementary Links</th>\n",
       "      <th>SRA Accession</th>\n",
       "      <th>Contact</th>\n",
       "      <th>Release Date</th>\n",
       "      <th>CellID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8066856</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>1169700</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10249431</td>\n",
       "      <td>494375</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895195</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "      <td>O1_TA_Hi_10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8066857</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>147409200</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10249430</td>\n",
       "      <td>56823661</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>sra,fastq</td>\n",
       "      <td>...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895196</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "      <td>O1_TA_Hi_11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows  41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Run Assay Type  AvgSpotLen      Bases   BioProject     BioSample  \\\n",
       "0  SRR8066856    RNA-Seq         150    1169700  PRJNA497128  SAMN10249431   \n",
       "1  SRR8066857    RNA-Seq         150  147409200  PRJNA497128  SAMN10249430   \n",
       "\n",
       "      Bytes Center Name Consent DATASTORE filetype  ...      Taxonomy  \\\n",
       "0    494375         GEO  public          fastq,sra  ...  Mus musculus   \n",
       "1  56823661         GEO  public          sra,fastq  ...  Mus musculus   \n",
       "\n",
       "  Channels Platform_y               Series Supplementary Types  \\\n",
       "0        1   GPL17021  GSE121364;GSE121437    SRA Run Selector   \n",
       "1        1   GPL17021  GSE121364;GSE121437    SRA Run Selector   \n",
       "\n",
       "                                 Supplementary Links SRA Accession  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895195   \n",
       "1  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895196   \n",
       "\n",
       "         Contact  Release Date       CellID  \n",
       "0  Felix Krueger  Aug 14, 2019  O1_TA_Hi_10  \n",
       "1  Felix Krueger  Aug 14, 2019  O1_TA_Hi_11  \n",
       "\n",
       "[2 rows x 41 columns]"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta_full = pd.merge(df_meta, df_meta_2, left_on='Sample Name', right_on='Accession')\n",
    "df_meta_full.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "fa2f8c61-defa-47ef-b5f6-2bfb1ac39344",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_meta_full['CellID'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "5cfa71c4-d47b-45d9-a0d2-2743ff2b0ecb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['lane3_Tom_Y4_TA_Hi_11_CTATCATG_L003_22540_7_R1_val_1_GRCm38_hisat2.bam',\n",
      "       'lane3_Tom_Y4_TA_Hi_12_TCTCAGTG_L003_22540_7_R1_val_1_GRCm38_hisat2.bam',\n",
      "       'lane3_Tom_Y4_TA_Hi_13_GCCTGCCC_L003_22540_7_R1_val_1_GRCm38_hisat2.bam',\n",
      "       'lane3_Tom_Y4_TA_Hi_1_TTCCTGCT_L003_22540_7_R1_val_1_GRCm38_hisat2.bam',\n",
      "       'lane3_Tom_Y4_TA_Hi_23_CGCCTGAT_L003_22540_7_R1_val_1_GRCm38_hisat2.bam'],\n",
      "      dtype='object')\n",
      "['Y4_TA_Hi_11', 'Y4_TA_Hi_12', 'Y4_TA_Hi_13', 'Y4_TA_Hi_1', 'Y4_TA_Hi_23']\n"
     ]
    }
   ],
   "source": [
    "sampleIDs = df.columns[3:]\n",
    "print(sampleIDs[:5])\n",
    "sampleIDs = ['_'.join(x.split('_')[2:6]) for x in sampleIDs]\n",
    "print(sampleIDs[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "00fd2839-3d02-452a-b5fe-ed61f3b0cac5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['geneName', 'chr', 'start', 'Y4_TA_Hi_11', 'Y4_TA_Hi_12', 'Y4_TA_Hi_13', 'Y4_TA_Hi_1']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneName</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>Y4_TA_Hi_11</th>\n",
       "      <th>Y4_TA_Hi_12</th>\n",
       "      <th>Y4_TA_Hi_13</th>\n",
       "      <th>Y4_TA_Hi_1</th>\n",
       "      <th>Y4_TA_Hi_23</th>\n",
       "      <th>Y4_TA_Hi_25</th>\n",
       "      <th>Y4_TA_Hi_27</th>\n",
       "      <th>...</th>\n",
       "      <th>O5_TA_Hi_84</th>\n",
       "      <th>O5_TA_Hi_85</th>\n",
       "      <th>O5_TA_Hi_86</th>\n",
       "      <th>O5_TA_Hi_87</th>\n",
       "      <th>O5_TA_Hi_89</th>\n",
       "      <th>O5_TA_Hi_90</th>\n",
       "      <th>O5_TA_Hi_91</th>\n",
       "      <th>O5_TA_Hi_92</th>\n",
       "      <th>O5_TA_Hi_94</th>\n",
       "      <th>O5_TA_Hi_95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>1</td>\n",
       "      <td>3205901</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gm1992</td>\n",
       "      <td>1</td>\n",
       "      <td>3466587</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rp1</td>\n",
       "      <td>1</td>\n",
       "      <td>4343507</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sox17</td>\n",
       "      <td>1</td>\n",
       "      <td>4490928</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>1</td>\n",
       "      <td>4773206</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  380 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  geneName chr    start  Y4_TA_Hi_11  Y4_TA_Hi_12  Y4_TA_Hi_13  Y4_TA_Hi_1  \\\n",
       "0     Xkr4   1  3205901            0            0            0           0   \n",
       "1   Gm1992   1  3466587            0            0            0           0   \n",
       "2      Rp1   1  4343507            0            0            0           0   \n",
       "3    Sox17   1  4490928            0            0            0           0   \n",
       "4   Mrpl15   1  4773206            0            0          121           0   \n",
       "\n",
       "   Y4_TA_Hi_23  Y4_TA_Hi_25  Y4_TA_Hi_27  ...  O5_TA_Hi_84  O5_TA_Hi_85  \\\n",
       "0            0            0            0  ...            0            0   \n",
       "1            0            0            0  ...            0            0   \n",
       "2            0            0            0  ...            0            0   \n",
       "3            0            0            0  ...            0            0   \n",
       "4            0            0            0  ...            0            0   \n",
       "\n",
       "   O5_TA_Hi_86  O5_TA_Hi_87  O5_TA_Hi_89  O5_TA_Hi_90  O5_TA_Hi_91  \\\n",
       "0            0            0            0            0            0   \n",
       "1            0            0            0            0            0   \n",
       "2            0            0            0            0            0   \n",
       "3            0            0            0            0            0   \n",
       "4            0            0            0            0            0   \n",
       "\n",
       "   O5_TA_Hi_92  O5_TA_Hi_94  O5_TA_Hi_95  \n",
       "0            0            0            0  \n",
       "1            0            0            0  \n",
       "2            0            0            0  \n",
       "3            0            0            0  \n",
       "4            0            0            0  \n",
       "\n",
       "[5 rows x 380 columns]"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_cols = df.columns[:3].tolist() + sampleIDs\n",
    "print(new_cols[:7])\n",
    "df.columns = new_cols\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "8c6f1038-dd77-49fa-aee7-67c7cf4a4402",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((377, 40), (768, 41), (25952, 380))"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta_full.set_index('CellID').loc[sampleIDs].shape, df_meta_full.shape, df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "18284e63-216c-4aee-9235-d37542003359",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GSM3432562    1\n",
       "GSM3432563    1\n",
       "GSM3433068    1\n",
       "GSM3433069    1\n",
       "GSM3433070    1\n",
       "             ..\n",
       "GSM3432820    1\n",
       "GSM3432821    1\n",
       "GSM3432822    1\n",
       "GSM3432823    1\n",
       "GSM3433329    1\n",
       "Name: Sample Name, Length: 768, dtype: int64"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta_full['Sample Name'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "da1c7e88-f23f-4462-b671-d5449321c2b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Age        Mouse_ID\n",
       "2 months   Y2          96\n",
       "           Y4          96\n",
       "           Y5          96\n",
       "           Y8          96\n",
       "24 months  O1          96\n",
       "           O5          96\n",
       "           O8          96\n",
       "2 months   Y7          93\n",
       "dtype: int64"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta.iloc[:,10:].head()\n",
    "df_meta[['Age', 'Mouse_ID']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c348142-2413-442b-8609-0cead18eb68c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "ba3d6bf2-a245-4e13-aafb-5e57d1598550",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>geneName</th>\n",
       "      <th>Xkr4</th>\n",
       "      <th>Gm1992</th>\n",
       "      <th>Rp1</th>\n",
       "      <th>Sox17</th>\n",
       "      <th>Mrpl15</th>\n",
       "      <th>Lypla1</th>\n",
       "      <th>Tcea1</th>\n",
       "      <th>Gm6104</th>\n",
       "      <th>Rgs20</th>\n",
       "      <th>Gm16041</th>\n",
       "      <th>...</th>\n",
       "      <th>AC211711.1</th>\n",
       "      <th>AC200894.1</th>\n",
       "      <th>AC192435.1</th>\n",
       "      <th>AC146633.1</th>\n",
       "      <th>AC183902.1</th>\n",
       "      <th>AC151712.3</th>\n",
       "      <th>AC151712.1</th>\n",
       "      <th>AC151712.4</th>\n",
       "      <th>AC151712.5</th>\n",
       "      <th>AC151712.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start</th>\n",
       "      <td>3205901</td>\n",
       "      <td>3466587</td>\n",
       "      <td>4343507</td>\n",
       "      <td>4490928</td>\n",
       "      <td>4773206</td>\n",
       "      <td>4807788</td>\n",
       "      <td>4857814</td>\n",
       "      <td>4879202</td>\n",
       "      <td>4909576</td>\n",
       "      <td>4970857</td>\n",
       "      <td>...</td>\n",
       "      <td>88804770</td>\n",
       "      <td>88952381</td>\n",
       "      <td>89328785</td>\n",
       "      <td>89846210</td>\n",
       "      <td>90402688</td>\n",
       "      <td>90754513</td>\n",
       "      <td>90755657</td>\n",
       "      <td>90784738</td>\n",
       "      <td>90838869</td>\n",
       "      <td>90843934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_11</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>69</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_12</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_13</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  25952 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "geneName        Xkr4   Gm1992      Rp1    Sox17   Mrpl15   Lypla1    Tcea1  \\\n",
       "chr                1        1        1        1        1        1        1   \n",
       "start        3205901  3466587  4343507  4490928  4773206  4807788  4857814   \n",
       "Y4_TA_Hi_11        0        0        0        0        0        0       69   \n",
       "Y4_TA_Hi_12        0        0        0        0        0        0        0   \n",
       "Y4_TA_Hi_13        0        0        0        0      121        0        0   \n",
       "\n",
       "geneName      Gm6104    Rgs20  Gm16041  ... AC211711.1 AC200894.1 AC192435.1  \\\n",
       "chr                1        1        1  ...          Y          Y          Y   \n",
       "start        4879202  4909576  4970857  ...   88804770   88952381   89328785   \n",
       "Y4_TA_Hi_11        0        0        0  ...          0          0          0   \n",
       "Y4_TA_Hi_12        0        0        0  ...          0          0          0   \n",
       "Y4_TA_Hi_13        0        0        0  ...          0          0          0   \n",
       "\n",
       "geneName    AC146633.1 AC183902.1 AC151712.3 AC151712.1 AC151712.4 AC151712.5  \\\n",
       "chr                  Y          Y          Y          Y          Y          Y   \n",
       "start         89846210   90402688   90754513   90755657   90784738   90838869   \n",
       "Y4_TA_Hi_11          0          0          0          0          0          0   \n",
       "Y4_TA_Hi_12          0          0          0          0          0          0   \n",
       "Y4_TA_Hi_13          0          0          0          0          0          0   \n",
       "\n",
       "geneName    AC151712.2  \n",
       "chr                  Y  \n",
       "start         90843934  \n",
       "Y4_TA_Hi_11          0  \n",
       "Y4_TA_Hi_12          0  \n",
       "Y4_TA_Hi_13          0  \n",
       "\n",
       "[5 rows x 25952 columns]"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_full = df.copy().set_index('geneName').T\n",
    "df_full.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "3ce9e199-6abc-4d23-bdfa-8297b60c82af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>Consent</th>\n",
       "      <th>DATASTORE filetype</th>\n",
       "      <th>...</th>\n",
       "      <th>Sample Type</th>\n",
       "      <th>Taxonomy</th>\n",
       "      <th>Channels</th>\n",
       "      <th>Platform_y</th>\n",
       "      <th>Series</th>\n",
       "      <th>Supplementary Types</th>\n",
       "      <th>Supplementary Links</th>\n",
       "      <th>SRA Accession</th>\n",
       "      <th>Contact</th>\n",
       "      <th>Release Date</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CellID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>O1_TA_Hi_10</th>\n",
       "      <td>SRR8066856</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>1169700</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10249431</td>\n",
       "      <td>494375</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895195</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O1_TA_Hi_11</th>\n",
       "      <td>SRR8066857</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>147409200</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10249430</td>\n",
       "      <td>56823661</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>sra,fastq</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895196</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O1_TA_Hi_12</th>\n",
       "      <td>SRR8066858</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>185247450</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10249429</td>\n",
       "      <td>66440460</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895197</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O1_TA_Hi_13</th>\n",
       "      <td>SRR8066859</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>187086300</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10249428</td>\n",
       "      <td>67360076</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895198</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O1_TA_Hi_14</th>\n",
       "      <td>SRR8066860</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>16032900</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10249427</td>\n",
       "      <td>5883082</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895199</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Run Assay Type  AvgSpotLen      Bases   BioProject  \\\n",
       "CellID                                                                   \n",
       "O1_TA_Hi_10  SRR8066856    RNA-Seq         150    1169700  PRJNA497128   \n",
       "O1_TA_Hi_11  SRR8066857    RNA-Seq         150  147409200  PRJNA497128   \n",
       "O1_TA_Hi_12  SRR8066858    RNA-Seq         150  185247450  PRJNA497128   \n",
       "O1_TA_Hi_13  SRR8066859    RNA-Seq         150  187086300  PRJNA497128   \n",
       "O1_TA_Hi_14  SRR8066860    RNA-Seq         150   16032900  PRJNA497128   \n",
       "\n",
       "                BioSample     Bytes Center Name Consent DATASTORE filetype  \\\n",
       "CellID                                                                       \n",
       "O1_TA_Hi_10  SAMN10249431    494375         GEO  public          fastq,sra   \n",
       "O1_TA_Hi_11  SAMN10249430  56823661         GEO  public          sra,fastq   \n",
       "O1_TA_Hi_12  SAMN10249429  66440460         GEO  public          fastq,sra   \n",
       "O1_TA_Hi_13  SAMN10249428  67360076         GEO  public          fastq,sra   \n",
       "O1_TA_Hi_14  SAMN10249427   5883082         GEO  public          fastq,sra   \n",
       "\n",
       "             ... Sample Type      Taxonomy Channels Platform_y  \\\n",
       "CellID       ...                                                 \n",
       "O1_TA_Hi_10  ...         SRA  Mus musculus        1   GPL17021   \n",
       "O1_TA_Hi_11  ...         SRA  Mus musculus        1   GPL17021   \n",
       "O1_TA_Hi_12  ...         SRA  Mus musculus        1   GPL17021   \n",
       "O1_TA_Hi_13  ...         SRA  Mus musculus        1   GPL17021   \n",
       "O1_TA_Hi_14  ...         SRA  Mus musculus        1   GPL17021   \n",
       "\n",
       "                          Series Supplementary Types  \\\n",
       "CellID                                                 \n",
       "O1_TA_Hi_10  GSE121364;GSE121437    SRA Run Selector   \n",
       "O1_TA_Hi_11  GSE121364;GSE121437    SRA Run Selector   \n",
       "O1_TA_Hi_12  GSE121364;GSE121437    SRA Run Selector   \n",
       "O1_TA_Hi_13  GSE121364;GSE121437    SRA Run Selector   \n",
       "O1_TA_Hi_14  GSE121364;GSE121437    SRA Run Selector   \n",
       "\n",
       "                                           Supplementary Links SRA Accession  \\\n",
       "CellID                                                                         \n",
       "O1_TA_Hi_10  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895195   \n",
       "O1_TA_Hi_11  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895196   \n",
       "O1_TA_Hi_12  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895197   \n",
       "O1_TA_Hi_13  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895198   \n",
       "O1_TA_Hi_14  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895199   \n",
       "\n",
       "                   Contact  Release Date  \n",
       "CellID                                    \n",
       "O1_TA_Hi_10  Felix Krueger  Aug 14, 2019  \n",
       "O1_TA_Hi_11  Felix Krueger  Aug 14, 2019  \n",
       "O1_TA_Hi_12  Felix Krueger  Aug 14, 2019  \n",
       "O1_TA_Hi_13  Felix Krueger  Aug 14, 2019  \n",
       "O1_TA_Hi_14  Felix Krueger  Aug 14, 2019  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 163,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = df_meta_full.copy()\n",
    "df_meta.set_index('CellID', inplace=True)\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "06d69f1d-6131-49c7-9710-79d6b684f001",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/Dropbox/Harvard/Science/scDNAm/data/GSE121364/GSE121364.h5ad\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "data_path = Path.joinpath(cwd, project_dir, \"{}.h5ad\".format(project_id))\n",
    "print(data_path)\n",
    "ad = sc.AnnData(df_full.iloc[2:,])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "f2b939e3-d4f2-4395-be3f-f9e2c0c4767f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm1992</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rp1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sox17</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mrpl15</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [Xkr4, Gm1992, Rp1, Sox17, Mrpl15]"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "2a92ccd6-0ed4-4e9e-9441-1156fde1336b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "      <td>1</td>\n",
       "      <td>3205901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm1992</th>\n",
       "      <td>1</td>\n",
       "      <td>3466587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rp1</th>\n",
       "      <td>1</td>\n",
       "      <td>4343507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sox17</th>\n",
       "      <td>1</td>\n",
       "      <td>4490928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mrpl15</th>\n",
       "      <td>1</td>\n",
       "      <td>4773206</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr    start\n",
       "geneName             \n",
       "Xkr4       1  3205901\n",
       "Gm1992     1  3466587\n",
       "Rp1        1  4343507\n",
       "Sox17      1  4490928\n",
       "Mrpl15     1  4773206"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad.var['chr'] = df_full.iloc[0,:].values\n",
    "ad.var['start'] = df_full.iloc[1,:].values\n",
    "ad.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "819b741b-eea1-4375-9401-d46c50b0d0e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_11</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_12</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_13</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_23</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [Y4_TA_Hi_11, Y4_TA_Hi_12, Y4_TA_Hi_13, Y4_TA_Hi_1, Y4_TA_Hi_23]"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "3e3631a2-ed3d-4c2f-84f4-e5d7e867f120",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(377, 5)"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.decomposition import PCA\n",
    "\n",
    "pca = PCA(n_components=5)\n",
    "z = pca.fit_transform(ad.X)\n",
    "z.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "00008933-9c6b-42c4-82ec-243cad38ebd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       ...,\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.],\n",
       "       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)"
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "daa9764e-92a3-4a8f-84e0-3360b5eb8464",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plt.scatter(z[:,0], z[:,1], c=ad.obs.Age.values == '2 months')\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "b48e2907-08e4-4d41-b3be-6b9da27b8d0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>Consent</th>\n",
       "      <th>DATASTORE filetype</th>\n",
       "      <th>...</th>\n",
       "      <th>Sample Type</th>\n",
       "      <th>Taxonomy</th>\n",
       "      <th>Channels</th>\n",
       "      <th>Platform_y</th>\n",
       "      <th>Series</th>\n",
       "      <th>Supplementary Types</th>\n",
       "      <th>Supplementary Links</th>\n",
       "      <th>SRA Accession</th>\n",
       "      <th>Contact</th>\n",
       "      <th>Release Date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_11</th>\n",
       "      <td>SRR8067241</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>107484600</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10250014</td>\n",
       "      <td>41604327</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>sra,fastq</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895580</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_12</th>\n",
       "      <td>SRR8067242</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>90305850</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10250013</td>\n",
       "      <td>35342837</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>sra,fastq</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895581</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_13</th>\n",
       "      <td>SRR8067243</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>97244550</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10250012</td>\n",
       "      <td>37644383</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895582</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_1</th>\n",
       "      <td>SRR8067250</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>39392850</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10250005</td>\n",
       "      <td>15256244</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895589</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_23</th>\n",
       "      <td>SRR8067254</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>116119050</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10250001</td>\n",
       "      <td>45741310</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>sra,fastq</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895593</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Run Assay Type  AvgSpotLen      Bases   BioProject  \\\n",
       "Y4_TA_Hi_11  SRR8067241    RNA-Seq         150  107484600  PRJNA497128   \n",
       "Y4_TA_Hi_12  SRR8067242    RNA-Seq         150   90305850  PRJNA497128   \n",
       "Y4_TA_Hi_13  SRR8067243    RNA-Seq         150   97244550  PRJNA497128   \n",
       "Y4_TA_Hi_1   SRR8067250    RNA-Seq         150   39392850  PRJNA497128   \n",
       "Y4_TA_Hi_23  SRR8067254    RNA-Seq         150  116119050  PRJNA497128   \n",
       "\n",
       "                BioSample     Bytes Center Name Consent DATASTORE filetype  \\\n",
       "Y4_TA_Hi_11  SAMN10250014  41604327         GEO  public          sra,fastq   \n",
       "Y4_TA_Hi_12  SAMN10250013  35342837         GEO  public          sra,fastq   \n",
       "Y4_TA_Hi_13  SAMN10250012  37644383         GEO  public          fastq,sra   \n",
       "Y4_TA_Hi_1   SAMN10250005  15256244         GEO  public          fastq,sra   \n",
       "Y4_TA_Hi_23  SAMN10250001  45741310         GEO  public          sra,fastq   \n",
       "\n",
       "             ... Sample Type      Taxonomy Channels Platform_y  \\\n",
       "Y4_TA_Hi_11  ...         SRA  Mus musculus        1   GPL17021   \n",
       "Y4_TA_Hi_12  ...         SRA  Mus musculus        1   GPL17021   \n",
       "Y4_TA_Hi_13  ...         SRA  Mus musculus        1   GPL17021   \n",
       "Y4_TA_Hi_1   ...         SRA  Mus musculus        1   GPL17021   \n",
       "Y4_TA_Hi_23  ...         SRA  Mus musculus        1   GPL17021   \n",
       "\n",
       "                          Series Supplementary Types  \\\n",
       "Y4_TA_Hi_11  GSE121364;GSE121437    SRA Run Selector   \n",
       "Y4_TA_Hi_12  GSE121364;GSE121437    SRA Run Selector   \n",
       "Y4_TA_Hi_13  GSE121364;GSE121437    SRA Run Selector   \n",
       "Y4_TA_Hi_1   GSE121364;GSE121437    SRA Run Selector   \n",
       "Y4_TA_Hi_23  GSE121364;GSE121437    SRA Run Selector   \n",
       "\n",
       "                                           Supplementary Links SRA Accession  \\\n",
       "Y4_TA_Hi_11  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895580   \n",
       "Y4_TA_Hi_12  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895581   \n",
       "Y4_TA_Hi_13  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895582   \n",
       "Y4_TA_Hi_1   https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895589   \n",
       "Y4_TA_Hi_23  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895593   \n",
       "\n",
       "                   Contact  Release Date  \n",
       "Y4_TA_Hi_11  Felix Krueger  Aug 14, 2019  \n",
       "Y4_TA_Hi_12  Felix Krueger  Aug 14, 2019  \n",
       "Y4_TA_Hi_13  Felix Krueger  Aug 14, 2019  \n",
       "Y4_TA_Hi_1   Felix Krueger  Aug 14, 2019  \n",
       "Y4_TA_Hi_23  Felix Krueger  Aug 14, 2019  \n",
       "\n",
       "[5 rows x 40 columns]"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad.obs = pd.concat((ad.obs, df_meta.loc[ad.obs.index,:]), axis=1)\n",
    "ad.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "7c5dc616-7dcd-4a07-95c4-c70c8d96bfc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in ['AvgSpotLen', 'Bases', 'Bytes', 'Channels']:\n",
    "    ad.obs[col] = ad.obs[col].astype(str)\n",
    "\n",
    "for col in ['chr', 'start']:\n",
    "    ad.var[col] = ad.var[col].astype(str)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "fcea39cc-9307-4c4c-9bb1-09ffb1b5e2d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Assay Type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'AvgSpotLen' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'BioProject' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Center Name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Consent' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE filetype' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE provider' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE region' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Instrument' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibraryLayout' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibrarySelection' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibrarySource' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Organism' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Platform_x' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'ReleaseDate' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'SRA Study' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Age' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Cell_type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Mouse_ID' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'source_name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'STRAIN' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Sample Type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Taxonomy' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Channels' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Platform_y' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Series' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Supplementary Types' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Contact' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Release Date' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'chr' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'start' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 377 entries, Y4_TA_Hi_11 to O5_TA_Hi_95\n",
      "Data columns (total 40 columns):\n",
      " #   Column               Non-Null Count  Dtype   \n",
      "---  ------               --------------  -----   \n",
      " 0   Run                  377 non-null    string  \n",
      " 1   Assay Type           377 non-null    category\n",
      " 2   AvgSpotLen           377 non-null    category\n",
      " 3   Bases                377 non-null    string  \n",
      " 4   BioProject           377 non-null    category\n",
      " 5   BioSample            377 non-null    string  \n",
      " 6   Bytes                377 non-null    string  \n",
      " 7   Center Name          377 non-null    category\n",
      " 8   Consent              377 non-null    category\n",
      " 9   DATASTORE filetype   377 non-null    category\n",
      " 10  DATASTORE provider   377 non-null    category\n",
      " 11  DATASTORE region     377 non-null    category\n",
      " 12  Experiment           377 non-null    string  \n",
      " 13  GEO_Accession (exp)  377 non-null    string  \n",
      " 14  Instrument           377 non-null    category\n",
      " 15  LibraryLayout        377 non-null    category\n",
      " 16  LibrarySelection     377 non-null    category\n",
      " 17  LibrarySource        377 non-null    category\n",
      " 18  Organism             377 non-null    category\n",
      " 19  Platform_x           377 non-null    category\n",
      " 20  ReleaseDate          377 non-null    category\n",
      " 21  Sample Name          377 non-null    string  \n",
      " 22  SRA Study            377 non-null    category\n",
      " 23  Age                  374 non-null    category\n",
      " 24  Cell_type            374 non-null    category\n",
      " 25  Mouse_ID             374 non-null    category\n",
      " 26  source_name          374 non-null    category\n",
      " 27  STRAIN               374 non-null    category\n",
      " 28  Accession            377 non-null    string  \n",
      " 29  Title                377 non-null    string  \n",
      " 30  Sample Type          377 non-null    category\n",
      " 31  Taxonomy             377 non-null    category\n",
      " 32  Channels             377 non-null    category\n",
      " 33  Platform_y           377 non-null    category\n",
      " 34  Series               377 non-null    category\n",
      " 35  Supplementary Types  377 non-null    category\n",
      " 36  Supplementary Links  377 non-null    string  \n",
      " 37  SRA Accession        377 non-null    string  \n",
      " 38  Contact              377 non-null    category\n",
      " 39  Release Date         377 non-null    category\n",
      "dtypes: category(29), string(11)\n",
      "memory usage: 57.5+ KB\n"
     ]
    }
   ],
   "source": [
    "ad.obs = ad.obs.convert_dtypes()\n",
    "# ad.obs = ad.obs.drop(['AvgSpotLen', 'Bases', 'Bytes', 'Channels'], axis=1)\n",
    "# ad.var = ad.var.drop(['chr', 'start'], axis=1)\n",
    "ad.var = ad.var.convert_dtypes()\n",
    "ad._sanitize()\n",
    "ad.obs.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "8aae1442-b3bd-43f8-bc38-8e6c4b0fb437",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ad.obs_names_make_unique()\n",
    "ad.var_names_make_unique()\n",
    "ad.write_h5ad(data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "id": "301475ad-a5c5-4d30-a87c-094a808c04ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad.to_df().to_csv('../scDNAm/data/GSE121364//For_Orr/X_table.csv')\n",
    "ad.var.to_csv('../scDNAm/data/GSE121364//For_Orr/Variables_table.csv')\n",
    "ad.obs.to_csv('../scDNAm/data/GSE121364/For_Orr/Observations_table.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "id": "50546081-fdbe-48e3-979d-fb1253da2464",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>Consent</th>\n",
       "      <th>DATASTORE filetype</th>\n",
       "      <th>...</th>\n",
       "      <th>Sample Type</th>\n",
       "      <th>Taxonomy</th>\n",
       "      <th>Channels</th>\n",
       "      <th>Platform_y</th>\n",
       "      <th>Series</th>\n",
       "      <th>Supplementary Types</th>\n",
       "      <th>Supplementary Links</th>\n",
       "      <th>SRA Accession</th>\n",
       "      <th>Contact</th>\n",
       "      <th>Release Date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_11</th>\n",
       "      <td>SRR8067241</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>107484600</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10250014</td>\n",
       "      <td>41604327</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>sra,fastq</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895580</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_12</th>\n",
       "      <td>SRR8067242</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>90305850</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10250013</td>\n",
       "      <td>35342837</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>sra,fastq</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895581</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_13</th>\n",
       "      <td>SRR8067243</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>97244550</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10250012</td>\n",
       "      <td>37644383</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895582</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_1</th>\n",
       "      <td>SRR8067250</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>39392850</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10250005</td>\n",
       "      <td>15256244</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895589</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y4_TA_Hi_23</th>\n",
       "      <td>SRR8067254</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>116119050</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10250001</td>\n",
       "      <td>45741310</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>sra,fastq</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895593</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O5_TA_Hi_90</th>\n",
       "      <td>SRR8067040</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>105824700</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10249402</td>\n",
       "      <td>38171506</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895379</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O5_TA_Hi_91</th>\n",
       "      <td>SRR8067041</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>107854050</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10249401</td>\n",
       "      <td>38724795</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895380</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O5_TA_Hi_92</th>\n",
       "      <td>SRR8067042</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>64730850</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10249400</td>\n",
       "      <td>22913257</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895381</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O5_TA_Hi_94</th>\n",
       "      <td>SRR8067044</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>158993700</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10249398</td>\n",
       "      <td>58439965</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>sra,fastq</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895383</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O5_TA_Hi_95</th>\n",
       "      <td>SRR8067045</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>168519000</td>\n",
       "      <td>PRJNA497128</td>\n",
       "      <td>SAMN10249397</td>\n",
       "      <td>61314359</td>\n",
       "      <td>GEO</td>\n",
       "      <td>public</td>\n",
       "      <td>fastq,sra</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121364;GSE121437</td>\n",
       "      <td>SRA Run Selector</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/Traces/study/?acc...</td>\n",
       "      <td>SRX4895384</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 14, 2019</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>377 rows  40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Run Assay Type AvgSpotLen      Bases   BioProject  \\\n",
       "Y4_TA_Hi_11  SRR8067241    RNA-Seq        150  107484600  PRJNA497128   \n",
       "Y4_TA_Hi_12  SRR8067242    RNA-Seq        150   90305850  PRJNA497128   \n",
       "Y4_TA_Hi_13  SRR8067243    RNA-Seq        150   97244550  PRJNA497128   \n",
       "Y4_TA_Hi_1   SRR8067250    RNA-Seq        150   39392850  PRJNA497128   \n",
       "Y4_TA_Hi_23  SRR8067254    RNA-Seq        150  116119050  PRJNA497128   \n",
       "...                 ...        ...        ...        ...          ...   \n",
       "O5_TA_Hi_90  SRR8067040    RNA-Seq        150  105824700  PRJNA497128   \n",
       "O5_TA_Hi_91  SRR8067041    RNA-Seq        150  107854050  PRJNA497128   \n",
       "O5_TA_Hi_92  SRR8067042    RNA-Seq        150   64730850  PRJNA497128   \n",
       "O5_TA_Hi_94  SRR8067044    RNA-Seq        150  158993700  PRJNA497128   \n",
       "O5_TA_Hi_95  SRR8067045    RNA-Seq        150  168519000  PRJNA497128   \n",
       "\n",
       "                BioSample     Bytes Center Name Consent DATASTORE filetype  \\\n",
       "Y4_TA_Hi_11  SAMN10250014  41604327         GEO  public          sra,fastq   \n",
       "Y4_TA_Hi_12  SAMN10250013  35342837         GEO  public          sra,fastq   \n",
       "Y4_TA_Hi_13  SAMN10250012  37644383         GEO  public          fastq,sra   \n",
       "Y4_TA_Hi_1   SAMN10250005  15256244         GEO  public          fastq,sra   \n",
       "Y4_TA_Hi_23  SAMN10250001  45741310         GEO  public          sra,fastq   \n",
       "...                   ...       ...         ...     ...                ...   \n",
       "O5_TA_Hi_90  SAMN10249402  38171506         GEO  public          fastq,sra   \n",
       "O5_TA_Hi_91  SAMN10249401  38724795         GEO  public          fastq,sra   \n",
       "O5_TA_Hi_92  SAMN10249400  22913257         GEO  public          fastq,sra   \n",
       "O5_TA_Hi_94  SAMN10249398  58439965         GEO  public          sra,fastq   \n",
       "O5_TA_Hi_95  SAMN10249397  61314359         GEO  public          fastq,sra   \n",
       "\n",
       "             ... Sample Type      Taxonomy Channels Platform_y  \\\n",
       "Y4_TA_Hi_11  ...         SRA  Mus musculus        1   GPL17021   \n",
       "Y4_TA_Hi_12  ...         SRA  Mus musculus        1   GPL17021   \n",
       "Y4_TA_Hi_13  ...         SRA  Mus musculus        1   GPL17021   \n",
       "Y4_TA_Hi_1   ...         SRA  Mus musculus        1   GPL17021   \n",
       "Y4_TA_Hi_23  ...         SRA  Mus musculus        1   GPL17021   \n",
       "...          ...         ...           ...      ...        ...   \n",
       "O5_TA_Hi_90  ...         SRA  Mus musculus        1   GPL17021   \n",
       "O5_TA_Hi_91  ...         SRA  Mus musculus        1   GPL17021   \n",
       "O5_TA_Hi_92  ...         SRA  Mus musculus        1   GPL17021   \n",
       "O5_TA_Hi_94  ...         SRA  Mus musculus        1   GPL17021   \n",
       "O5_TA_Hi_95  ...         SRA  Mus musculus        1   GPL17021   \n",
       "\n",
       "                          Series Supplementary Types  \\\n",
       "Y4_TA_Hi_11  GSE121364;GSE121437    SRA Run Selector   \n",
       "Y4_TA_Hi_12  GSE121364;GSE121437    SRA Run Selector   \n",
       "Y4_TA_Hi_13  GSE121364;GSE121437    SRA Run Selector   \n",
       "Y4_TA_Hi_1   GSE121364;GSE121437    SRA Run Selector   \n",
       "Y4_TA_Hi_23  GSE121364;GSE121437    SRA Run Selector   \n",
       "...                          ...                 ...   \n",
       "O5_TA_Hi_90  GSE121364;GSE121437    SRA Run Selector   \n",
       "O5_TA_Hi_91  GSE121364;GSE121437    SRA Run Selector   \n",
       "O5_TA_Hi_92  GSE121364;GSE121437    SRA Run Selector   \n",
       "O5_TA_Hi_94  GSE121364;GSE121437    SRA Run Selector   \n",
       "O5_TA_Hi_95  GSE121364;GSE121437    SRA Run Selector   \n",
       "\n",
       "                                           Supplementary Links SRA Accession  \\\n",
       "Y4_TA_Hi_11  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895580   \n",
       "Y4_TA_Hi_12  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895581   \n",
       "Y4_TA_Hi_13  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895582   \n",
       "Y4_TA_Hi_1   https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895589   \n",
       "Y4_TA_Hi_23  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895593   \n",
       "...                                                        ...           ...   \n",
       "O5_TA_Hi_90  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895379   \n",
       "O5_TA_Hi_91  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895380   \n",
       "O5_TA_Hi_92  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895381   \n",
       "O5_TA_Hi_94  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895383   \n",
       "O5_TA_Hi_95  https://www.ncbi.nlm.nih.gov/Traces/study/?acc...    SRX4895384   \n",
       "\n",
       "                   Contact  Release Date  \n",
       "Y4_TA_Hi_11  Felix Krueger  Aug 14, 2019  \n",
       "Y4_TA_Hi_12  Felix Krueger  Aug 14, 2019  \n",
       "Y4_TA_Hi_13  Felix Krueger  Aug 14, 2019  \n",
       "Y4_TA_Hi_1   Felix Krueger  Aug 14, 2019  \n",
       "Y4_TA_Hi_23  Felix Krueger  Aug 14, 2019  \n",
       "...                    ...           ...  \n",
       "O5_TA_Hi_90  Felix Krueger  Aug 14, 2019  \n",
       "O5_TA_Hi_91  Felix Krueger  Aug 14, 2019  \n",
       "O5_TA_Hi_92  Felix Krueger  Aug 14, 2019  \n",
       "O5_TA_Hi_94  Felix Krueger  Aug 14, 2019  \n",
       "O5_TA_Hi_95  Felix Krueger  Aug 14, 2019  \n",
       "\n",
       "[377 rows x 40 columns]"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39198869-91d4-4258-855a-cf5bdb2e90b3",
   "metadata": {},
   "source": [
    "# GSE121436 scDNAm "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "77ac21d8-8490-4a01-81bc-a1c846782210",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwd = Path.cwd()\n",
    "# cwd = Path('/Users/tarkhov/Dropbox/Harvard/Science/Church/')\n",
    "project_dir = Path.joinpath(cwd, Path('data/GSE121436/'))\n",
    "project_id = 'GSE121436'\n",
    "project_samples_dir = 'GSE121436_RAW'\n",
    "metainfo_path = 'SraRunTable.txt'\n",
    "metainfo_path_2 = 'sample.csv'\n",
    "gl_metainfo_path = Path.joinpath(project_dir, metainfo_path)\n",
    "gl_metainfo_path_2 = Path.joinpath(project_dir, metainfo_path_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c1e11593-7c45-4448-97db-19f3ac699096",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(275, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Title</th>\n",
       "      <th>Sample Type</th>\n",
       "      <th>Taxonomy</th>\n",
       "      <th>Channels</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Series</th>\n",
       "      <th>Supplementary Types</th>\n",
       "      <th>Supplementary Links</th>\n",
       "      <th>SRA Accession</th>\n",
       "      <th>Contact</th>\n",
       "      <th>Release Date</th>\n",
       "      <th>CellID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSM3436261</td>\n",
       "      <td>O1_TA_Hi_10 [scBS-seq]</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121436;GSE121437</td>\n",
       "      <td>COV;SRA Run Selector</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...</td>\n",
       "      <td>SRX4901189</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 22, 2019</td>\n",
       "      <td>O1_TA_Hi_10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSM3436262</td>\n",
       "      <td>O1_TA_Hi_11 [scBS-seq]</td>\n",
       "      <td>SRA</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>1</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121436;GSE121437</td>\n",
       "      <td>COV;SRA Run Selector</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...</td>\n",
       "      <td>SRX4901190</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 22, 2019</td>\n",
       "      <td>O1_TA_Hi_11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Accession                   Title Sample Type      Taxonomy  Channels  \\\n",
       "0  GSM3436261  O1_TA_Hi_10 [scBS-seq]         SRA  Mus musculus         1   \n",
       "1  GSM3436262  O1_TA_Hi_11 [scBS-seq]         SRA  Mus musculus         1   \n",
       "\n",
       "   Platform               Series   Supplementary Types  \\\n",
       "0  GPL17021  GSE121436;GSE121437  COV;SRA Run Selector   \n",
       "1  GPL17021  GSE121436;GSE121437  COV;SRA Run Selector   \n",
       "\n",
       "                                 Supplementary Links SRA Accession  \\\n",
       "0  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...    SRX4901189   \n",
       "1  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...    SRX4901190   \n",
       "\n",
       "         Contact  Release Date       CellID  \n",
       "0  Felix Krueger  Aug 22, 2019  O1_TA_Hi_10  \n",
       "1  Felix Krueger  Aug 22, 2019  O1_TA_Hi_11  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta_2 = pd.read_csv(gl_metainfo_path_2, sep=',')\n",
    "df_meta_2['CellID'] = df_meta_2['Title'].apply(lambda x: x.split(' ')[0])\n",
    "print(df_meta_2.shape)\n",
    "df_meta_2.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8a298f40-3ea5-42ab-badf-7ba1fed6d8ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['O1', 'O8'], dtype=object)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_csv(gl_metainfo_path, sep=',')\n",
    "# df_meta[df_meta['Age'] == '2 months']['Mouse_ID'].unique()\n",
    "df_meta[df_meta['Age'] == '24 months']['Mouse_ID'].unique()#head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "7aee8181-43f9-41e0-b59e-abbd456fe16d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(275, 28)\n",
      "(275, 41)\n",
      "(275, 42)\n",
      "(550, 43)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Age</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Cell_type</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>...</th>\n",
       "      <th>Platform_y</th>\n",
       "      <th>Series</th>\n",
       "      <th>Supplementary Types</th>\n",
       "      <th>Supplementary Links</th>\n",
       "      <th>SRA Accession</th>\n",
       "      <th>Contact</th>\n",
       "      <th>Release Date</th>\n",
       "      <th>CellID</th>\n",
       "      <th>Age_months</th>\n",
       "      <th>Data Type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name PerCov</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM3436261_percentage</th>\n",
       "      <td>SRR8073317</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>250</td>\n",
       "      <td>891398250</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254811</td>\n",
       "      <td>351440127</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121436;GSE121437</td>\n",
       "      <td>COV;SRA Run Selector</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...</td>\n",
       "      <td>SRX4901189</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 22, 2019</td>\n",
       "      <td>O1_TA_Hi_10</td>\n",
       "      <td>24 months</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3436262_percentage</th>\n",
       "      <td>SRR8073318</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>250</td>\n",
       "      <td>775564750</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254810</td>\n",
       "      <td>307776943</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121436;GSE121437</td>\n",
       "      <td>COV;SRA Run Selector</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...</td>\n",
       "      <td>SRX4901190</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 22, 2019</td>\n",
       "      <td>O1_TA_Hi_11</td>\n",
       "      <td>24 months</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3436263_percentage</th>\n",
       "      <td>SRR8073319</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>250</td>\n",
       "      <td>844106750</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254809</td>\n",
       "      <td>335133628</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121436;GSE121437</td>\n",
       "      <td>COV;SRA Run Selector</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...</td>\n",
       "      <td>SRX4901191</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 22, 2019</td>\n",
       "      <td>O1_TA_Hi_12</td>\n",
       "      <td>24 months</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3436264_percentage</th>\n",
       "      <td>SRR8073320</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>250</td>\n",
       "      <td>947647750</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254808</td>\n",
       "      <td>376647456</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121436;GSE121437</td>\n",
       "      <td>COV;SRA Run Selector</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...</td>\n",
       "      <td>SRX4901192</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 22, 2019</td>\n",
       "      <td>O1_TA_Hi_13</td>\n",
       "      <td>24 months</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3436265_percentage</th>\n",
       "      <td>SRR8073321</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>250</td>\n",
       "      <td>917775500</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254807</td>\n",
       "      <td>368154088</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121436;GSE121437</td>\n",
       "      <td>COV;SRA Run Selector</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...</td>\n",
       "      <td>SRX4901193</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 22, 2019</td>\n",
       "      <td>O1_TA_Hi_14</td>\n",
       "      <td>24 months</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Run   Age     Assay Type  AvgSpotLen      Bases  \\\n",
       "Name PerCov                                                                     \n",
       "GSM3436261_percentage  SRR8073317  24.0  Bisulfite-Seq         250  891398250   \n",
       "GSM3436262_percentage  SRR8073318  24.0  Bisulfite-Seq         250  775564750   \n",
       "GSM3436263_percentage  SRR8073319  24.0  Bisulfite-Seq         250  844106750   \n",
       "GSM3436264_percentage  SRR8073320  24.0  Bisulfite-Seq         250  947647750   \n",
       "GSM3436265_percentage  SRR8073321  24.0  Bisulfite-Seq         250  917775500   \n",
       "\n",
       "                        BioProject     BioSample      Bytes  \\\n",
       "Name PerCov                                                   \n",
       "GSM3436261_percentage  PRJNA497312  SAMN10254811  351440127   \n",
       "GSM3436262_percentage  PRJNA497312  SAMN10254810  307776943   \n",
       "GSM3436263_percentage  PRJNA497312  SAMN10254809  335133628   \n",
       "GSM3436264_percentage  PRJNA497312  SAMN10254808  376647456   \n",
       "GSM3436265_percentage  PRJNA497312  SAMN10254807  368154088   \n",
       "\n",
       "                               Cell_type Center Name  ... Platform_y  \\\n",
       "Name PerCov                                           ...              \n",
       "GSM3436261_percentage  Muscle stem cells         GEO  ...   GPL17021   \n",
       "GSM3436262_percentage  Muscle stem cells         GEO  ...   GPL17021   \n",
       "GSM3436263_percentage  Muscle stem cells         GEO  ...   GPL17021   \n",
       "GSM3436264_percentage  Muscle stem cells         GEO  ...   GPL17021   \n",
       "GSM3436265_percentage  Muscle stem cells         GEO  ...   GPL17021   \n",
       "\n",
       "                                    Series   Supplementary Types  \\\n",
       "Name PerCov                                                        \n",
       "GSM3436261_percentage  GSE121436;GSE121437  COV;SRA Run Selector   \n",
       "GSM3436262_percentage  GSE121436;GSE121437  COV;SRA Run Selector   \n",
       "GSM3436263_percentage  GSE121436;GSE121437  COV;SRA Run Selector   \n",
       "GSM3436264_percentage  GSE121436;GSE121437  COV;SRA Run Selector   \n",
       "GSM3436265_percentage  GSE121436;GSE121437  COV;SRA Run Selector   \n",
       "\n",
       "                                                     Supplementary Links  \\\n",
       "Name PerCov                                                                \n",
       "GSM3436261_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...   \n",
       "GSM3436262_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...   \n",
       "GSM3436263_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...   \n",
       "GSM3436264_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...   \n",
       "GSM3436265_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...   \n",
       "\n",
       "                      SRA Accession        Contact  Release Date       CellID  \\\n",
       "Name PerCov                                                                     \n",
       "GSM3436261_percentage    SRX4901189  Felix Krueger  Aug 22, 2019  O1_TA_Hi_10   \n",
       "GSM3436262_percentage    SRX4901190  Felix Krueger  Aug 22, 2019  O1_TA_Hi_11   \n",
       "GSM3436263_percentage    SRX4901191  Felix Krueger  Aug 22, 2019  O1_TA_Hi_12   \n",
       "GSM3436264_percentage    SRX4901192  Felix Krueger  Aug 22, 2019  O1_TA_Hi_13   \n",
       "GSM3436265_percentage    SRX4901193  Felix Krueger  Aug 22, 2019  O1_TA_Hi_14   \n",
       "\n",
       "                      Age_months   Data Type  \n",
       "Name PerCov                                   \n",
       "GSM3436261_percentage  24 months  percentage  \n",
       "GSM3436262_percentage  24 months  percentage  \n",
       "GSM3436263_percentage  24 months  percentage  \n",
       "GSM3436264_percentage  24 months  percentage  \n",
       "GSM3436265_percentage  24 months  percentage  \n",
       "\n",
       "[5 rows x 43 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta = pd.read_csv(gl_metainfo_path, sep=',')\n",
    "print(df_meta.shape)\n",
    "df_meta = pd.merge(df_meta, df_meta_2, left_on='Sample Name', right_on='Accession')\n",
    "print(df_meta.shape)\n",
    "df_meta['Age_months'] = df_meta['Age'].copy()\n",
    "df_meta['Age'] = df_meta['Age_months'].apply(lambda x: float(x.split(' ')[0]))\n",
    "print(df_meta.shape)\n",
    "df_meta_2 = df_meta.copy()\n",
    "df_meta['Name PerCov'] = df_meta['Sample Name'].str.cat(['percentage'] * df_meta.shape[0], sep='_')\n",
    "df_meta['Data Type'] = 'percentage'\n",
    "df_meta_2['Name PerCov'] = df_meta_2['Sample Name'].str.cat(['coverage'] * df_meta.shape[0], sep='_')\n",
    "df_meta_2['Data Type'] = 'coverage'\n",
    "df_meta['Name PerCov'] = df_meta['Sample Name'].str.cat(['percentage'] * df_meta.shape[0], sep='_')\n",
    "df_meta = pd.concat((df_meta, df_meta_2), axis=0)\n",
    "df_meta.set_index('Name PerCov', inplace=True)\n",
    "print(df_meta.shape)\n",
    "df_meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8639c3d8-9820-4e26-a375-0d8fd1b749d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# np.intersect1d(df_meta_full.CellID.values, df_meta_2.CellID.values).shape, \n",
    "# np.intersect1d(sampleIDs, df_meta_2.CellID.values).shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "12858817-3335-4215-87fa-1abd002cfd98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Age</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Cell_type</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>...</th>\n",
       "      <th>Platform_y</th>\n",
       "      <th>Series</th>\n",
       "      <th>Supplementary Types</th>\n",
       "      <th>Supplementary Links</th>\n",
       "      <th>SRA Accession</th>\n",
       "      <th>Contact</th>\n",
       "      <th>Release Date</th>\n",
       "      <th>CellID</th>\n",
       "      <th>Age_months</th>\n",
       "      <th>Data Type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name PerCov</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM3436480_percentage</th>\n",
       "      <td>SRR8073536</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>933869412</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10255048</td>\n",
       "      <td>440589845</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121436;GSE121437</td>\n",
       "      <td>COV;SRA Run Selector</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...</td>\n",
       "      <td>SRX4901408</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 22, 2019</td>\n",
       "      <td>Y8_TA_Hi_47</td>\n",
       "      <td>2 months</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3436480_coverage</th>\n",
       "      <td>SRR8073536</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>933869412</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10255048</td>\n",
       "      <td>440589845</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>GPL17021</td>\n",
       "      <td>GSE121436;GSE121437</td>\n",
       "      <td>COV;SRA Run Selector</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...</td>\n",
       "      <td>SRX4901408</td>\n",
       "      <td>Felix Krueger</td>\n",
       "      <td>Aug 22, 2019</td>\n",
       "      <td>Y8_TA_Hi_47</td>\n",
       "      <td>2 months</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows  43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Run  Age     Assay Type  AvgSpotLen      Bases  \\\n",
       "Name PerCov                                                                    \n",
       "GSM3436480_percentage  SRR8073536  2.0  Bisulfite-Seq         252  933869412   \n",
       "GSM3436480_coverage    SRR8073536  2.0  Bisulfite-Seq         252  933869412   \n",
       "\n",
       "                        BioProject     BioSample      Bytes  \\\n",
       "Name PerCov                                                   \n",
       "GSM3436480_percentage  PRJNA497312  SAMN10255048  440589845   \n",
       "GSM3436480_coverage    PRJNA497312  SAMN10255048  440589845   \n",
       "\n",
       "                               Cell_type Center Name  ... Platform_y  \\\n",
       "Name PerCov                                           ...              \n",
       "GSM3436480_percentage  Muscle stem cells         GEO  ...   GPL17021   \n",
       "GSM3436480_coverage    Muscle stem cells         GEO  ...   GPL17021   \n",
       "\n",
       "                                    Series   Supplementary Types  \\\n",
       "Name PerCov                                                        \n",
       "GSM3436480_percentage  GSE121436;GSE121437  COV;SRA Run Selector   \n",
       "GSM3436480_coverage    GSE121436;GSE121437  COV;SRA Run Selector   \n",
       "\n",
       "                                                     Supplementary Links  \\\n",
       "Name PerCov                                                                \n",
       "GSM3436480_percentage  ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...   \n",
       "GSM3436480_coverage    ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3436...   \n",
       "\n",
       "                      SRA Accession        Contact  Release Date       CellID  \\\n",
       "Name PerCov                                                                     \n",
       "GSM3436480_percentage    SRX4901408  Felix Krueger  Aug 22, 2019  Y8_TA_Hi_47   \n",
       "GSM3436480_coverage      SRX4901408  Felix Krueger  Aug 22, 2019  Y8_TA_Hi_47   \n",
       "\n",
       "                      Age_months   Data Type  \n",
       "Name PerCov                                   \n",
       "GSM3436480_percentage   2 months  percentage  \n",
       "GSM3436480_coverage     2 months    coverage  \n",
       "\n",
       "[2 rows x 43 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta[df_meta['Sample Name'] == 'GSM3436480']#.iloc[275:280,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f0110f13-aaf3-4f54-b50e-a5c0b46d3c9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/Dropbox/Harvard/Science/scDNAm/data/GSE121436/GSE121436_coverage_0_binned_by_1.h5ad\n"
     ]
    }
   ],
   "source": [
    "data_path = Path.joinpath(cwd, project_dir, \"{}_coverage_0_binned_by_1.h5ad\".format(project_id))\n",
    "print(data_path)\n",
    "ad = sc.read_h5ad(data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "416449f9-9b16-4da3-8714-773cd62032cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs  n_vars = 35584147  550"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "82135e2e-d763-41c0-89c1-779711fe2b47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/Dropbox/Harvard/Science/scDNAm/data/GSE121436/GSE121436_wmeta_mm10thresh_5_common_cpgs.h5ad\n"
     ]
    }
   ],
   "source": [
    "dump_key=\"thresh_5_common_cpgs\"\n",
    "data_path = Path.joinpath(cwd, project_dir, f\"GSE121436_wmeta_mm10{dump_key}.h5ad\")\n",
    "print(data_path)\n",
    "ad = sc.read_h5ad(data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "510bccb1-39cc-4bab-b31b-5479847bea05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs  n_vars = 550  5999943\n",
       "    obs: 'Run', 'Age', 'Assay Type', 'AvgSpotLen', 'Bases', 'BioProject', 'BioSample', 'Bytes', 'Cell_type', 'Center Name', 'Consent', 'DATASTORE filetype', 'DATASTORE provider', 'DATASTORE region', 'Experiment', 'GEO_Accession (exp)', 'Instrument', 'LibraryLayout', 'LibrarySelection', 'LibrarySource', 'Mouse_ID', 'Organism', 'Platform', 'ReleaseDate', 'Sample Name', 'source_name', 'SRA Study', 'strain', 'Age_months', 'Data Type', '_full_binary'\n",
       "    var: '_common_cpgs_binary', '_common_cpgs_order', '_cov>30cells_binary', '_cov>30cells_order', 'r_scMiceMuscle_full_cov>30cells_meth', 'p_scMiceMuscle_full_cov>30cells_meth', 'b_scMiceMuscle_full_cov>30cells_meth_non_sign'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f9f3f162-7e6c-4c6e-ac9a-144786d66c42",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Age</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Cell_type</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>...</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>source_name</th>\n",
       "      <th>SRA Study</th>\n",
       "      <th>strain</th>\n",
       "      <th>Age_months</th>\n",
       "      <th>Data Type</th>\n",
       "      <th>_full_binary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM3436480_percentage</th>\n",
       "      <td>SRR8073536</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>933869412</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10255048</td>\n",
       "      <td>440589845</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436480</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>2 months</td>\n",
       "      <td>percentage</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3436480_coverage</th>\n",
       "      <td>SRR8073536</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>933869412</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10255048</td>\n",
       "      <td>440589845</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436480</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>2 months</td>\n",
       "      <td>coverage</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3436314_percentage</th>\n",
       "      <td>SRR8073370</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>612125640</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254952</td>\n",
       "      <td>257904041</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436314</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>24 months</td>\n",
       "      <td>percentage</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3436314_coverage</th>\n",
       "      <td>SRR8073370</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>612125640</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254952</td>\n",
       "      <td>257904041</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436314</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>24 months</td>\n",
       "      <td>coverage</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3436450_percentage</th>\n",
       "      <td>SRR8073506</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>742126140</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254990</td>\n",
       "      <td>345931510</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436450</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>2 months</td>\n",
       "      <td>percentage</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3436531_coverage</th>\n",
       "      <td>SRR8073587</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>1207480428</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10255062</td>\n",
       "      <td>589667584</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436531</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>2 months</td>\n",
       "      <td>coverage</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3436268_percentage</th>\n",
       "      <td>SRR8073324</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>250</td>\n",
       "      <td>1012848250</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254804</td>\n",
       "      <td>405102838</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436268</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>24 months</td>\n",
       "      <td>percentage</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3436268_coverage</th>\n",
       "      <td>SRR8073324</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>250</td>\n",
       "      <td>1012848250</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254804</td>\n",
       "      <td>405102838</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436268</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>24 months</td>\n",
       "      <td>coverage</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3436365_percentage</th>\n",
       "      <td>SRR8073421</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>250</td>\n",
       "      <td>1348635500</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10255146</td>\n",
       "      <td>566490196</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436365</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>2 months</td>\n",
       "      <td>percentage</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM3436365_coverage</th>\n",
       "      <td>SRR8073421</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>250</td>\n",
       "      <td>1348635500</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10255146</td>\n",
       "      <td>566490196</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>GEO</td>\n",
       "      <td>...</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436365</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>2 months</td>\n",
       "      <td>coverage</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>550 rows  31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              Run   Age     Assay Type  AvgSpotLen  \\\n",
       "GSM3436480_percentage  SRR8073536   2.0  Bisulfite-Seq         252   \n",
       "GSM3436480_coverage    SRR8073536   2.0  Bisulfite-Seq         252   \n",
       "GSM3436314_percentage  SRR8073370  24.0  Bisulfite-Seq         252   \n",
       "GSM3436314_coverage    SRR8073370  24.0  Bisulfite-Seq         252   \n",
       "GSM3436450_percentage  SRR8073506   2.0  Bisulfite-Seq         252   \n",
       "...                           ...   ...            ...         ...   \n",
       "GSM3436531_coverage    SRR8073587   2.0  Bisulfite-Seq         252   \n",
       "GSM3436268_percentage  SRR8073324  24.0  Bisulfite-Seq         250   \n",
       "GSM3436268_coverage    SRR8073324  24.0  Bisulfite-Seq         250   \n",
       "GSM3436365_percentage  SRR8073421   2.0  Bisulfite-Seq         250   \n",
       "GSM3436365_coverage    SRR8073421   2.0  Bisulfite-Seq         250   \n",
       "\n",
       "                            Bases   BioProject     BioSample      Bytes  \\\n",
       "GSM3436480_percentage   933869412  PRJNA497312  SAMN10255048  440589845   \n",
       "GSM3436480_coverage     933869412  PRJNA497312  SAMN10255048  440589845   \n",
       "GSM3436314_percentage   612125640  PRJNA497312  SAMN10254952  257904041   \n",
       "GSM3436314_coverage     612125640  PRJNA497312  SAMN10254952  257904041   \n",
       "GSM3436450_percentage   742126140  PRJNA497312  SAMN10254990  345931510   \n",
       "...                           ...          ...           ...        ...   \n",
       "GSM3436531_coverage    1207480428  PRJNA497312  SAMN10255062  589667584   \n",
       "GSM3436268_percentage  1012848250  PRJNA497312  SAMN10254804  405102838   \n",
       "GSM3436268_coverage    1012848250  PRJNA497312  SAMN10254804  405102838   \n",
       "GSM3436365_percentage  1348635500  PRJNA497312  SAMN10255146  566490196   \n",
       "GSM3436365_coverage    1348635500  PRJNA497312  SAMN10255146  566490196   \n",
       "\n",
       "                               Cell_type Center Name  ...      Organism  \\\n",
       "GSM3436480_percentage  Muscle stem cells         GEO  ...  Mus musculus   \n",
       "GSM3436480_coverage    Muscle stem cells         GEO  ...  Mus musculus   \n",
       "GSM3436314_percentage  Muscle stem cells         GEO  ...  Mus musculus   \n",
       "GSM3436314_coverage    Muscle stem cells         GEO  ...  Mus musculus   \n",
       "GSM3436450_percentage  Muscle stem cells         GEO  ...  Mus musculus   \n",
       "...                                  ...         ...  ...           ...   \n",
       "GSM3436531_coverage    Muscle stem cells         GEO  ...  Mus musculus   \n",
       "GSM3436268_percentage  Muscle stem cells         GEO  ...  Mus musculus   \n",
       "GSM3436268_coverage    Muscle stem cells         GEO  ...  Mus musculus   \n",
       "GSM3436365_percentage  Muscle stem cells         GEO  ...  Mus musculus   \n",
       "GSM3436365_coverage    Muscle stem cells         GEO  ...  Mus musculus   \n",
       "\n",
       "                       Platform           ReleaseDate Sample Name  \\\n",
       "GSM3436480_percentage  ILLUMINA  2019-08-27T00:00:00Z  GSM3436480   \n",
       "GSM3436480_coverage    ILLUMINA  2019-08-27T00:00:00Z  GSM3436480   \n",
       "GSM3436314_percentage  ILLUMINA  2019-08-27T00:00:00Z  GSM3436314   \n",
       "GSM3436314_coverage    ILLUMINA  2019-08-27T00:00:00Z  GSM3436314   \n",
       "GSM3436450_percentage  ILLUMINA  2019-08-27T00:00:00Z  GSM3436450   \n",
       "...                         ...                   ...         ...   \n",
       "GSM3436531_coverage    ILLUMINA  2019-08-27T00:00:00Z  GSM3436531   \n",
       "GSM3436268_percentage  ILLUMINA  2019-08-27T00:00:00Z  GSM3436268   \n",
       "GSM3436268_coverage    ILLUMINA  2019-08-27T00:00:00Z  GSM3436268   \n",
       "GSM3436365_percentage  ILLUMINA  2019-08-27T00:00:00Z  GSM3436365   \n",
       "GSM3436365_coverage    ILLUMINA  2019-08-27T00:00:00Z  GSM3436365   \n",
       "\n",
       "                                     source_name  SRA Study        strain  \\\n",
       "GSM3436480_percentage  Tibialis anterior muscles  SRP166031  MGI: 5308730   \n",
       "GSM3436480_coverage    Tibialis anterior muscles  SRP166031  MGI: 5308730   \n",
       "GSM3436314_percentage  Tibialis anterior muscles  SRP166031  MGI: 5308730   \n",
       "GSM3436314_coverage    Tibialis anterior muscles  SRP166031  MGI: 5308730   \n",
       "GSM3436450_percentage  Tibialis anterior muscles  SRP166031  MGI: 5308730   \n",
       "...                                          ...        ...           ...   \n",
       "GSM3436531_coverage    Tibialis anterior muscles  SRP166031  MGI: 5308730   \n",
       "GSM3436268_percentage  Tibialis anterior muscles  SRP166031  MGI: 5308730   \n",
       "GSM3436268_coverage    Tibialis anterior muscles  SRP166031  MGI: 5308730   \n",
       "GSM3436365_percentage  Tibialis anterior muscles  SRP166031  MGI: 5308730   \n",
       "GSM3436365_coverage    Tibialis anterior muscles  SRP166031  MGI: 5308730   \n",
       "\n",
       "                      Age_months   Data Type _full_binary  \n",
       "GSM3436480_percentage   2 months  percentage         True  \n",
       "GSM3436480_coverage     2 months    coverage         True  \n",
       "GSM3436314_percentage  24 months  percentage         True  \n",
       "GSM3436314_coverage    24 months    coverage         True  \n",
       "GSM3436450_percentage   2 months  percentage         True  \n",
       "...                          ...         ...          ...  \n",
       "GSM3436531_coverage     2 months    coverage         True  \n",
       "GSM3436268_percentage  24 months  percentage         True  \n",
       "GSM3436268_coverage    24 months    coverage         True  \n",
       "GSM3436365_percentage   2 months  percentage         True  \n",
       "GSM3436365_coverage     2 months    coverage         True  \n",
       "\n",
       "[550 rows x 31 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "71c10ea6-a962-4f35-9f24-b08f05a15b11",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad.obs = pd.concat((ad.obs, df_meta.loc[ad.obs.index,:]), axis=1)\n",
    "ad.X = ad.X.astype(np.float32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "875c2df2-fd59-4ea5-835e-4b5e7693c100",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "      <td>1</td>\n",
       "      <td>3205901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm1992</th>\n",
       "      <td>1</td>\n",
       "      <td>3466587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rp1</th>\n",
       "      <td>1</td>\n",
       "      <td>4343507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sox17</th>\n",
       "      <td>1</td>\n",
       "      <td>4490928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mrpl15</th>\n",
       "      <td>1</td>\n",
       "      <td>4773206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC151712.3</th>\n",
       "      <td>Y</td>\n",
       "      <td>90754513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC151712.1</th>\n",
       "      <td>Y</td>\n",
       "      <td>90755657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC151712.4</th>\n",
       "      <td>Y</td>\n",
       "      <td>90784738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC151712.5</th>\n",
       "      <td>Y</td>\n",
       "      <td>90838869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC151712.2</th>\n",
       "      <td>Y</td>\n",
       "      <td>90843934</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25952 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           chr     start\n",
       "geneName                \n",
       "Xkr4         1   3205901\n",
       "Gm1992       1   3466587\n",
       "Rp1          1   4343507\n",
       "Sox17        1   4490928\n",
       "Mrpl15       1   4773206\n",
       "...         ..       ...\n",
       "AC151712.3   Y  90754513\n",
       "AC151712.1   Y  90755657\n",
       "AC151712.4   Y  90784738\n",
       "AC151712.5   Y  90838869\n",
       "AC151712.2   Y  90843934\n",
       "\n",
       "[25952 rows x 2 columns]"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ad.obs[['Age', 'Mouse_ID']]\n",
    "ad.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "85d818e7-2689-473a-a716-2f1a3e8c457e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dump_key=\"thresh_5_common_cpgs\"\n",
    "data_path = Path.joinpath(cwd, project_dir, f\"GSE121436_wmeta_mm10_with_CellID_{dump_key}.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2dc45a76-6345-42bd-802f-c9e06cadf91c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Run' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Age' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Assay Type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'AvgSpotLen' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Bases' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'BioProject' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'BioSample' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Bytes' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Cell_type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Center Name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Consent' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE filetype' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE provider' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE region' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Experiment' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'GEO_Accession (exp)' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Instrument' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibraryLayout' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibrarySelection' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibrarySource' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Mouse_ID' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Organism' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'ReleaseDate' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Sample Name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'source_name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'SRA Study' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'strain' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Age_months' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Data Type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Run' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Age' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Assay Type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'AvgSpotLen' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Bases' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'BioProject' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'BioSample' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Bytes' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Cell_type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Center Name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Consent' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE filetype' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE provider' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE region' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Experiment' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'GEO_Accession (exp)' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Instrument' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibraryLayout' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibrarySelection' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibrarySource' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Mouse_ID' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Organism' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Platform_x' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'ReleaseDate' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Sample Name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'source_name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'SRA Study' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'strain' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Accession' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Title' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Sample Type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Taxonomy' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Platform_y' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Series' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Supplementary Types' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Supplementary Links' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'SRA Accession' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Contact' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Release Date' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'CellID' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Age_months' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Data Type' as categorical\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Unable to create dataset (name already exists)\n\nAbove error raised while writing key 'Run' of <class 'h5py._hl.group.Group'> from /.\n\nAbove error raised while writing key 'obs' of <class 'h5py._hl.files.File'> from /.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "File \u001b[0;32m~/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_io/utils.py:209\u001b[0m, in \u001b[0;36mreport_write_key_on_error.<locals>.func_wrapper\u001b[0;34m(elem, key, val, *args, **kwargs)\u001b[0m\n\u001b[1;32m    208\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 209\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43melem\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mval\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    210\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_io/h5ad.py:270\u001b[0m, in \u001b[0;36mwrite_series\u001b[0;34m(group, key, series, dataset_kwargs)\u001b[0m\n\u001b[1;32m    269\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m series\u001b[38;5;241m.\u001b[39mdtype \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mobject\u001b[39m:  \u001b[38;5;66;03m# Assuming its string\u001b[39;00m\n\u001b[0;32m--> 270\u001b[0m     \u001b[43mgroup\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate_dataset\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    271\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    272\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdata\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mseries\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    273\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdtype\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mh5py\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mspecial_dtype\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvlen\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    274\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mdataset_kwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    275\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    276\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m is_categorical_dtype(series):\n\u001b[1;32m    277\u001b[0m     \u001b[38;5;66;03m# This should work for categorical Index and Series\u001b[39;00m\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/py310/lib/python3.10/site-packages/h5py/_hl/group.py:149\u001b[0m, in \u001b[0;36mGroup.create_dataset\u001b[0;34m(self, name, shape, dtype, data, **kwds)\u001b[0m\n\u001b[1;32m    147\u001b[0m         group \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrequire_group(parent_path)\n\u001b[0;32m--> 149\u001b[0m dsid \u001b[38;5;241m=\u001b[39m \u001b[43mdataset\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmake_new_dset\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgroup\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mshape\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    150\u001b[0m dset \u001b[38;5;241m=\u001b[39m dataset\u001b[38;5;241m.\u001b[39mDataset(dsid)\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/py310/lib/python3.10/site-packages/h5py/_hl/dataset.py:142\u001b[0m, in \u001b[0;36mmake_new_dset\u001b[0;34m(parent, shape, dtype, data, name, chunks, compression, shuffle, fletcher32, maxshape, compression_opts, fillvalue, scaleoffset, track_times, external, track_order, dcpl, allow_unknown_filter)\u001b[0m\n\u001b[1;32m    139\u001b[0m     sid \u001b[38;5;241m=\u001b[39m h5s\u001b[38;5;241m.\u001b[39mcreate_simple(shape, maxshape)\n\u001b[0;32m--> 142\u001b[0m dset_id \u001b[38;5;241m=\u001b[39m \u001b[43mh5d\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate\u001b[49m\u001b[43m(\u001b[49m\u001b[43mparent\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mid\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtid\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msid\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdcpl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdcpl\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    144\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (data \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;129;01mand\u001b[39;00m (\u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(data, Empty)):\n",
      "File \u001b[0;32mh5py/_objects.pyx:54\u001b[0m, in \u001b[0;36mh5py._objects.with_phil.wrapper\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mh5py/_objects.pyx:55\u001b[0m, in \u001b[0;36mh5py._objects.with_phil.wrapper\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mh5py/h5d.pyx:87\u001b[0m, in \u001b[0;36mh5py.h5d.create\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Unable to create dataset (name already exists)",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "File \u001b[0;32m~/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_io/utils.py:209\u001b[0m, in \u001b[0;36mreport_write_key_on_error.<locals>.func_wrapper\u001b[0;34m(elem, key, val, *args, **kwargs)\u001b[0m\n\u001b[1;32m    208\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 209\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43melem\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mval\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    210\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_io/h5ad.py:263\u001b[0m, in \u001b[0;36mwrite_dataframe\u001b[0;34m(f, key, df, dataset_kwargs)\u001b[0m\n\u001b[1;32m    262\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m col_name, (_, series) \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(col_names, df\u001b[38;5;241m.\u001b[39mitems()):\n\u001b[0;32m--> 263\u001b[0m     \u001b[43mwrite_series\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgroup\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcol_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mseries\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdataset_kwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdataset_kwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_io/utils.py:212\u001b[0m, in \u001b[0;36mreport_write_key_on_error.<locals>.func_wrapper\u001b[0;34m(elem, key, val, *args, **kwargs)\u001b[0m\n\u001b[1;32m    211\u001b[0m parent \u001b[38;5;241m=\u001b[39m _get_parent(elem)\n\u001b[0;32m--> 212\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;28mtype\u001b[39m(e)(\n\u001b[1;32m    213\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00me\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    214\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAbove error raised while writing key \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m of \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(elem)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    215\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m from \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mparent\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    216\u001b[0m ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01me\u001b[39;00m\n",
      "\u001b[0;31mValueError\u001b[0m: Unable to create dataset (name already exists)\n\nAbove error raised while writing key 'Run' of <class 'h5py._hl.group.Group'> from /.",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[0;32mIn [16]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m ad\u001b[38;5;241m.\u001b[39mobs_names_make_unique()\n\u001b[1;32m      2\u001b[0m ad\u001b[38;5;241m.\u001b[39mvar_names_make_unique()\n\u001b[0;32m----> 3\u001b[0m \u001b[43mad\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrite_h5ad\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata_path\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1912\u001b[0m, in \u001b[0;36mAnnData.write_h5ad\u001b[0;34m(self, filename, compression, compression_opts, force_dense, as_dense)\u001b[0m\n\u001b[1;32m   1909\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m filename \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1910\u001b[0m     filename \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfilename\n\u001b[0;32m-> 1912\u001b[0m \u001b[43m_write_h5ad\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1913\u001b[0m \u001b[43m    \u001b[49m\u001b[43mPath\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilename\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1914\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1915\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1916\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression_opts\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcompression_opts\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1917\u001b[0m \u001b[43m    \u001b[49m\u001b[43mforce_dense\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mforce_dense\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1918\u001b[0m \u001b[43m    \u001b[49m\u001b[43mas_dense\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mas_dense\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1919\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1921\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39misbacked:\n\u001b[1;32m   1922\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfile\u001b[38;5;241m.\u001b[39mfilename \u001b[38;5;241m=\u001b[39m filename\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_io/h5ad.py:111\u001b[0m, in \u001b[0;36mwrite_h5ad\u001b[0;34m(filepath, adata, force_dense, as_dense, dataset_kwargs, **kwargs)\u001b[0m\n\u001b[1;32m    109\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    110\u001b[0m     write_attribute(f, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mraw\u001b[39m\u001b[38;5;124m\"\u001b[39m, adata\u001b[38;5;241m.\u001b[39mraw, dataset_kwargs\u001b[38;5;241m=\u001b[39mdataset_kwargs)\n\u001b[0;32m--> 111\u001b[0m \u001b[43mwrite_attribute\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mobs\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43madata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdataset_kwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdataset_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    112\u001b[0m write_attribute(f, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvar\u001b[39m\u001b[38;5;124m\"\u001b[39m, adata\u001b[38;5;241m.\u001b[39mvar, dataset_kwargs\u001b[38;5;241m=\u001b[39mdataset_kwargs)\n\u001b[1;32m    113\u001b[0m write_attribute(f, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mobsm\u001b[39m\u001b[38;5;124m\"\u001b[39m, adata\u001b[38;5;241m.\u001b[39mobsm, dataset_kwargs\u001b[38;5;241m=\u001b[39mdataset_kwargs)\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/py310/lib/python3.10/functools.py:889\u001b[0m, in \u001b[0;36msingledispatch.<locals>.wrapper\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    885\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m args:\n\u001b[1;32m    886\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfuncname\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m requires at least \u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m    887\u001b[0m                     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m1 positional argument\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m--> 889\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mdispatch\u001b[49m\u001b[43m(\u001b[49m\u001b[43margs\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;18;43m__class__\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkw\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_io/h5ad.py:130\u001b[0m, in \u001b[0;36mwrite_attribute_h5ad\u001b[0;34m(f, key, value, *args, **kwargs)\u001b[0m\n\u001b[1;32m    128\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m key \u001b[38;5;129;01min\u001b[39;00m f:\n\u001b[1;32m    129\u001b[0m     \u001b[38;5;28;01mdel\u001b[39;00m f[key]\n\u001b[0;32m--> 130\u001b[0m \u001b[43m_write_method\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mtype\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_io/utils.py:212\u001b[0m, in \u001b[0;36mreport_write_key_on_error.<locals>.func_wrapper\u001b[0;34m(elem, key, val, *args, **kwargs)\u001b[0m\n\u001b[1;32m    210\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    211\u001b[0m     parent \u001b[38;5;241m=\u001b[39m _get_parent(elem)\n\u001b[0;32m--> 212\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;28mtype\u001b[39m(e)(\n\u001b[1;32m    213\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00me\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    214\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAbove error raised while writing key \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m of \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(elem)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    215\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m from \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mparent\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    216\u001b[0m     ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01me\u001b[39;00m\n",
      "\u001b[0;31mValueError\u001b[0m: Unable to create dataset (name already exists)\n\nAbove error raised while writing key 'Run' of <class 'h5py._hl.group.Group'> from /.\n\nAbove error raised while writing key 'obs' of <class 'h5py._hl.files.File'> from /."
     ]
    }
   ],
   "source": [
    "ad.obs_names_make_unique()\n",
    "ad.var_names_make_unique()\n",
    "ad.write_h5ad(data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c92c013-f6b4-4321-bf78-8d7bc838c9b0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "d2d054fb-603e-4ee7-85a5-061967a4a652",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import meth.tools as mtools\n",
    "# df = ad.obs.copy()\n",
    "# df.reset_index(inplace=True)\n",
    "# df['index'] = df['index'].apply(lambda x: mtools.convert_between_assemblies(x))\n",
    "# idx = df['index'].values != 'None'\n",
    "# print(np.sum(idx), idx.shape[0])\n",
    "# ad = ad[idx,:].copy()\n",
    "# df = df[df['index'] != 'None']\n",
    "# df.set_index('index', inplace=True)    \n",
    "# ad.obs = df\n",
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39c35125-7cdd-4ff2-990b-7c8766f5d4ba",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e6c8c17-3072-434b-add1-fe6768f90f8d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8b3ad70-911d-4028-8268-bf7ca8865f24",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9a2021b-82be-4f38-bd1a-07711a6287bd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "db021b16-fc09-4f4b-9693-52037ebf0621",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr10_100000122</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_100000144</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_100000272</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_100000277</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10_100000455</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY_999736</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY_999764</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY_9998876</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY_9999488</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY_9999489</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>35584147 rows  0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [chr10_100000122, chr10_100000144, chr10_100000272, chr10_100000277, chr10_100000455, chr10_10000068, chr10_10000069, chr10_100000895, chr10_100001081, chr10_100001095, chr10_100001328, chr10_100001648, chr10_100001649, chr10_100001652, chr10_100001653, chr10_100001685, chr10_100001687, chr10_100001688, chr10_100002052, chr10_100002205, chr10_100002206, chr10_100002296, chr10_100002297, chr10_100002463, chr10_100002464, chr10_100003076, chr10_100003077, chr10_100003116, chr10_100003117, chr10_100003132, chr10_100003313, chr10_100003314, chr10_100003350, chr10_100003351, chr10_100003578, chr10_100003621, chr10_100003622, chr10_100003681, chr10_100003682, chr10_100003936, chr10_100003937, chr10_100003998, chr10_100004052, chr10_100004127, chr10_100004128, chr10_100004206, chr10_10000421, chr10_10000422, chr10_100004300, chr10_100004301, chr10_100004315, chr10_100004316, chr10_100004324, chr10_100004325, chr10_100004372, chr10_100004373, chr10_100004406, chr10_100004416, chr10_100004417, chr10_100004445, chr10_100004446, chr10_100004496, chr10_100004497, chr10_100004505, chr10_100004506, chr10_100004549, chr10_100004550, chr10_100004561, chr10_100004562, chr10_100004578, chr10_100004579, chr10_100004580, chr10_100004581, chr10_100004593, chr10_100004594, chr10_100004707, chr10_100004708, chr10_100004777, chr10_100004778, chr10_100004783, chr10_100004784, chr10_100004830, chr10_100004831, chr10_100004934, chr10_100004935, chr10_100005126, chr10_100005127, chr10_100005135, chr10_100005136, chr10_100005162, chr10_100005163, chr10_100005197, chr10_100005200, chr10_100005220, chr10_10000588, chr10_10000589, chr10_10000608, chr10_10000609, chr10_100006259, chr10_100006301, ...]\n",
       "\n",
       "[35584147 rows x 0 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "4d52a7b1-a2d4-45f0-ac2b-1dffc5f10d77",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad.var = pd.concat((ad.var, df_meta.loc[ad.var.index,:]), axis=1)\n",
    "ad.X = ad.X.astype(np.float32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "948201f5-074a-4625-b24a-166048fbbfff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['O1' 'O5' 'O8' 'Y2' 'Y7' 'Y8']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/v_/gtsrj3xs0_131ltxwg51r_f40000gn/T/ipykernel_16405/3837489246.py:12: RuntimeWarning: invalid value encountered in true_divide\n",
      "  xbulk[:,2*i] = xmeth[:,idx].sum(axis=1).flatten() / xcov[:,idx].sum(axis=1).flatten()\n"
     ]
    }
   ],
   "source": [
    "x = ad.X[:,::2]\n",
    "xcov = ad.X[:,1::2]\n",
    "xmeth = x.multiply(xcov)\n",
    "\n",
    "mouse_ids = ad.var.loc[::2,:].Mouse_ID.values\n",
    "mice = np.unique(mouse_ids)\n",
    "print(mice)\n",
    "xbulk = np.zeros((ad.X.shape[0], 2* len(mice)))\n",
    "#xcovbulk = np.zeros((ad.X.shape[0], len(mice)))\n",
    "for i,mouse_id in enumerate(mice):\n",
    "    idx = mouse_ids == mouse_id\n",
    "    xbulk[:,2*i] = xmeth[:,idx].sum(axis=1).flatten() / xcov[:,idx].sum(axis=1).flatten()\n",
    "    xbulk[:,2*i+1] = xcov[:,idx].sum(axis=1).flatten()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "51402978-c8bf-414c-9d80-ee4e5f4ad08f",
   "metadata": {},
   "outputs": [],
   "source": [
    "mice_index = np.array([[f\"{x}_percentage\", f\"{x}_coverage\"] for x in mice]).flatten()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "3ca88497-3f83-4d5f-bb90-25b8651260b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0. , 1. , nan, ..., 0. , nan, 0. ],\n",
       "       [1. , 1. , nan, ..., 0. , nan, 0. ],\n",
       "       [0.5, 2. , nan, ..., 2. , 1. , 1. ],\n",
       "       ...,\n",
       "       [0. , 2. , nan, ..., 0. , nan, 0. ],\n",
       "       [1. , 1. , nan, ..., 4. , nan, 0. ],\n",
       "       [nan, 0. , nan, ..., 0. , nan, 0. ]])"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "359b55d2-77ad-461a-9251-d9647e050345",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12, 30)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mouse_ID</th>\n",
       "      <th>Run</th>\n",
       "      <th>Age</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Cell_type</th>\n",
       "      <th>...</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>source_name</th>\n",
       "      <th>SRA Study</th>\n",
       "      <th>strain</th>\n",
       "      <th>Age_months</th>\n",
       "      <th>Data Type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name PerCov</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>O1_percentage</th>\n",
       "      <td>O1</td>\n",
       "      <td>SRR8073317</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>250</td>\n",
       "      <td>891398250</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254811</td>\n",
       "      <td>351440127</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>...</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436261</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>24 months</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O1_coverage</th>\n",
       "      <td>O1</td>\n",
       "      <td>SRR8073317</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>250</td>\n",
       "      <td>891398250</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254811</td>\n",
       "      <td>351440127</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>...</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436261</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>24 months</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O5_percentage</th>\n",
       "      <td>O5</td>\n",
       "      <td>SRR8073544</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>1151661168</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10255099</td>\n",
       "      <td>568791665</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>...</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436488</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>2 months</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O5_coverage</th>\n",
       "      <td>O5</td>\n",
       "      <td>SRR8073544</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>1151661168</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10255099</td>\n",
       "      <td>568791665</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>...</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436488</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>2 months</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O8_percentage</th>\n",
       "      <td>O8</td>\n",
       "      <td>SRR8073360</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>1361646216</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254897</td>\n",
       "      <td>572974528</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>...</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436304</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>24 months</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O8_coverage</th>\n",
       "      <td>O8</td>\n",
       "      <td>SRR8073360</td>\n",
       "      <td>24.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>1361646216</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254897</td>\n",
       "      <td>572974528</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>...</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436304</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>24 months</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y2_percentage</th>\n",
       "      <td>Y2</td>\n",
       "      <td>SRR8073408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>250</td>\n",
       "      <td>961632750</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10255104</td>\n",
       "      <td>404404212</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>...</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436352</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>2 months</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y2_coverage</th>\n",
       "      <td>Y2</td>\n",
       "      <td>SRR8073408</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>250</td>\n",
       "      <td>961632750</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10255104</td>\n",
       "      <td>404404212</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>...</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436352</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>2 months</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y7_percentage</th>\n",
       "      <td>Y7</td>\n",
       "      <td>SRR8073448</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>911255688</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254934</td>\n",
       "      <td>418481168</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>...</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436392</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>2 months</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y7_coverage</th>\n",
       "      <td>Y7</td>\n",
       "      <td>SRR8073448</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>911255688</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10254934</td>\n",
       "      <td>418481168</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>...</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436392</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>2 months</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y8_percentage</th>\n",
       "      <td>Y8</td>\n",
       "      <td>SRR8073496</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>495294408</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10255000</td>\n",
       "      <td>226573086</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>...</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436440</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>2 months</td>\n",
       "      <td>percentage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y8_coverage</th>\n",
       "      <td>Y8</td>\n",
       "      <td>SRR8073496</td>\n",
       "      <td>2.0</td>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>252</td>\n",
       "      <td>495294408</td>\n",
       "      <td>PRJNA497312</td>\n",
       "      <td>SAMN10255000</td>\n",
       "      <td>226573086</td>\n",
       "      <td>Muscle stem cells</td>\n",
       "      <td>...</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>Mus musculus</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2019-08-27T00:00:00Z</td>\n",
       "      <td>GSM3436440</td>\n",
       "      <td>Tibialis anterior muscles</td>\n",
       "      <td>SRP166031</td>\n",
       "      <td>MGI: 5308730</td>\n",
       "      <td>2 months</td>\n",
       "      <td>coverage</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12 rows  30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Mouse_ID         Run   Age     Assay Type  AvgSpotLen  \\\n",
       "Name PerCov                                                           \n",
       "O1_percentage       O1  SRR8073317  24.0  Bisulfite-Seq         250   \n",
       "O1_coverage         O1  SRR8073317  24.0  Bisulfite-Seq         250   \n",
       "O5_percentage       O5  SRR8073544   2.0  Bisulfite-Seq         252   \n",
       "O5_coverage         O5  SRR8073544   2.0  Bisulfite-Seq         252   \n",
       "O8_percentage       O8  SRR8073360  24.0  Bisulfite-Seq         252   \n",
       "O8_coverage         O8  SRR8073360  24.0  Bisulfite-Seq         252   \n",
       "Y2_percentage       Y2  SRR8073408   2.0  Bisulfite-Seq         250   \n",
       "Y2_coverage         Y2  SRR8073408   2.0  Bisulfite-Seq         250   \n",
       "Y7_percentage       Y7  SRR8073448   2.0  Bisulfite-Seq         252   \n",
       "Y7_coverage         Y7  SRR8073448   2.0  Bisulfite-Seq         252   \n",
       "Y8_percentage       Y8  SRR8073496   2.0  Bisulfite-Seq         252   \n",
       "Y8_coverage         Y8  SRR8073496   2.0  Bisulfite-Seq         252   \n",
       "\n",
       "                    Bases   BioProject     BioSample      Bytes  \\\n",
       "Name PerCov                                                       \n",
       "O1_percentage   891398250  PRJNA497312  SAMN10254811  351440127   \n",
       "O1_coverage     891398250  PRJNA497312  SAMN10254811  351440127   \n",
       "O5_percentage  1151661168  PRJNA497312  SAMN10255099  568791665   \n",
       "O5_coverage    1151661168  PRJNA497312  SAMN10255099  568791665   \n",
       "O8_percentage  1361646216  PRJNA497312  SAMN10254897  572974528   \n",
       "O8_coverage    1361646216  PRJNA497312  SAMN10254897  572974528   \n",
       "Y2_percentage   961632750  PRJNA497312  SAMN10255104  404404212   \n",
       "Y2_coverage     961632750  PRJNA497312  SAMN10255104  404404212   \n",
       "Y7_percentage   911255688  PRJNA497312  SAMN10254934  418481168   \n",
       "Y7_coverage     911255688  PRJNA497312  SAMN10254934  418481168   \n",
       "Y8_percentage   495294408  PRJNA497312  SAMN10255000  226573086   \n",
       "Y8_coverage     495294408  PRJNA497312  SAMN10255000  226573086   \n",
       "\n",
       "                       Cell_type  ... LibrarySource      Organism  Platform  \\\n",
       "Name PerCov                       ...                                         \n",
       "O1_percentage  Muscle stem cells  ...       GENOMIC  Mus musculus  ILLUMINA   \n",
       "O1_coverage    Muscle stem cells  ...       GENOMIC  Mus musculus  ILLUMINA   \n",
       "O5_percentage  Muscle stem cells  ...       GENOMIC  Mus musculus  ILLUMINA   \n",
       "O5_coverage    Muscle stem cells  ...       GENOMIC  Mus musculus  ILLUMINA   \n",
       "O8_percentage  Muscle stem cells  ...       GENOMIC  Mus musculus  ILLUMINA   \n",
       "O8_coverage    Muscle stem cells  ...       GENOMIC  Mus musculus  ILLUMINA   \n",
       "Y2_percentage  Muscle stem cells  ...       GENOMIC  Mus musculus  ILLUMINA   \n",
       "Y2_coverage    Muscle stem cells  ...       GENOMIC  Mus musculus  ILLUMINA   \n",
       "Y7_percentage  Muscle stem cells  ...       GENOMIC  Mus musculus  ILLUMINA   \n",
       "Y7_coverage    Muscle stem cells  ...       GENOMIC  Mus musculus  ILLUMINA   \n",
       "Y8_percentage  Muscle stem cells  ...       GENOMIC  Mus musculus  ILLUMINA   \n",
       "Y8_coverage    Muscle stem cells  ...       GENOMIC  Mus musculus  ILLUMINA   \n",
       "\n",
       "                        ReleaseDate Sample Name                source_name  \\\n",
       "Name PerCov                                                                  \n",
       "O1_percentage  2019-08-27T00:00:00Z  GSM3436261  Tibialis anterior muscles   \n",
       "O1_coverage    2019-08-27T00:00:00Z  GSM3436261  Tibialis anterior muscles   \n",
       "O5_percentage  2019-08-27T00:00:00Z  GSM3436488  Tibialis anterior muscles   \n",
       "O5_coverage    2019-08-27T00:00:00Z  GSM3436488  Tibialis anterior muscles   \n",
       "O8_percentage  2019-08-27T00:00:00Z  GSM3436304  Tibialis anterior muscles   \n",
       "O8_coverage    2019-08-27T00:00:00Z  GSM3436304  Tibialis anterior muscles   \n",
       "Y2_percentage  2019-08-27T00:00:00Z  GSM3436352  Tibialis anterior muscles   \n",
       "Y2_coverage    2019-08-27T00:00:00Z  GSM3436352  Tibialis anterior muscles   \n",
       "Y7_percentage  2019-08-27T00:00:00Z  GSM3436392  Tibialis anterior muscles   \n",
       "Y7_coverage    2019-08-27T00:00:00Z  GSM3436392  Tibialis anterior muscles   \n",
       "Y8_percentage  2019-08-27T00:00:00Z  GSM3436440  Tibialis anterior muscles   \n",
       "Y8_coverage    2019-08-27T00:00:00Z  GSM3436440  Tibialis anterior muscles   \n",
       "\n",
       "               SRA Study        strain Age_months   Data Type  \n",
       "Name PerCov                                                    \n",
       "O1_percentage  SRP166031  MGI: 5308730  24 months  percentage  \n",
       "O1_coverage    SRP166031  MGI: 5308730  24 months    coverage  \n",
       "O5_percentage  SRP166031  MGI: 5308730   2 months  percentage  \n",
       "O5_coverage    SRP166031  MGI: 5308730   2 months    coverage  \n",
       "O8_percentage  SRP166031  MGI: 5308730  24 months  percentage  \n",
       "O8_coverage    SRP166031  MGI: 5308730  24 months    coverage  \n",
       "Y2_percentage  SRP166031  MGI: 5308730   2 months  percentage  \n",
       "Y2_coverage    SRP166031  MGI: 5308730   2 months    coverage  \n",
       "Y7_percentage  SRP166031  MGI: 5308730   2 months  percentage  \n",
       "Y7_coverage    SRP166031  MGI: 5308730   2 months    coverage  \n",
       "Y8_percentage  SRP166031  MGI: 5308730   2 months  percentage  \n",
       "Y8_coverage    SRP166031  MGI: 5308730   2 months    coverage  \n",
       "\n",
       "[12 rows x 30 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta_bulk = df_meta.drop_duplicates('Mouse_ID').set_index('Mouse_ID').loc[mice].copy()\n",
    "df_meta_bulk.reset_index(inplace=True)\n",
    "df_meta_2 = df_meta_bulk.copy()\n",
    "df_meta_bulk['Name PerCov'] = df_meta_bulk['Mouse_ID'].str.cat(['percentage'] * df_meta_bulk.shape[0], sep='_')\n",
    "df_meta_bulk['Data Type'] = 'percentage'\n",
    "df_meta_2['Name PerCov'] = df_meta_2['Mouse_ID'].str.cat(['coverage'] * df_meta_bulk.shape[0], sep='_')\n",
    "df_meta_2['Data Type'] = 'coverage'\n",
    "#df_meta['Name PerCov'] = df_meta['Sample Name'].str.cat(['percentage'] * df_meta.shape[0], sep='_')\n",
    "df_meta_bulk = pd.concat((df_meta_bulk, df_meta_2), axis=0)\n",
    "df_meta_bulk.set_index('Name PerCov', inplace=True)\n",
    "print(df_meta_bulk.shape)\n",
    "df_meta_bulk = df_meta_bulk.loc[mice_index]\n",
    "df_meta_bulk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6d353494-940a-4311-8893-18ac1fd86545",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Mouse_ID' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Run' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Assay Type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'BioProject' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'BioSample' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Cell_type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Center Name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Consent' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE filetype' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE provider' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE region' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Experiment' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'GEO_Accession (exp)' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Instrument' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibraryLayout' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibrarySelection' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibrarySource' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Organism' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Platform' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'ReleaseDate' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Sample Name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'source_name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'SRA Study' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'strain' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Age_months' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Data Type' as categorical\n"
     ]
    }
   ],
   "source": [
    "adbu = sc.AnnData(sp.csr_matrix(np.nan_to_num(xbulk)), ad.obs, df_meta_bulk)\n",
    "adbu.obs_names_make_unique()\n",
    "adbu.var_names_make_unique()\n",
    "\n",
    "adbu.write_h5ad(Path.joinpath(project_dir, \"{}_bulked_wmeta_mm10.h5ad\".format(project_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "8e032e3c-d6a4-491c-bdff-35e44199693d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[0. , 1. , 0. , ..., 0. , 0. , 0. ],\n",
       "        [1. , 1. , 0. , ..., 0. , 0. , 0. ],\n",
       "        [0.5, 2. , 0. , ..., 2. , 1. , 1. ],\n",
       "        ...,\n",
       "        [0. , 2. , 0. , ..., 0. , 0. , 0. ],\n",
       "        [1. , 1. , 0. , ..., 4. , 0. , 0. ],\n",
       "        [0. , 0. , 0. , ..., 0. , 0. , 0. ]], dtype=float32)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f6e267a8-ec41-47ad-b749-f7aae3a68c4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad.obs_names_make_unique()\n",
    "ad.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e36df4bb-7bd1-4c0b-b00f-a9c8bc96426e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Run' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Assay Type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'BioProject' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'BioSample' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Cell_type' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Center Name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Consent' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE filetype' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE provider' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'DATASTORE region' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Experiment' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'GEO_Accession (exp)' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Instrument' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibraryLayout' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibrarySelection' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'LibrarySource' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Mouse_ID' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Organism' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Platform' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'ReleaseDate' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Sample Name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'source_name' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'SRA Study' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'strain' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Age_months' as categorical\n",
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/anndata/_core/anndata.py:1228: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Reordering categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Data Type' as categorical\n"
     ]
    }
   ],
   "source": [
    "ad.write_h5ad(Path.joinpath(project_dir, \"{}_wmeta.h5ad\".format(project_id)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "cd97828e-7896-4456-999d-7a11925c4432",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<987360x550 sparse matrix of type '<class 'numpy.float32'>'\n",
       "\twith 2773548 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ad.obs.reset_index(inplace=True)\n",
    "# ad.obs['index'] = ad.obs['index'].apply(lambda x: \"chr\" + x)\n",
    "# ad.obs.set_index('index', inplace=True)\n",
    "\n",
    "# ad.var.head()\n",
    "ad.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "193a5098-d4a9-48da-8d8a-12f5b606d38c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ad.var[ad.var['Age'] == 2.0]['Mouse_ID'].unique()\n",
    "# ad.var[ad.var['Age'] == 24.0]['Mouse_ID'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c22fb516-3542-4cf7-a001-365b75715a11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<100x2 sparse matrix of type '<class 'numpy.float32'>'\n",
       "\twith 21 stored elements in Compressed Sparse Row format>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad.X[:100,:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6731a84d-aa83-4ba4-88a0-90475df58fc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyliftover import LiftOver\n",
    "\n",
    "lo_mm9_to_mm10 = LiftOver('meth/mm9ToMm10.over.chain.gz')\n",
    "lo_mm10_to_mm39 = LiftOver('meth/mm10ToMm39.over.chain.gz')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "87f6c4f5-dd00-4dc3-addb-e3b424bdbff9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pyliftover.intervaltree.IntervalTree at 0x3a55c3520>"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lo_mm9_to_mm10.chain_file.chain_index['chr1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "7dda0c40-2789-4a51-8b82-75a0700bd0fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad.write_h5ad(Path.joinpath(project_dir, \"{}_wmeta_mm10.h5ad\".format(project_id)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3649636-5d39-43d4-8b2f-4f0d72910692",
   "metadata": {},
   "source": [
    "# GSE121436  preprocessing raw files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "3259b866-6192-445a-ae0b-d5a66f337cfa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "275\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/tarkhov/opt/anaconda3/envs/py310/lib/python3.10/site-packages/IPython/core/interactiveshell.py:3251: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM3436480_percentage    float64\n",
      "GSM3436480_coverage        int64\n",
      "dtype: object\n",
      "GSM3436480_percentage    float16\n",
      "GSM3436480_coverage      float16\n",
      "dtype: object\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM3436480_percentage</th>\n",
       "      <th>GSM3436480_coverage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10_100001640</th>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_100001650</th>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_100001680</th>\n",
       "      <td>100.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_10000420</th>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_100004300</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              GSM3436480_percentage  GSM3436480_coverage\n",
       "index                                                   \n",
       "10_100001640                  100.0                  1.0\n",
       "10_100001650                  100.0                  1.0\n",
       "10_100001680                  100.0                  2.0\n",
       "10_10000420                   100.0                  1.0\n",
       "10_100004300                    0.0                  1.0"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import sys\n",
    "import pwd\n",
    "from pathlib import Path\n",
    "from tqdm import tqdm, trange\n",
    "from tqdm.contrib.concurrent import process_map\n",
    "import meth.tools as mtools\n",
    "\n",
    "cwd = Path.cwd()\n",
    "project_dir = Path('data/GSE121436')\n",
    "project_id = 'GSE121436'\n",
    "project_samples_dir = 'GSE121436_RAW'\n",
    "path_samples = Path.joinpath(project_dir, project_samples_dir)\n",
    "gl_path_samples = Path.joinpath(cwd, path_samples)\n",
    "gl_project_dir = Path.joinpath(cwd, project_dir)\n",
    "all_samples = os.listdir(gl_path_samples)\n",
    "print(len(all_samples))\n",
    "# print(all_samples)\n",
    "df_glob = None\n",
    "id_to_gsm = dict(zip([x.split('_')[1] for x in all_samples], [x.split('_')[0] for x in all_samples]))\n",
    "\n",
    "coverage = 0\n",
    "bin_by = 10\n",
    "\n",
    "sample_path_tuples = [Path.joinpath(gl_path_samples, file) for file in all_samples]\n",
    "path_sample = sample_path_tuples[0]\n",
    "df = pd.read_csv(path_sample, sep='\\t', header=None)\n",
    "gsm = path_sample.name.split('_')[0]\n",
    "df[0] = df[0].astype(str)\n",
    "df['index'] = df.iloc[:,0].str.cat(df.iloc[:,1].astype(str), sep=\"_\")\n",
    "df.set_index('index', inplace=True)\n",
    "df['{}_percentage'.format(gsm)] = df.iloc[:,3]\n",
    "df['{}_coverage'.format(gsm)] = df.iloc[:,4] + df.iloc[:,5]\n",
    "df.drop(labels=np.arange(6), axis=1, inplace=True)\n",
    "df = df[df.iloc[:,1] >= coverage]\n",
    "df = mtools.get_binned_cpgs(df, bin_by=bin_by)\n",
    "df['{}_percentage'.format(gsm)] = 1. * df['{}_percentage'.format(gsm)].values / df['{}_coverage'.format(gsm)].values\n",
    "df.dropna(axis=0, how='any', inplace=True)\n",
    "print(df.dtypes)\n",
    "df = df.astype(np.float16)\n",
    "print(df.dtypes)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "bf323f62-fc25-486d-8804-73e3b458f176",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.3.5'"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "0a062bdd-f7ee-4ed2-a74b-ce57b62b17ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = pd.DataFrame({'{}_percentage'.format(gsm) : pd.arrays.SparseArray(df['{}_percentage'.format(gsm)].values, fill_value=np.nan),\n",
    "                    '{}_coverage'.format(gsm) : pd.arrays.SparseArray(df['{}_coverage'.format(gsm)].values, fill_value=np.nan),\n",
    "                   }, index=df.index)\n",
    "# dfs = pd.DataFrame({'{}_percentage'.format(gsm) : pd.arrays.SparseArray(df['{}_percentage'.format(gsm)].values),\n",
    "#                     '{}_coverage'.format(gsm) : pd.arrays.SparseArray(df['{}_coverage'.format(gsm)].values),\n",
    "#                    }, index=df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "5ff511ab-43e4-43a7-ad19-7f3dfc42b31e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dfs2 = pd.DataFrame({'{}_percentage'.format(gsm) : pd.arrays.SparseArray(df['{}_percentage'.format(gsm)].values),\n",
    "#                     '{}_coverage'.format(gsm) : pd.arrays.SparseArray(df['{}_coverage'.format(gsm)].values),\n",
    "#                    }, index=df.index)\n",
    "dfs2 = pd.DataFrame({'{}_percentage'.format(gsm) : pd.arrays.SparseArray(df['{}_percentage'.format(gsm)].values, fill_value=np.nan),\n",
    "                    '{}_coverage'.format(gsm) : pd.arrays.SparseArray(df['{}_coverage'.format(gsm)].values, fill_value=np.nan),\n",
    "                   }, index=df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "fac781c0-a3ef-4356-9451-709bd382ab32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GSM3436314_percentage    Sparse[float64, nan]\n",
       "GSM3436314_coverage          Sparse[int64, 0]\n",
       "dtype: object"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dfs\n",
    "dfs2.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "cc2961c5-2565-417c-816d-256454007588",
   "metadata": {},
   "outputs": [],
   "source": [
    "an1 = sc.AnnData(dfs)\n",
    "an2 = sc.AnnData(dfs2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "d0485714-73c7-4d2d-97f8-56b2cac9630c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs  n_vars = 1453605  4"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "an1, an2\n",
    "anres = sc.concat((an1,an2, an3), axis=1, join='outer')\n",
    "anres"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "5911d9b5-524f-4567-8686-23592a9652d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# anres.obs, an2.obs, an1.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "51627d90-a170-47a2-8c49-cf1e018a0f4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM3436480_percentage</th>\n",
       "      <th>GSM3436480_coverage</th>\n",
       "      <th>GSM3436314_percentage</th>\n",
       "      <th>GSM3436314_coverage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10_100001640</th>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_100001650</th>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_100001680</th>\n",
       "      <td>100.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_10000420</th>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_100004300</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              GSM3436480_percentage  GSM3436480_coverage  \\\n",
       "index                                                      \n",
       "10_100001640                  100.0                  1.0   \n",
       "10_100001650                  100.0                  1.0   \n",
       "10_100001680                  100.0                  2.0   \n",
       "10_10000420                   100.0                  1.0   \n",
       "10_100004300                    0.0                  1.0   \n",
       "\n",
       "              GSM3436314_percentage  GSM3436314_coverage  \n",
       "index                                                     \n",
       "10_100001640                    NaN                  NaN  \n",
       "10_100001650                    NaN                  NaN  \n",
       "10_100001680                    NaN                  NaN  \n",
       "10_10000420                     NaN                  NaN  \n",
       "10_100004300                    NaN                  NaN  "
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfres = pd.concat((dfs, dfs2), axis=1)\n",
    "dfres2 = dfres.fillna(0)\n",
    "dfres.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "6efa5aab-db7f-4408-879f-68e7d9eafa6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM3436480_percentage</th>\n",
       "      <th>GSM3436480_coverage</th>\n",
       "      <th>GSM3436314_percentage</th>\n",
       "      <th>GSM3436314_coverage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10_100001640</th>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_100001650</th>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_100001680</th>\n",
       "      <td>100.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_10000420</th>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_100004300</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              GSM3436480_percentage  GSM3436480_coverage  \\\n",
       "index                                                      \n",
       "10_100001640                  100.0                  1.0   \n",
       "10_100001650                  100.0                  1.0   \n",
       "10_100001680                  100.0                  2.0   \n",
       "10_10000420                   100.0                  1.0   \n",
       "10_100004300                    0.0                  1.0   \n",
       "\n",
       "              GSM3436314_percentage  GSM3436314_coverage  \n",
       "index                                                     \n",
       "10_100001640                    0.0                  0.0  \n",
       "10_100001650                    0.0                  0.0  \n",
       "10_100001680                    0.0                  0.0  \n",
       "10_10000420                     0.0                  0.0  \n",
       "10_100004300                    0.0                  0.0  "
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfres2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "f3b0d28d-5347-4def-9f1e-5ec3e5c68e3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, ...]\n",
       "Fill: nan\n",
       "IntIndex\n",
       "Indices: array([   1382,    2945,    3130, ..., 1453602, 1453603, 1453604],\n",
       "      dtype=int32)"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfres['GSM3436314_coverage'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "5360820f-521d-4d67-b3bc-218fe9c6897e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...]\n",
       "Fill: 0\n",
       "IntIndex\n",
       "Indices: array([   1382,    2945,    3130, ..., 1453602, 1453603, 1453604],\n",
       "      dtype=int32)"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfres2['GSM3436314_coverage'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "c49aeaf5-0e7d-494f-b9cd-c2ba0d4e56ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GSM3436480_percentage    Sparse[float64, 0]\n",
       "GSM3436480_coverage      Sparse[float64, 0]\n",
       "GSM3436314_percentage    Sparse[float64, 0]\n",
       "GSM3436314_coverage      Sparse[float64, 0]\n",
       "dtype: object"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfres2.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "f5a8e662-68c5-4d04-b473-8c508fc37342",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GSM3436480_percentage    Sparse[float64, nan]\n",
       "GSM3436480_coverage      Sparse[float64, nan]\n",
       "GSM3436314_percentage    Sparse[float64, nan]\n",
       "GSM3436314_coverage      Sparse[float64, nan]\n",
       "dtype: object"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfres.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "e06a281f-db51-4344-9866-094474344e77",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df sparse size = 62839836\n",
      "df2 sparse size = 196816\n",
      "dfsum sparse size = 46534320\n",
      "dfsum2 sparse size = 46534320\n"
     ]
    }
   ],
   "source": [
    "print(\"df sparse size =\", dfs.memory_usage().sum())\n",
    "print(\"df2 sparse size =\", dfs2.memory_usage().sum())\n",
    "print(\"dfsum sparse size =\", dfres.memory_usage().sum())\n",
    "print(\"dfsum2 sparse size =\", dfres2.memory_usage().sum())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "ae6d89c7-074c-4daf-bfac-126a8aa24ad6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df sparse size = 62839836\n",
      "df2 sparse size = 196816\n",
      "dfsum sparse size = 46534320\n",
      "dfsum2 sparse size = 46534320\n"
     ]
    }
   ],
   "source": [
    "print(\"df sparse size =\", dfs.memory_usage().sum())\n",
    "print(\"df2 sparse size =\", dfs2.memory_usage().sum())\n",
    "print(\"dfsum sparse size =\", dfres.memory_usage().sum())\n",
    "print(\"dfsum2 sparse size =\", dfres2.memory_usage().sum())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "97a3663f-458b-4c9a-bd05-2bdb34cfe4b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Dat\n",
       "100    0\n",
       "101    1\n",
       "102    2\n",
       "103    3\n",
       "104    4"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(np.arange(10), columns=['Dat'])\n",
    "df.index = np.arange(10) + 100\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "70bd596c-b402-4ae1-b451-df9003f005d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dat_sp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Dat_sp\n",
       "0       0\n",
       "1       1\n",
       "2       2\n",
       "3       3\n",
       "4       4"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs = pd.DataFrame({'Dat_sp': pd.arrays.SparseArray(df['Dat'].values)},\n",
    "                  )#index=df.index)\n",
    "dfs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90db8172-c705-4e4e-8b26-3800d0f0c1fb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python3.10 (biology)",
   "language": "python",
   "name": "py310"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
